// @project-sunbird/common-consumption-v8/esm5/lib/content-details/toc-child-item/toc-child-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { TocCardType } from '../../card/models';
var TocChildItemComponent = /** @class */ (function () {
    function TocChildItemComponent() {
        this.activeMimeTypeFilter = ['all'];
        this.type = TocCardType.TEXTBOOK;
        this.contentStatus = [];
        this.disabled = '';
        this.scoreLabel = '';
        this.tocCardClick = new EventEmitter();
        this.playButtonClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    TocChildItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} activeMimeType
     * @param {?} mimeType
     * @param {?} content
     * @return {?}
     */
    TocChildItemComponent.prototype.hasMimeType = /**
     * @param {?} activeMimeType
     * @param {?} mimeType
     * @param {?} content
     * @return {?}
     */
    function (activeMimeType, mimeType, content) {
        if (!activeMimeType) {
            return true;
        }
        else {
            if (activeMimeType.indexOf('all') > -1) {
                // if (content.contentData.mimeType === MimeType.COLLECTION && !content.children) {
                //     return false;
                // }
                return true;
            }
            return !!activeMimeType.find((/**
             * @param {?} m
             * @return {?}
             */
            function (m) { return m === mimeType; }));
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TocChildItemComponent.prototype.tocCardClickHandler = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.tocCardClick.emit(tslib_1.__assign({}, event));
    };
    /**
     * @param {?} tocData
     * @param {?} item
     * @return {?}
     */
    TocChildItemComponent.prototype.createUniqueId = /**
     * @param {?} tocData
     * @param {?} item
     * @return {?}
     */
    function (tocData, item) {
        item['sbUniqueIdentifier'] = tocData.identifier + '_' + item.identifier;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TocChildItemComponent.prototype.onPlayButtonClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.playButtonClick.emit(tslib_1.__assign({}, event));
    };
    TocChildItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-toc-child-item',
                    template: "<ng-container *ngIf=\"childData | hasMimeType: activeMimeTypeFilter\">\r\n  <div *ngIf=\"childData?.mimeType === 'application/vnd.ekstep.content-collection' && childData?.children\">\r\n    <h4 class=\"sbchapter__title text-left sb__ellipsis sb__ellipsis--two ml-16 mb-8\" title=\"{{childData?.name}}\">{{childData?.name}}</h4>\r\n  </div>\r\n</ng-container>\r\n\r\n<ng-container *ngIf=\"childData?.mimeType !== 'application/vnd.ekstep.content-collection' && hasMimeType(activeMimeTypeFilter, childData?.mimeType, childData) ||\r\n  (childData?.contentData?.trackable?.enabled === 'Yes' || childData?.trackable?.enabled === 'Yes')\">\r\n  <sb-toc-card [content]=\"childData\" (tocCardClick)=\"tocCardClickHandler($event)\" [activeContent]=\"activeContent\"\r\n    [type]=\"type\" [contentStatus]=\"contentStatus\" [playBtnConfig]=\"playBtnConfig\" (playButtonClick)=\"onPlayButtonClick($event)\"\r\n    [platform]=\"platform\" [trackableDefaultImage]=\"trackableDefaultImage\" [maxAttempts]=\"maxAttempts\" [disabled]=\"disabled\" [scoreLabel]=\"scoreLabel\">\r\n  </sb-toc-card>\r\n</ng-container>\r\n\r\n<div *ngIf=\"childData?.children && !(childData?.contentData?.trackable?.enabled === 'Yes' || childData?.trackable?.enabled === 'Yes')\">\r\n  <div *ngFor=\"let child of childData?.children; let j = index;\" [id]=\"child.identifier\">\r\n    {{ createUniqueId(childData, child) }}\r\n    <sb-toc-child-item [childData]=\"child\" [activeContent]=\"activeContent\" (tocCardClick)=\"tocCardClickHandler($event)\"\r\n      [activeMimeTypeFilter]=\"activeMimeTypeFilter\" [type]=\"type\" [contentStatus]=\"contentStatus\" [playBtnConfig]=\"playBtnConfig\" \r\n      (playButtonClick)=\"onPlayButtonClick($event)\" [platform]=\"platform\" [trackableDefaultImage]=\"trackableDefaultImage\">\r\n    </sb-toc-child-item>\r\n  </div>\r\n</div>",
                    styles: [".sbchapter__title{font-size:.875rem;line-height:1.375rem}"]
                }] }
    ];
    /** @nocollapse */
    TocChildItemComponent.ctorParameters = function () { return []; };
    TocChildItemComponent.propDecorators = {
        childData: [{ type: Input }],
        activeContent: [{ type: Input }],
        activeMimeTypeFilter: [{ type: Input }],
        type: [{ type: Input }],
        contentStatus: [{ type: Input }],
        playBtnConfig: [{ type: Input }],
        platform: [{ type: Input }],
        trackableDefaultImage: [{ type: Input }],
        maxAttempts: [{ type: Input }],
        disabled: [{ type: Input }],
        scoreLabel: [{ type: Input }],
        tocCardClick: [{ type: Output }],
        playButtonClick: [{ type: Output }]
    };
    return TocChildItemComponent;
}());
export { TocChildItemComponent };
if (false) {
    /** @type {?} */
    TocChildItemComponent.prototype.childData;
    /** @type {?} */
    TocChildItemComponent.prototype.activeContent;
    /** @type {?} */
    TocChildItemComponent.prototype.activeMimeTypeFilter;
    /** @type {?} */
    TocChildItemComponent.prototype.type;
    /** @type {?} */
    TocChildItemComponent.prototype.contentStatus;
    /** @type {?} */
    TocChildItemComponent.prototype.playBtnConfig;
    /** @type {?} */
    TocChildItemComponent.prototype.platform;
    /** @type {?} */
    TocChildItemComponent.prototype.trackableDefaultImage;
    /** @type {?} */
    TocChildItemComponent.prototype.maxAttempts;
    /** @type {?} */
    TocChildItemComponent.prototype.disabled;
    /** @type {?} */
    TocChildItemComponent.prototype.scoreLabel;
    /** @type {?} */
    TocChildItemComponent.prototype.tocCardClick;
    /** @type {?} */
    TocChildItemComponent.prototype.playButtonClick;
}
