// @project-sunbird/common-consumption-v8/esm5/lib/content-details/toc-curriculum/toc-curriculum.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, ViewChildren, QueryList, Output, EventEmitter } from '@angular/core';
import { MimeTypeMasterData } from '../../pipes-module/mime-type';
import { staticData } from './toc-data';
var TocCurriculumComponent = /** @class */ (function () {
    function TocCurriculumComponent() {
        var _this = this;
        this.tocData = staticData;
        this.activeMimeTypeFilter = ['all'];
        this.noContentMessage = 'No content available';
        this.isAccordion = false;
        this.isChapterListing = false;
        this.type = undefined;
        this.contentStatus = [];
        this.shadowColor = "hsl(0, 0%, 80%)";
        this.progressColor = "#024f9d";
        this.layoutConfig = {};
        this.disabled = '';
        this.tocCardClick = new EventEmitter();
        this.noContent = new EventEmitter();
        this.scoreLabel = '';
        this.isMimeTypeFilterChanged = false;
        this.isSameMimeTypeInChildren = ((/**
         * @param {?} mimeTypesCount
         * @param {?} activeMimeType
         * @return {?}
         */
        function (mimeTypesCount, activeMimeType) {
            /** @type {?} */
            var contentMimeType = Object.keys(JSON.parse(mimeTypesCount));
            return Boolean(activeMimeType.filter((/**
             * @param {?} value
             * @return {?}
             */
            function (value) { return contentMimeType.includes(value); })).length);
        }));
        this.isSameMimeType = (/**
         * @param {?} currentContent
         * @return {?}
         */
        function (currentContent) { return _this.activeMimeTypeFilter.find((/**
         * @param {?} mimeType
         * @return {?}
         */
        function (mimeType) { return mimeType === currentContent.mimeType; })); });
        this.isChildrenPresent = (/**
         * @param {?} currentContent
         * @return {?}
         */
        function (currentContent) { return Boolean(currentContent.children && currentContent.children.length); });
    }
    Object.defineProperty(TocCurriculumComponent.prototype, "MimeTypeMasterData", {
        get: /**
         * @return {?}
         */
        function () { return MimeTypeMasterData; },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TocCurriculumComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.setActiveContent();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    TocCurriculumComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.activeMimeTypeFilter) {
            this.isMimeTypeFilterChanged = false;
        }
        else if (changes.tocData) {
            this.setActiveContent();
        }
    };
    /**
     * @return {?}
     */
    TocCurriculumComponent.prototype.setActiveContent = /**
     * @return {?}
     */
    function () {
        if (this.tocData && this.tocData.children) {
            /** @type {?} */
            var flattenDeepContents = this.flattenDeep(this.tocData.children);
            this.activeContent = this.firstNonCollectionContent(flattenDeepContents);
            if (!this.activeContent) {
                this.noContent.emit({ message: 'No Content Available' });
            }
        }
    };
    /**
     * @param {?} content
     * @return {?}
     */
    TocCurriculumComponent.prototype.filterChildren = /**
     * @param {?} content
     * @return {?}
     */
    function (content) {
        // Check for the ActiveMimeType
        if (this.activeMimeTypeFilter.includes('all')) {
            if (content.mimeType === MimeTypeMasterData.COLLECTION) {
                return this.isShowContent(content, true);
            }
            return true; // Return true for all the content except collection mimeType
        }
        else if (content.mimeType && this.isSameMimeType(content)) {
            return true;
        }
        else if (content.mimeType === MimeTypeMasterData.COLLECTION) {
            return this.isShowContent(content, false);
        }
        else {
            return false;
        }
    };
    /**
     * @private
     * @param {?} content
     * @param {?} isActiveFilterAll
     * @return {?}
     */
    TocCurriculumComponent.prototype.isShowContent = /**
     * @private
     * @param {?} content
     * @param {?} isActiveFilterAll
     * @return {?}
     */
    function (content, isActiveFilterAll) {
        var _this = this;
        if (content.mimeTypesCount && !isActiveFilterAll) {
            return this.isSameMimeTypeInChildren(content.mimeTypesCount, this.activeMimeTypeFilter);
        }
        else if (this.isChildrenPresent(content)) {
            /** @type {?} */
            var contentList = this.flattenDeep(content.children);
            return isActiveFilterAll ? contentList.some((/**
             * @param {?} c
             * @return {?}
             */
            function (c) { return c.mimeType !== MimeTypeMasterData.COLLECTION; })) :
                contentList.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return _this.isSameMimeType(c); }));
        }
        else {
            return false;
        }
    };
    /**
     * @param {?} event
     * @param {?} item
     * @return {?}
     */
    TocCurriculumComponent.prototype.onTocChapterClick = /**
     * @param {?} event
     * @param {?} item
     * @return {?}
     */
    function (event, item) {
        /*const rollup = this.getRollup(this.tocData, event.data.sbUniqueIdentifier);
        if (rollup.length) {
          rollup.pop();
        }*/
        this.tocCardClick.emit(tslib_1.__assign({}, event, { item: item }));
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TocCurriculumComponent.prototype.onTocCardClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var rollup = this.getRollup(this.tocData, event.data.sbUniqueIdentifier);
        if (rollup.length) {
            rollup.pop();
        }
        this.tocCardClick.emit(tslib_1.__assign({}, event, { rollup: rollup }));
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TocCurriculumComponent.prototype.collapsedChangeHandler = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
    };
    /**
     * @private
     * @param {?} contents
     * @return {?}
     */
    TocCurriculumComponent.prototype.firstNonCollectionContent = /**
     * @private
     * @param {?} contents
     * @return {?}
     */
    function (contents) {
        return contents.find((/**
         * @param {?} content
         * @return {?}
         */
        function (content) { return content.mimeType !== 'application/vnd.ekstep.content-collection'; }));
    };
    /**
     * @private
     * @param {?} contents
     * @return {?}
     */
    TocCurriculumComponent.prototype.flattenDeep = /**
     * @private
     * @param {?} contents
     * @return {?}
     */
    function (contents) {
        var _this = this;
        if (contents) {
            return contents.reduce((/**
             * @param {?} acc
             * @param {?} val
             * @return {?}
             */
            function (acc, val) {
                if (val.children) {
                    acc.push(val);
                    return acc.concat(_this.flattenDeep(val.children));
                }
                else {
                    return acc.concat(val);
                }
            }), []);
        }
    };
    /**
     * @param {?} index
     * @param {?} item
     * @return {?}
     */
    TocCurriculumComponent.prototype.isExpanded = /**
     * @param {?} index
     * @param {?} item
     * @return {?}
     */
    function (index, item) {
        return Boolean(index === 0 || item && item.mimeType !== MimeTypeMasterData.COLLECTION);
    };
    /**
     * @param {?} item
     * @param {?} index
     * @return {?}
     */
    TocCurriculumComponent.prototype.isShowBody = /**
     * @param {?} item
     * @param {?} index
     * @return {?}
     */
    function (item, index) {
        var _this = this;
        if (item) {
            /** @type {?} */
            var isShowAllMimeType = (/**
             * @return {?}
             */
            function () { return _this.activeMimeTypeFilter.indexOf('all') > -1; });
            /** @type {?} */
            var isCollection = (/**
             * @return {?}
             */
            function () { return item.mimeType === MimeTypeMasterData.COLLECTION; });
            /** @type {?} */
            var isMathchesMimeType = (/**
             * @return {?}
             */
            function () { return _this.activeMimeTypeFilter.indexOf(item.mimeType) > -1; });
            /** @type {?} */
            var isLastContent = (/**
             * @return {?}
             */
            function () { return _this.tocData.children.length === index + 1; });
            if (this.tocData && isShowAllMimeType() || (isCollection() || isMathchesMimeType())) {
                this.isMimeTypeFilterChanged = true;
                return true;
            }
            else if (this.activeMimeTypeFilter.indexOf(item.mimeType) < 0 && !isLastContent()) {
                return false;
            }
            else if (this.activeMimeTypeFilter.indexOf(item.mimeType) < 0 && isLastContent()) {
                if (this.isMimeTypeFilterChanged) {
                    return false;
                }
                this.isMimeTypeFilterChanged = false;
                return true;
            }
        }
        return false;
    };
    /**
     * @param {?} tree
     * @param {?} identifier
     * @return {?}
     */
    TocCurriculumComponent.prototype.getRollup = /**
     * @param {?} tree
     * @param {?} identifier
     * @return {?}
     */
    function (tree, identifier) {
        /** @type {?} */
        var rollup = [tree.identifier];
        if (tree.sbUniqueIdentifier === identifier) {
            return rollup;
        }
        if (!tree.children || !tree.children.length) {
            return [];
        }
        /** @type {?} */
        var notDone = true;
        /** @type {?} */
        var childRollup;
        /** @type {?} */
        var index = 0;
        while (notDone && tree.children[index]) {
            childRollup = this.getRollup(tree.children[index], identifier);
            if (childRollup && childRollup.length) {
                notDone = false;
            }
            index++;
        }
        if (childRollup && childRollup.length) {
            rollup.push.apply(rollup, tslib_1.__spread(childRollup));
            return rollup;
        }
        else {
            return [];
        }
    };
    /**
     * @param {?} tocData
     * @param {?} item
     * @return {?}
     */
    TocCurriculumComponent.prototype.createUniqueId = /**
     * @param {?} tocData
     * @param {?} item
     * @return {?}
     */
    function (tocData, item) {
        item['sbUniqueIdentifier'] = tocData.identifier + '_' + item.identifier;
    };
    /**
     * @param {?=} item
     * @return {?}
     */
    TocCurriculumComponent.prototype.fetchProgress = /**
     * @param {?=} item
     * @return {?}
     */
    function (item) {
        /** @type {?} */
        var bgColor = this.progressColor;
        /** @type {?} */
        var widthStyle = item.progressPercentage != null ? item.progressPercentage + "%" : "2%";
        return {
            width: widthStyle,
            'background-color': bgColor
        };
    };
    /**
     * @param {?=} item
     * @return {?}
     */
    TocCurriculumComponent.prototype.showCompleted = /**
     * @param {?=} item
     * @return {?}
     */
    function (item) {
        if (item != null && item.progressPercentage != null && item.progressPercentage === 100) {
            return true;
        }
        return false;
    };
    /**
     * @param {?} item
     * @return {?}
     */
    TocCurriculumComponent.prototype.fetchProgressShadow = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        /** @type {?} */
        var bgColor = this.shadowColor;
        /** @type {?} */
        var displayStatus = item.progressPercentage != null && item.progressPercentage != "0" ? 'block' : 'none';
        return {
            'background': bgColor,
            display: displayStatus
        };
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TocCurriculumComponent.prototype.onPlayButtonClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var rollup = this.getRollup(this.tocData, event.data.sbUniqueIdentifier);
        if (rollup.length) {
            rollup.pop();
        }
        this.tocCardClick.emit(tslib_1.__assign({}, event, { rollup: rollup }));
    };
    TocCurriculumComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-toc-curriculum',
                    template: "<div *ngIf=\"isAccordion\">\r\n    <div class=\"\" *ngIf=\"tocData?.mimeType === MimeTypeMasterData?.COLLECTION\">\r\n      <sb-toc-item *ngIf=\"type\" [tocData]=\"tocData\" [type]=\"type\" [recogniseCollection]=\"true\"\r\n       (tocCardClick)=\"onTocCardClick($event)\" [contentStatus]=\"contentStatus\" (playButtonClick)=\"onPlayButtonClick($event)\">\r\n      </sb-toc-item>\r\n      <sb-toc-item *ngIf=\"!type\" [tocData]=\"tocData\" [recogniseCollection]=\"true\"></sb-toc-item>\r\n    </div>\r\n    <div class=\"\" *ngIf=\"tocData?.mimeType !== MimeTypeMasterData?.COLLECTION\">\r\n        <sb-toc-card *ngIf=\"type\" [content]=\"tocData\" [type]=\"type\" (tocCardClick)=\"onTocCardClick($event)\"[platform]=\"platform\"\r\n         [activeContent]=\"activeContent\" [contentStatus]=\"contentStatus\" [playBtnConfig]=\"playBtnConfig\" (playButtonClick)=\"onPlayButtonClick($event)\"\r\n         [trackableDefaultImage]=\"trackableDefaultImage\" [scoreLabel]=\"scoreLabel\" [maxAttempts]=\"maxAttempts\" [disabled]=\"disabled\">\r\n          </sb-toc-card>\r\n          <sb-toc-card *ngIf=\"!type\" [content]=\"tocData\"  (tocCardClick)=\"onTocCardClick($event)\" [platform]=\"platform\"\r\n            [activeContent]=\"activeContent\" [playBtnConfig]=\"playBtnConfig\" (playButtonClick)=\"onPlayButtonClick($event)\"\r\n            [trackableDefaultImage]=\"trackableDefaultImage\" [scoreLabel]=\"scoreLabel\" [maxAttempts]=\"maxAttempts\" [disabled]=\"disabled\">\r\n          </sb-toc-card>\r\n    </div>\r\n</div>\r\n<div  *ngIf=\"isChapterListing\">\r\n  <div class=\"sb-curiculum-responsive-grid\">\r\n    <div class=\"sb-curiculum__border\" *ngFor=\"let item of tocData?.children; let index = index;\" (click)=\"onTocChapterClick($event,item)\">\r\n      {{ createUniqueId(tocData, item) }}\r\n      <div class=\"sb-curiculum\" style=\"border:none;\">\r\n      <div class=\"sb--completed\" *ngIf=\"showCompleted(item)\">\r\n          <img alt=\"\" src=\"assets/common-consumption/images/completed.svg\" />\r\n      </div>\r\n      <div class=\"sb-curiculum__chapter\">\r\n        <div class=\"title\">{{item?.name}}</div>\r\n      </div>\r\n      </div>\r\n      <div class=\"sb--card__progressbar\" [ngStyle]=\"fetchProgressShadow(item)\">\r\n        <span [ngStyle]=\"fetchProgress(item)\"></span>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<div class=\"\" *ngIf=\"!isAccordion && !isChapterListing\">\r\n  \r\n  <div class=\"\">\r\n      <div class=\"sb-curiculum\" *ngFor=\"let item of tocData?.children; let index = index;\">\r\n      {{ createUniqueId(tocData, item) }}\r\n      <div class=\"sb-curiculum__chapter\" *ngIf=\"item?.mimeType === MimeTypeMasterData?.COLLECTION && item?.contentType !== 'TextBook' && item?.contentType !== 'Course'\" \r\n         (click)=\"onTocChapterClick($event,item)\">\r\n        <div class=\"title\">{{item?.name}}</div>\r\n        <div *ngIf=\"item && (item | hasMimeType: activeMimeTypeFilter);else noContent\" #chapterContainer>\r\n          <ng-container *ngIf=\"item?.mimeType !== MimeTypeMasterData?.COLLECTION && !item?.children?.length\" >\r\n            <div class=\"child-content-padding\">\r\n              <sb-toc-card *ngIf=\"type\" [content]=\"item\" [type]=\"type\" (tocCardClick)=\"onTocCardClick($event)\" [platform]=\"platform\"\r\n               [activeContent]=\"activeContent\" [contentStatus]=\"contentStatus\" [playBtnConfig]=\"playBtnConfig\" (playButtonClick)=\"onPlayButtonClick($event)\"\r\n               [trackableDefaultImage]=\"trackableDefaultImage\" [scoreLabel]=\"scoreLabel\" [maxAttempts]=\"maxAttempts\" [disabled]=\"disabled\">\r\n              </sb-toc-card>\r\n              <sb-toc-card *ngIf=\"!type\" [content]=\"item\" (tocCardClick)=\"onTocCardClick($event)\" [platform]=\"platform\"\r\n                [activeContent]=\"activeContent\" [playBtnConfig]=\"playBtnConfig\" (playButtonClick)=\"onPlayButtonClick($event)\"\r\n                [trackableDefaultImage]=\"trackableDefaultImage\" [scoreLabel]=\"scoreLabel\" [maxAttempts]=\"maxAttempts\" [disabled]=\"disabled\">\r\n              </sb-toc-card>\r\n            </div>\r\n          </ng-container>\r\n          <ng-container *ngIf=\"item?.children?.length\">\r\n            <ng-container *ngFor=\"let child of item?.children; let j = index;\">\r\n              {{ createUniqueId(item, child) }}\r\n              <div class=\"sbchapter\" *ngIf=\"filterChildren(child) && child.item?.contentType === 'TextBook'\">\r\n                <sb-library-card></sb-library-card>\r\n              </div>\r\n              <div class=\"sbchapter\" *ngIf=\"filterChildren(child) && child.item?.contentType === 'Course'\">\r\n                <sb-course-card></sb-course-card>\r\n              </div>\r\n              <div class=\"sbchapter\" *ngIf=\"filterChildren(child)\">\r\n                <sb-toc-child-item *ngIf=\"type\" [childData]=\"child\" [type]=\"type\" [activeMimeTypeFilter]=\"activeMimeTypeFilter\"\r\n                  [activeContent]=\"activeContent\" [contentStatus]=\"contentStatus\" (tocCardClick)=\"onTocCardClick($event)\" \r\n                  [playBtnConfig]=\"playBtnConfig\" (playButtonClick)=\"onPlayButtonClick($event)\" [platform]=\"platform\"\r\n                  [trackableDefaultImage]=\"trackableDefaultImage\">\r\n                </sb-toc-child-item>\r\n                <sb-toc-child-item *ngIf=\"!type\" [childData]=\"child\" [activeMimeTypeFilter]=\"activeMimeTypeFilter\"\r\n                  [activeContent]=\"activeContent\" (tocCardClick)=\"onTocCardClick($event)\" [playBtnConfig]=\"playBtnConfig\"\r\n                  (playButtonClick)=\"onPlayButtonClick($event)\" [platform]=\"platform\" [trackableDefaultImage]=\"trackableDefaultImage\">\r\n                </sb-toc-child-item>\r\n              </div>\r\n            </ng-container>\r\n          </ng-container>\r\n        </div>\r\n        <ng-template #noContent>\r\n          <div class=\"heading\">{{noContentMessage}}</div>\r\n        </ng-template>\r\n      </div>\r\n      <div class=\"sb-curiculum__chapter\" *ngIf=\"item?.mimeType !== MimeTypeMasterData?.COLLECTION\">\r\n          <sb-toc-card *ngIf=\"type\" [content]=\"item\" [type]=\"type\" (tocCardClick)=\"onTocCardClick($event)\" [platform]=\"platform\"\r\n          [activeContent]=\"activeContent\" [contentStatus]=\"contentStatus\" [playBtnConfig]=\"playBtnConfig\" (playButtonClick)=\"onPlayButtonClick($event)\"\r\n          [trackableDefaultImage]=\"trackableDefaultImage\" [scoreLabel]=\"scoreLabel\" [maxAttempts]=\"maxAttempts\" [disabled]=\"disabled\">\r\n            </sb-toc-card>\r\n      </div>\r\n      <div class=\"sb-curiculum__chapter\" *ngIf=\"item?.mimeType === MimeTypeMasterData?.COLLECTION && item?.contentType === 'TextBook'\">\r\n          <sb-library-card [content]=\"item\"></sb-library-card>\r\n      </div>\r\n      <div class=\"sb-curiculum__chapter\" *ngIf=\"item?.mimeType === MimeTypeMasterData?.COLLECTION && item?.contentType === 'Course'\">\r\n          <sb-course-card [course]=\"item\"></sb-course-card>\r\n        </div>\r\n    </div>\r\n    </div>\r\n</div>",
                    styles: [".sb--card__progressbar{margin:1rem!important;width:unset!important}.sb-curiculum-responsive-grid{padding:.5rem .5rem 1rem;display:block}.sb-curiculum-responsive-grid .sb-curiculum__border{background:var(--white);box-shadow:0 2px 7px rgba(var(--rc-rgba-black),.16);margin:.5rem 0;padding:.125rem 0}.sb-curiculum-responsive-grid .sb-curiculum__border.active,.sb-curiculum-responsive-grid .sb-curiculum__border:active,.sb-curiculum-responsive-grid .sb-curiculum__border:hover{box-shadow:0 2px 7px rgba(var(--rc-009D53),.65)}.sb-curiculum-responsive-grid .sb-curiculum__border .sb-curiculum.active,.sb-curiculum-responsive-grid .sb-curiculum__border .sb-curiculum:active,.sb-curiculum-responsive-grid .sb-curiculum__border .sb-curiculum:hover{box-shadow:none}.sb-curiculum-responsive-grid .sb-curiculum{position:relative;border-radius:.5rem;padding:.5rem 2.125rem 0 1rem;border:1px solid var(--primary-200);background:var(--white);font-weight:700;display:flex;min-height:48px;flex-direction:column}.sb-curiculum-responsive-grid .sb-curiculum.active,.sb-curiculum-responsive-grid .sb-curiculum:active,.sb-curiculum-responsive-grid .sb-curiculum:hover{box-shadow:0 2px 7px rgba(var(--rc-009D53),.65)}.sb-curiculum-responsive-grid .sb-curiculum__chapter{display:flex;align-items:flex-start;flex:1}.sb-curiculum-responsive-grid .sb-curiculum__chapter .title{font-size:.875rem;color:#3a3a3a;line-height:1.25rem;display:-webkit-box;-webkit-line-clamp:3;overflow:hidden;text-overflow:ellipsis;word-break:break-all}.sb-curiculum-responsive-grid .sb--completed{position:absolute;top:.875rem;right:.5rem}", ".loading-animation{-webkit-animation-duration:1s;animation-duration:1s;-webkit-animation-fill-mode:forwards;animation-fill-mode:forwards;-webkit-animation-iteration-count:infinite;animation-iteration-count:infinite;-webkit-animation-name:placeload;animation-name:placeload;-webkit-animation-timing-function:linear;animation-timing-function:linear;background:var(--lazyload-bg);background:var(--grey-nuance-lighter);background:linear-gradient(to right,var(--grey-nuance-lighter) 8%,var(--grey-nuance-darker) 18%,var(--grey-nuance-lighter) 33%);background-size:75rem 6.5rem;position:relative}@-webkit-keyframes placeload{0%{background-position:-29.25rem 0}100%{background-position:29.25rem 0}}@keyframes placeload{0%{background-position:-29.25rem 0}100%{background-position:29.25rem 0}}.sb--card--highlighter{height:.75rem;width:100%;margin-top:.3125rem}:host .layout-library{display:flex;flex-direction:column}:host .header{display:flex;flex-direction:row;padding:0;align-items:center}[dir=ltr] :host .header,[dir=rtl] :host .header{margin:1.5rem 0 1rem}:host .header .title{font-family:inherit;text-overflow:ellipsis;overflow:hidden;flex:1;height:1.5rem;color:var(--primary-color);font-size:1.125rem;font-weight:700;text-transform:capitalize;line-height:1.625rem;display:flex;align-items:center}:host .header .title i{vertical-align:middle;height:1.0625rem;width:.875rem;top:0}[dir=ltr] :host .header .title i{margin-right:.5rem}[dir=rtl] :host .header .title i{margin-left:.5rem}:host .header .title .badge{display:inline-block;background:var(--rc-d4d3d3);color:var(--gray-800);text-align:center;padding:0 .75rem;border-radius:.25rem;height:1.5rem;line-height:1.5rem;font-size:1rem;font-weight:400}[dir=ltr] :host .header .title .badge{margin-left:.5rem}[dir=rtl] :host .header .title .badge{margin-right:.5rem}:host .cards-containers{display:-ms-grid;display:grid;grid-template-columns:repeat(auto-fill,minmax(292px,1fr));grid-gap:1rem;grid-row-gap:1.5rem}:host .cards-containers--show-1{grid-template-columns:repeat(auto-fill,minmax(calc(100% / 1.5),1fr))}:host .cards-containers--show-2{grid-template-columns:repeat(auto-fill,minmax(calc(100% / 2.5),1fr))}:host .cards-containers--show-3{grid-template-columns:repeat(auto-fill,minmax(calc(100% / 3.5),1fr))}:host .cards-containers--show-4{grid-template-columns:repeat(auto-fill,minmax(calc(100% / 4.5),1fr))}:host .cards-containers--show-5{grid-template-columns:repeat(auto-fill,minmax(calc(100% / 5.5),1fr))}:host .cards-container{display:flex;flex-direction:row;flex-wrap:wrap;margin:-1.5rem -.5rem 0}:host .cards-container .item{box-sizing:border-box;margin-top:1.5rem}[dir=ltr] :host .cards-container .item,[dir=rtl] :host .cards-container .item{padding:0 .5rem}@media (max-width:767px){:host .cards-container .item{margin-top:1rem}}:host .cards-container .item.show-1{max-width:100%}:host .cards-container .item.show-2{flex:0 0 50%;max-width:50%}:host .cards-container .item.show-3{flex:0 0 33.33%;max-width:33.33%}:host .cards-container .item.show-4{flex:0 0 25%;max-width:25%}:host .cards-container .item.show-5{flex:0 0 20%;max-width:20%}@media (max-width:991px){:host .cards-containers--show-3,:host .cards-containers--show-4,:host .cards-containers--show-5{grid-template-columns:repeat(auto-fill,minmax(calc(100% / 2.5),1fr))}:host .cards-container .item.show-2,:host .cards-container .item.show-3,:host .cards-container .item.show-4,:host .cards-container .item.show-5{flex:0 0 50%;max-width:50%}}@media (max-width:575px){:host .cards-containers--show-2,:host .cards-containers--show-3,:host .cards-containers--show-4,:host .cards-containers--show-5{grid-template-columns:repeat(auto-fill,minmax(calc(100% / 1.5),1fr))}:host .cards-container .item.show-2,:host .cards-container .item.show-3,:host .cards-container .item.show-4,:host .cards-container .item.show-5{flex:0 0 100%;max-width:100%}}@media (min-width:992px) and (max-width:1279px){:host .cards-containers--show-4,:host .cards-containers--show-5{grid-template-columns:repeat(auto-fill,minmax(calc(100% / 3.5),1fr))}:host .cards-container .item.show-4,:host .cards-container .item.show-5{flex:0 0 33.33%;max-width:33.33%}}", ".loading-animation{-webkit-animation-duration:1s;animation-duration:1s;-webkit-animation-fill-mode:forwards;animation-fill-mode:forwards;-webkit-animation-iteration-count:infinite;animation-iteration-count:infinite;-webkit-animation-name:placeload;animation-name:placeload;-webkit-animation-timing-function:linear;animation-timing-function:linear;background:var(--lazyload-bg);background:var(--grey-nuance-lighter);background:linear-gradient(to right,var(--grey-nuance-lighter) 8%,var(--grey-nuance-darker) 18%,var(--grey-nuance-lighter) 33%);background-size:75rem 6.5rem;position:relative}@-webkit-keyframes placeload{0%{background-position:-29.25rem 0}100%{background-position:29.25rem 0}}@keyframes placeload{0%{background-position:-29.25rem 0}100%{background-position:29.25rem 0}}.sb--card--highlighter{height:.75rem;width:100%;margin-top:.3125rem}:host .sb--card{width:100%;background-color:var(--white);display:flex;flex-direction:column;cursor:pointer;position:relative;margin-top:0;border:0;contain:content;border-radius:.125rem}:host .sb--card:hover{transform:translate(0,0);box-shadow:0 .125rem .4375rem 0 rgba(var(--rc-rgba-black),.16)}@media (max-width:767px){:host .sb--card{box-shadow:0 .125rem .4375rem rgba(var(--rc-rgba-black),.16)}}:host .sb--card__main-area{display:flex;flex-direction:row;align-items:center;height:5.5rem}:host .sb--card__img{height:4.5rem;width:4.5rem;overflow:hidden;flex:0 0 4.5rem;margin:.5rem}:host .sb--card__img .img-container{display:flex;align-items:center;background-color:var(--gray-100);height:4.5rem;width:4.5rem;justify-content:center}:host .sb--card__img .img-container img{max-width:100%}:host .sb--card__info{flex:1;overflow:hidden}[dir=ltr] :host .sb--card__info{padding-right:.5rem}[dir=rtl] :host .sb--card__info{padding-left:.5rem}:host .sb--card__title{color:var(--primary-color);font-size:.875rem;margin:0;line-height:1.25rem}:host .sb--card__meta1{color:var(--gray-800);font-size:.75rem;display:flex;flex-direction:row;align-items:center;flex:1;line-height:1.25rem;margin-top:.25rem}:host .sb--card .data_1{max-width:50%}:host .sb--card .data_2{flex:1}:host .sb--card .dot-divider{width:.25rem;height:.25rem;min-width:.25rem;max-width:.25rem;min-height:.25rem;max-height:.25rem;border-radius:100%;background-color:var(--gray-200);display:inline-block;margin:0 .5rem}:host .sb--card__moreinfo{border-radius:0 0 .125rem .125rem;background-color:var(--rc-FAFAFA);display:flex;flex-direction:row;align-items:stretch;justify-content:space-between;min-height:3.5rem;border-top:0 solid #cacbce}:host .sb--card__meta2{color:var(--gray-800);font-size:.75rem;line-height:.875rem;display:flex;flex-direction:column;flex-wrap:wrap;justify-content:space-around;padding:.5rem}:host .sb--card__meta2_data .label{font-weight:400}:host .sb--card__meta2_data .value{font-weight:700}:host .sb--card__org{color:var(--gray-200)}:host .sb--card__tags{display:flex;flex-direction:row;align-items:center;justify-content:flex-end;flex:1}:host .sb--card__badge{display:flex}:host .sb--card__badge img{width:1.25rem;height:1.25rem}:host .sb--card__type{background:var(--gray-200);position:relative;height:1.25rem;color:var(--white);font-size:.625rem;display:flex;align-items:center}[dir=ltr] :host .sb--card__type{padding-left:.25rem;padding-right:.5rem;margin-left:1rem}[dir=rtl] :host .sb--card__type{padding-left:.5rem;padding-right:.25rem;margin-right:1rem}:host .sb--card__type:before{content:\"\";width:0;height:0;border-width:.625rem .3125rem;border-style:solid;border-color:var(--gray-200) var(--gray-200) var(--gray-200) transparent;position:absolute;left:-10px}html[dir=rtl] :host .sb--card__type:before{right:-.625rem;left:inherit;border-color:var(--gray-200) transparent var(--gray-200) var(--gray-200)}:host .sb--card--recently-viewed .sb--card__img{height:5.5rem;width:5.5rem;flex:0 0 5.5rem;margin:0}:host .sb--card--recently-viewed .sb--card__img .img-container{display:flex;align-items:center;background-color:var(--gray-100);height:5.5rem;width:5.5rem}:host .sb--card--recently-viewed .sb--card__info{padding:.5rem;flex:1}:host .sb--card.selected{border:1px solid #024f9d;border-radius:4px;background-color:rgba(2,79,157,.2)}:host .sb--card.offline .sb--card__img .img-container{display:flex;align-items:center;background-color:var(--tertiary-color);opacity:.5}:host .sb--card.offline .sb--card__img .img-container img{margin:auto;font-size:1.5rem;color:var(--white)}:host .sb--card.offline .sb--card__info,:host .sb--card.offline .sb--card__moreinfo{opacity:.5}:host .sb--card__progressbar{height:.25rem;margin:0;border-radius:0;box-shadow:inset 0 -.0625rem .0625rem rgba(var(--rc-rgba-white),.3);width:100%;background:var(--gray-100)}:host .sb--card__progressbar>span{display:block;height:100%;background-color:var(--secondary-200);background-image:-webkit-gradient(center bottom,var(--secondary-400),var(--secondary-200));box-shadow:inset 0 .125rem .5625rem var(--gray-100),inset 0 -.125rem .375rem rgba(var(--rc-rgba-black),.4);position:relative;overflow:hidden}:host .animate>span:after{content:\"\";position:absolute;top:0;left:0;bottom:0;right:0;background-image:-webkit-gradient(linear,0 0,100% 100%,color-stop(.25,rgba(var(--rc-rgba-white),.2)),color-stop(.25,transparent),color-stop(.5,transparent),color-stop(.5,rgba(var(--rc-rgba-white),.2)),color-stop(.75,rgba(var(--rc-rgba-white),.2)),color-stop(.75,transparent),to(transparent));z-index:1;background-size:3.125rem 3.125rem;overflow:hidden;-webkit-animation:2s linear infinite move;border-radius:1.25rem .5rem .5rem 1.25rem}@-webkit-keyframes move{0%{background-position:0 0}100%{background-position:3.125rem 3.125rem}}:host .completed>span{background-color:var(--secondary-color);background-image:-webkit-linear-gradient(var(--secondary-100),var(--secondary-400))}:host .inprogress>span{background-color:var(--tertiary-color);background-image:-webkit-linear-gradient(var(--tertiary-100),var(--tertiary-400))}:host .notcompleted>span{background-color:var(--red-100);background-image:-webkit-linear-gradient(var(--red-100),var(--red-400))}:host .nostripes>span,:host .nostripes>span:after{-webkit-animation:none;-moz-animation:none;background-image:none}:host .border-top-r-left-right{border-top-left-radius:0;border-top-right-radius:0}::ng-deep .card-hover{display:none;transition:.2s ease-in}::ng-deep .sb--card:hover .card-hover{display:block;z-index:1001}::ng-deep .sb--card:hover .card-hover .child-content{color:var(--white);position:absolute;left:0;right:0;top:0;bottom:0;opacity:.95;border-radius:.125rem;background-color:var(--primary-color);z-index:1;display:flex;align-items:center;transition:.2s ease-in;flex-direction:column;justify-content:space-evenly}::ng-deep .sb--card:hover .card-hover .action-container{display:flex;width:100%;justify-content:center;align-self:center;justify-self:center;flex:1;align-items:center}::ng-deep .sb--card:hover .card-hover .action-item{padding:0 1rem;flex:1;text-align:center}::ng-deep .sb--card:hover .card-hover .action-item .sb-btn.sb-btn-primary[disabled]{background-color:transparent;opacity:.5;border:0;color:var(--white)}::ng-deep .sb--card:hover .card-hover .action-item:not(:last-child){border-right:.0625rem solid rgba(var(--rc-rgba-white),.5)}::ng-deep .sb--card:hover .card-hover .hover-meta-info{font-size:.625rem;opacity:.8;color:var(--white);line-height:.75rem;flex:0 0 25%}.menu::after{content:'\\2807';font-size:1.5rem;color:var(--gray-200);text-align:center;align-items:center;display:flex;justify-content:center;border-radius:50%;padding-left:.375rem;cursor:pointer;width:2rem;height:2rem}.menu:hover::after{background:var(--gray-0)}"]
                }] }
    ];
    /** @nocollapse */
    TocCurriculumComponent.ctorParameters = function () { return []; };
    TocCurriculumComponent.propDecorators = {
        tocData: [{ type: Input }],
        activeMimeTypeFilter: [{ type: Input }],
        noContentMessage: [{ type: Input }],
        isAccordion: [{ type: Input }],
        isChapterListing: [{ type: Input }],
        type: [{ type: Input }],
        contentStatus: [{ type: Input }],
        shadowColor: [{ type: Input }],
        progressColor: [{ type: Input }],
        refresh: [{ type: Input }],
        layoutConfig: [{ type: Input }],
        playBtnConfig: [{ type: Input }],
        platform: [{ type: Input }],
        trackableDefaultImage: [{ type: Input }],
        maxAttempts: [{ type: Input }],
        disabled: [{ type: Input }],
        divs: [{ type: ViewChild, args: ['chapter', { static: false },] }],
        chapterContainer: [{ type: ViewChildren, args: ['chapterContainer',] }],
        activeContent: [{ type: Input }],
        tocCardClick: [{ type: Output }],
        noContent: [{ type: Output }],
        scoreLabel: [{ type: Input }]
    };
    return TocCurriculumComponent;
}());
export { TocCurriculumComponent };
if (false) {
    /** @type {?} */
    TocCurriculumComponent.prototype.tocData;
    /** @type {?} */
    TocCurriculumComponent.prototype.activeMimeTypeFilter;
    /** @type {?} */
    TocCurriculumComponent.prototype.noContentMessage;
    /** @type {?} */
    TocCurriculumComponent.prototype.isAccordion;
    /** @type {?} */
    TocCurriculumComponent.prototype.isChapterListing;
    /** @type {?} */
    TocCurriculumComponent.prototype.type;
    /** @type {?} */
    TocCurriculumComponent.prototype.contentStatus;
    /** @type {?} */
    TocCurriculumComponent.prototype.shadowColor;
    /** @type {?} */
    TocCurriculumComponent.prototype.progressColor;
    /** @type {?} */
    TocCurriculumComponent.prototype.refresh;
    /** @type {?} */
    TocCurriculumComponent.prototype.layoutConfig;
    /** @type {?} */
    TocCurriculumComponent.prototype.playBtnConfig;
    /** @type {?} */
    TocCurriculumComponent.prototype.platform;
    /** @type {?} */
    TocCurriculumComponent.prototype.trackableDefaultImage;
    /** @type {?} */
    TocCurriculumComponent.prototype.maxAttempts;
    /** @type {?} */
    TocCurriculumComponent.prototype.disabled;
    /** @type {?} */
    TocCurriculumComponent.prototype.divs;
    /** @type {?} */
    TocCurriculumComponent.prototype.chapterContainer;
    /** @type {?} */
    TocCurriculumComponent.prototype.activeContent;
    /** @type {?} */
    TocCurriculumComponent.prototype.tocCardClick;
    /** @type {?} */
    TocCurriculumComponent.prototype.noContent;
    /** @type {?} */
    TocCurriculumComponent.prototype.scoreLabel;
    /** @type {?} */
    TocCurriculumComponent.prototype.isMimeTypeFilterChanged;
    /**
     * @type {?}
     * @private
     */
    TocCurriculumComponent.prototype.isSameMimeTypeInChildren;
    /**
     * @type {?}
     * @private
     */
    TocCurriculumComponent.prototype.isSameMimeType;
    /**
     * @type {?}
     * @private
     */
    TocCurriculumComponent.prototype.isChildrenPresent;
}
