// @project-sunbird/common-consumption-v8/esm5/lib/card/card-hover/card-hover.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
var CardHoverComponent = /** @class */ (function () {
    function CardHoverComponent() {
        this.hoverActionClick = new EventEmitter();
        this.text = 'click';
    }
    /**
     * @return {?}
     */
    CardHoverComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    CardHoverComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} item
     * @param {?} content
     * @param {?} event
     * @return {?}
     */
    CardHoverComponent.prototype.actionClick = /**
     * @param {?} item
     * @param {?} content
     * @param {?} event
     * @return {?}
     */
    function (item, content, event) {
        this.hoverActionClick.emit({ event: event, hover: item, content: content });
        event.stopPropagation();
    };
    CardHoverComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-card-hover',
                    template: "\r\n<!-- <div class=\"card-hover\"> -->\r\n    <div class=\"action-container\">\r\n        <ng-container *ngFor=\"let item of hoverData?.actions\">\r\n            <div class=\"action-item\" *ngIf=\"item.type === 'save'\">\r\n                <button class=\"sb-btn-grow\" (click)=\"actionClick(item, content, $event);\" [disabled]=\"item.disabled\">\r\n                    <i class=\"icon-svg icon-svg--md\">\r\n                        <svg class=\"icon\">\r\n                            <use xlink:href=\"assets/common-consumption/images/sprite.svg#pendrive\"></use>\r\n                        </svg>\r\n                    </i>\r\n                  <span class=\"sb__ellipsis\">{{item.label}}</span>\r\n                </button>\r\n            </div>\r\n            <div class=\"action-item\" *ngIf=\"item.type === 'download'\">\r\n                <button class=\"sb-btn-grow\" (click)=\"actionClick(item, content, $event);\" [disabled]=\"item.disabled\">\r\n                    <i class=\"icon-svg icon-svg--md\">\r\n                        <svg class=\"icon\">\r\n                            <use xlink:href=\"assets/common-consumption/images/sprite.svg#outline-cloud_download\"></use>\r\n                        </svg>\r\n                    </i>\r\n                    <span class=\"sb__ellipsis\">{{item.label}}</span>\r\n                  </button>\r\n            </div>\r\n            <div class=\"action-item\" *ngIf=\"item.type === 'open'\">\r\n                <button class=\"sb-btn-grow\" (click)=\"actionClick(item, content, $event);\" [disabled]=\"item.disabled\">\r\n                    <i class=\"icon-svg icon-svg--md\">\r\n                        <svg class=\"icon\">\r\n                            <use xlink:href=\"assets/common-consumption/images/sprite.svg#book-open\"></use>\r\n                        </svg>\r\n                    </i>\r\n                    <span class=\"sb__ellipsis\">{{item.label}}</span>\r\n                  </button>\r\n            </div>\r\n            <div class=\"action-item\" *ngIf=\"item.type === 'view'\">\r\n                <button class=\"sb-btn-grow\" (click)=\"actionClick(item, content, $event);\" [disabled]=\"item.disabled\">\r\n                    <i class=\"icon-svg icon-svg--sm\">\r\n                        <svg class=\"icon\">\r\n                            <use xlink:href=\"assets/common-consumption/images/sprite.svg#eye\"></use>\r\n                        </svg>\r\n                    </i>\r\n                    <span class=\"sb__ellipsis\">{{item.label}}</span>\r\n                  </button>\r\n            </div>\r\n            <div class=\"action-item\" *ngIf=\"item.type === 'addToGroup'\">\r\n                <button class=\"sb-btn-grow\" (click)=\"actionClick(item, content, $event);\" [disabled]=\"item.disabled\">\r\n                    <i class=\"icon-svg icon-svg--sm\">\r\n                        <svg class=\"icon\">\r\n                            <use xlink:href=\"assets/common-consumption/images/sprite.svg#add-icon\"></use>\r\n                        </svg>\r\n                    </i>\r\n                    <span class=\"sb__ellipsis\">{{item.label}}</span>\r\n                  </button>\r\n            </div>\r\n            <div  class=\"d-flex flex-ai-center flex-jc-center fnormal sb-color-white\" *ngIf=\"item.type === 'text'\">\r\n                {{item?.label}}\r\n                </div>\r\n        </ng-container>\r\n        \r\n        <!-- <div class=\"action-item\">\r\n            <button class=\"sb-btn sb-btn-link sb-btn-link- sb-btn-normal sb-btn-square sb-color-white w-auto sb-btn-grow\" (click)=\"test('button3'); $event.stopPropagation()\">\r\n                <i class=\"icon-svg icon-svg--sm\">\r\n                    <svg class=\"icon\">\r\n                        <use xlink:href=\"assets/common-consumption/images/sprite.svg#book\"></use>\r\n                    </svg>\r\n                </i>\r\n                <span>{{text}}</span>\r\n              </button>\r\n        </div> -->\r\n    </div>\r\n    <div class=\"hover-meta-info\" *ngIf=\"hoverData?.note\">\r\n        {{hoverData?.note}}\r\n    </div>\r\n<!-- </div> -->",
                    styles: [".sb-btn-grow{background:0 0;border:0 solid transparent;font-size:.75rem;color:var(--white);cursor:pointer;outline:0!important;transition:.2s ease-in-out}.sb-btn-grow:disabled{opacity:.4;cursor:default}.sb-btn-grow:hover:not([disabled]){background:0 0;transform:scale(1.15)}"]
                }] }
    ];
    /** @nocollapse */
    CardHoverComponent.ctorParameters = function () { return []; };
    CardHoverComponent.propDecorators = {
        hoverData: [{ type: Input }],
        content: [{ type: Input }],
        hoverActionClick: [{ type: Output }]
    };
    return CardHoverComponent;
}());
export { CardHoverComponent };
if (false) {
    /** @type {?} */
    CardHoverComponent.prototype.hoverData;
    /** @type {?} */
    CardHoverComponent.prototype.content;
    /** @type {?} */
    CardHoverComponent.prototype.hoverActionClick;
    /** @type {?} */
    CardHoverComponent.prototype.text;
}
