// @project-sunbird/common-consumption-v8/esm5/lib/layout/in-app-notification/in-app-notification.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { notificationViewConfig } from './in-app-notification-data';
var InAppNotificationComponent = /** @class */ (function () {
    function InAppNotificationComponent() {
        this.notificationList = [];
        this.inAppNotificationConfig = notificationViewConfig;
        this.notificationClick = new EventEmitter();
        this.deleteNotificationClick = new EventEmitter();
        this.clearAllClick = new EventEmitter();
        this.showMore = new EventEmitter();
        this.showLess = new EventEmitter();
    }
    /**
     * @return {?}
     */
    InAppNotificationComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.displayItemCount = this.inAppNotificationConfig.minNotificationViewCount;
        console.log(this.notificationList);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    InAppNotificationComponent.prototype.notificationClickHandler = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        console.log(this.notificationList);
        this.notificationClick.emit(event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    InAppNotificationComponent.prototype.deleteNotificationHandler = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.deleteNotificationClick.emit(event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    InAppNotificationComponent.prototype.clearAllNotifications = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var eventData = {
            event: event,
            data: this.notificationList
        };
        this.clearAllClick.emit(eventData);
    };
    /**
     * @return {?}
     */
    InAppNotificationComponent.prototype.onShowMore = /**
     * @return {?}
     */
    function () {
        this.showMore.emit(true);
        if (this.notificationList.length) {
            this.displayItemCount = this.notificationList.length;
        }
    };
    /**
     * @return {?}
     */
    InAppNotificationComponent.prototype.onShowLess = /**
     * @return {?}
     */
    function () {
        this.showLess.emit(true);
        this.displayItemCount = this.inAppNotificationConfig.minNotificationViewCount;
    };
    InAppNotificationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-in-app-notification',
                    template: "<div class=\"sb-notification-container\">\r\n  <div class=\"sb-notification-header\">\r\n    <div class=\"title\" *ngIf=\"inAppNotificationConfig?.title\">{{inAppNotificationConfig?.title}}</div>\r\n    <div class=\"sub-title\" *ngIf=\"inAppNotificationConfig?.subTitle\">{{inAppNotificationConfig?.subTitle}}</div>\r\n  </div>\r\n  <div class=\"clear-text\" *ngIf=\"inAppNotificationConfig.clearText\">\r\n    <a (click)=\"clearAllNotifications($event)\">{{inAppNotificationConfig.clearText}}</a>\r\n  </div>\r\n  <div class=\"sb-notification-list-items\">\r\n    <div *ngFor=\"let notification of notificationList | slice:0:displayItemCount; let i=index\" >\r\n      <sb-notification-card [notification]=\"notification\" (notificationClick)=\"notificationClickHandler($event)\"\r\n      (deleteNotificationClick)=\"deleteNotificationHandler($event)\"></sb-notification-card>\r\n    </div>\r\n  </div>\r\n  <ng-container *ngIf=\"inAppNotificationConfig?.minNotificationViewCount < this.notificationList?.length\">\r\n    <div class=\"text-center more-text mt-16\" *ngIf=\"displayItemCount < this.notificationList?.length\">\r\n      <a (click)=\"onShowMore()\">{{inAppNotificationConfig?.moreText}}</a>\r\n    </div>\r\n    <div class=\"text-center more-text mt-16\" *ngIf=\"displayItemCount === this.notificationList?.length\">\r\n      <a (click)=\"onShowLess()\">{{inAppNotificationConfig?.lessText}}</a>\r\n    </div>\r\n  </ng-container>\r\n</div>",
                    styles: [":host .sb-notification-container .sb-notification-header{color:var(--gray)}:host .sb-notification-container .sb-notification-header .title{font-size:1rem;font-weight:700;margin-bottom:.25rem}:host .sb-notification-container .sb-notification-header .sub-title{font-size:.75rem;margin-bottom:1rem}:host .sb-notification-container .sb-notification-list-items{max-height:calc(100vh - 275px);overflow-y:auto}:host .sb-notification-container .clear-text a{color:var(--danger-color);font-size:.875rem;position:absolute;right:1.5rem;top:1.5rem;text-decoration:none}html[dir=rtl] :host .sb-notification-container .clear-text a{right:auto;left:1.5rem}:host .sb-notification-container .more-text a{font-size:.75rem;font-weight:700;cursor:pointer;color:var(--primary-400)}"]
                }] }
    ];
    /** @nocollapse */
    InAppNotificationComponent.ctorParameters = function () { return []; };
    InAppNotificationComponent.propDecorators = {
        notificationList: [{ type: Input }],
        inAppNotificationConfig: [{ type: Input }],
        notificationClick: [{ type: Output }],
        deleteNotificationClick: [{ type: Output }],
        clearAllClick: [{ type: Output }],
        showMore: [{ type: Output }],
        showLess: [{ type: Output }]
    };
    return InAppNotificationComponent;
}());
export { InAppNotificationComponent };
if (false) {
    /** @type {?} */
    InAppNotificationComponent.prototype.notificationList;
    /** @type {?} */
    InAppNotificationComponent.prototype.inAppNotificationConfig;
    /** @type {?} */
    InAppNotificationComponent.prototype.notificationClick;
    /** @type {?} */
    InAppNotificationComponent.prototype.deleteNotificationClick;
    /** @type {?} */
    InAppNotificationComponent.prototype.clearAllClick;
    /** @type {?} */
    InAppNotificationComponent.prototype.showMore;
    /** @type {?} */
    InAppNotificationComponent.prototype.showLess;
    /** @type {?} */
    InAppNotificationComponent.prototype.displayItemCount;
}
