// @project-sunbird/common-consumption-v8/esm5/lib/layout/in-app-notification/in-app-notification-data.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var notificationList = [
    {
        "identifier": "c2d72c04-3010-45d7-bc13-29a3105dca26",
        "data": {
            "actionType": "contentURL",
            "contentURL": "https://staging.ntp.net.in/resources/play/content/do_212669767503060992174",
            "description": "Notfication description",
            "openPlayer": true,
            "title": "InApp Notification 4"
        },
        "id": "c2d72c04-3010-45d7-bc13-29a3105dca26",
        "category": "Notification",
        "priority": 1,
        "userId": "555caa55-8dc8-4d8f-a8c5-159ade057f90",
        "createdOn": 1605589382634,
        "status": "unread"
    },
    {
        "identifier": "03ae8b70-e6ea-4615-9545-bf9e5ee2f30b",
        "data": {
            "actionType": "contentURL",
            "contentURL": "https://staging.ntp.net.in/resources/play/content/do_212669767503060992174",
            "description": "Notfication description",
            "openPlayer": true,
            "title": "InApp Notification 7"
        },
        "id": "03ae8b70-e6ea-4615-9545-bf9e5ee2f30b",
        "category": "Notification",
        "priority": 1,
        "userId": "555caa55-8dc8-4d8f-a8c5-159ade057f90",
        "createdOn": 1605589397767,
        "status": "unread"
    }
];
/** @type {?} */
export var notificationViewConfig = {
    title: 'Notification',
    subTitle: 'New Notification (s)',
    clearText: 'Clear',
    moreText: 'See more',
    lessText: 'See less',
    minNotificationViewCount: 5
};
