// @project-sunbird/common-consumption-v8/esm5/lib/accordion/accordion-body/accordion-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Host } from '@angular/core';
import { trigger, state, style, transition, animate } from '@angular/animations';
import { AccordionItemComponent } from '../accordion-item/accordion-item.component';
/** @type {?} */
var accordionItemBodyTrigger = trigger('accordionItemBody', [
    state('collapsed', style({
        overflow: 'hidden',
        visibility: 'hidden',
        height: 0,
    })),
    state('expanded', style({
        overflow: 'hidden',
        visibility: 'visible',
    })),
    transition('collapsed => expanded', animate('100ms ease-in')),
    transition('expanded => collapsed', animate('100ms ease-out')),
]);
var AccordionBodyComponent = /** @class */ (function () {
    function AccordionBodyComponent(accordionItem) {
        this.accordionItem = accordionItem;
    }
    Object.defineProperty(AccordionBodyComponent.prototype, "state", {
        get: /**
         * @return {?}
         */
        function () {
            return this.accordionItem.collapsed ? 'collapsed' : 'expanded';
        },
        enumerable: true,
        configurable: true
    });
    AccordionBodyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-accordion-body',
                    template: "<div [@accordionItemBody]=\"{ value: state }\">\r\n  <div class=\"sbaccordion__panel-content\">\r\n    <ng-content></ng-content>\r\n  </div>\r\n</div>",
                    animations: [accordionItemBodyTrigger],
                    styles: [".sbaccordion__panel-content{border:.0625rem solid var(--primary-300);background:var(--gray-0);min-height:2rem;padding:1rem;overflow:hidden;overflow-y:auto;max-height:60vh;margin-top:-1px;border-bottom-left-radius:.125rem;border-bottom-right-radius:.125rem}.accordion-item-header-collapsed .sbaccordion__panel-header__title{font-weight:400}"]
                }] }
    ];
    /** @nocollapse */
    AccordionBodyComponent.ctorParameters = function () { return [
        { type: AccordionItemComponent, decorators: [{ type: Host }] }
    ]; };
    return AccordionBodyComponent;
}());
export { AccordionBodyComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AccordionBodyComponent.prototype.accordionItem;
}
