// @project-sunbird/common-consumption-v8/esm2015/lib/layout/member-list/member-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input } from '@angular/core';
export class MemberListComponent {
    constructor() {
        this.isMenu = false;
        this.layoutConfig = { size: 'medium', isBold: false, isSelectable: true, view: "vertical" };
        this.menuClick = new EventEmitter();
        this.cardClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} event
     * @param {?} data
     * @return {?}
     */
    onCardClick(event, data) {
        this.cardClick.emit({ event: event, data: { data } });
    }
    /**
     * @param {?} event
     * @param {?} data
     * @return {?}
     */
    onMenuClick(event, data) {
        this.menuClick.emit({ event: event, data: { data } });
    }
}
MemberListComponent.decorators = [
    { type: Component, args: [{
                selector: 'sb-member-list',
                template: "<div class=\"members-list\" *ngIf=\"layoutConfig.view == 'horizontal'\">\r\n  <span *ngFor=\"let member of memberList;let index = index\">\r\n      <sb-member-card [title]=\"member.title\" [initial]=\"member.initial\" [isAdmin]=\"member.isAdmin\" \r\n      [isMenu]=\"isMenu\" [progressDisplay]=\"member.progressDisplay\" [selected]=\"member.selected\" [cardImg]=\"cardImg\" [config]=\"layoutConfig\" [identifier]=\"member.id\" [indexOfMember]=\"index\" (cardClick)=\"onCardClick($event, member)\" (menuClick)=\"onMenuClick($event, member)\">\r\n      </sb-member-card>\r\n  </span>\r\n</div>\r\n<div class=\"members-inline\" *ngIf=\"layoutConfig.view == 'vertical'\">\r\n    <span *ngFor=\"let member of memberList;let index = index\">\r\n      <sb-member-card [title]=\"member.title\" [initial]=\"member.initial\" [isAdmin]=\"member.isAdmin\" \r\n        [isMenu]=\"isMenu\" [progressDisplay]=\"member.progressDisplay\" [cardImg]=\"cardImg\" [config]=\"layoutConfig\" [identifier]=\"member.id\" [indexOfMember]=\"index\" (cardClick)=\"onCardClick($event, member)\" (menuClick)=\"onMenuClick($event, member)\">\r\n      </sb-member-card>\r\n    </span>\r\n</div>\r\n",
                styles: [".members-list{width:100%;background-color:var(--white)}.members-inline{display:flex;flex-wrap:wrap}"]
            }] }
];
/** @nocollapse */
MemberListComponent.ctorParameters = () => [];
MemberListComponent.propDecorators = {
    memberList: [{ type: Input }],
    isMenu: [{ type: Input }],
    cardImg: [{ type: Input }],
    layoutConfig: [{ type: Input }],
    menuClick: [{ type: Output }],
    cardClick: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    MemberListComponent.prototype.memberList;
    /** @type {?} */
    MemberListComponent.prototype.isMenu;
    /** @type {?} */
    MemberListComponent.prototype.cardImg;
    /** @type {?} */
    MemberListComponent.prototype.layoutConfig;
    /** @type {?} */
    MemberListComponent.prototype.menuClick;
    /** @type {?} */
    MemberListComponent.prototype.cardClick;
}
