// @project-sunbird/common-consumption-v8/esm5/lib/content-details/player-actions/player-actions.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { ActionButtonType } from './model';
/** @type {?} */
var IMAGE_PATH = 'assets/common-consumption/images/sprite.svg';
/** @type {?} */
var DEFAULT_BUTTON = {
    name: ActionButtonType.FULL_SCREEN,
    label: 'Fullscreen',
    iconPath: IMAGE_PATH + "#" + ActionButtonType.FULL_SCREEN,
    disabled: false
};
var PlayerActionsComponent = /** @class */ (function () {
    function PlayerActionsComponent() {
        this.actionButtons = [];
        this.buttonClick = new EventEmitter();
        this.showFullscreen = false;
        this.fullscreenButton = DEFAULT_BUTTON;
    }
    Object.defineProperty(PlayerActionsComponent.prototype, "ActionButtonType", {
        get: /**
         * @return {?}
         */
        function () { return ActionButtonType; },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PlayerActionsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.actionButtons) {
            this.actionButtons.forEach((/**
             * @param {?} item
             * @param {?} index
             * @param {?} object
             * @return {?}
             */
            function (item, index, object) {
                item.iconPath = item.iconPath ? item.iconPath : IMAGE_PATH + "#" + item.name;
                if (item.name === ActionButtonType.FULL_SCREEN) {
                    _this.showFullscreen = true;
                    _this.fullscreenButton = item;
                    object.splice(index, 1);
                }
            }));
        }
    };
    /**
     * Emits event when user clicks on the button
     * @param event Contains default mouse click event
     * @param actionButton Data for the clicked button
     */
    /**
     * Emits event when user clicks on the button
     * @param {?} event Contains default mouse click event
     * @param {?} actionButton Data for the clicked button
     * @return {?}
     */
    PlayerActionsComponent.prototype.onButtonClick = /**
     * Emits event when user clicks on the button
     * @param {?} event Contains default mouse click event
     * @param {?} actionButton Data for the clicked button
     * @return {?}
     */
    function (event, actionButton) {
        this.buttonClick.emit({ event: event, data: actionButton });
    };
    PlayerActionsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-player-actions',
                    template: "<div class=\"d-flex flex-w-wrap content-video__player__tools\">\r\n  <div *ngFor=\"let item of actionButtons\">\r\n    <button class=\"sb-btn sb-btn-link sb-btn-link-gray sb-btn-sm sb-btn-square\"\r\n      [class.sb-btn-link-disabled]=\"item?.disabled\" [hidden]=\"item?.disabled\" (click)=\"onButtonClick($event, item)\">\r\n      <i class=\"icon-svg icon-svg--sm icon-back\">\r\n        <svg class=\"icon icon-svg--primary\">\r\n          <use [attr.xlink:href]=\"item?.iconPath\"></use>\r\n        </svg>\r\n      </i>\r\n      <span class=\"fxsmall\">{{item?.label}}</span>\r\n    </button>\r\n  </div>\r\n\r\n  <div class=\"d-flex flex-ai-center content-video__player__screenpart\" *ngIf=\"showFullscreen\">\r\n    <div class=\"separator\"></div>\r\n    <button class=\"sb-btn sb-btn-link sb-btn-link-gray sb-btn-sm sb-btn-square\"\r\n      [class.sb-btn-link-disabled]=\"fullscreenButton?.disabled\" [hidden]=\"fullscreenButton?.disabled\"\r\n      (click)=\"onButtonClick($event, fullscreenButton)\">\r\n      <i class=\"icon-svg icon-svg--sm icon-back\">\r\n        <svg class=\"icon icon-svg--primary\">\r\n          <use [attr.xlink:href]=\"fullscreenButton?.iconPath\"></use>\r\n        </svg>\r\n      </i>\r\n      <span class=\"fxsmall\">{{fullscreenButton?.label}}</span>\r\n    </button>\r\n  </div>\r\n</div>",
                    styles: [":host .content-video__player__screenpart{padding-left:0}:host .content-video__player__screenpart .separator{border-left:.03125rem solid var(--gray-100);height:2.5rem;width:.0625rem}[dir=ltr] :host .content-video__player__screenpart .separator,[dir=rtl] :host .content-video__player__screenpart .separator{margin:0 .5rem}:host .sb-btn-link-disabled,:host .sb-btn-link-disabled[disabled]{opacity:.5;cursor:initial;background:0 0;color:var(--gray-200);border:none}:host .sb-btn-link-disabled:hover,:host .sb-btn-link-disabled[disabled]:hover{background:0 0;color:var(--gray-200);border:0 solid var(--gray-100);cursor:default;opacity:.5}:host .sb-btn-link-disabled:focus,:host .sb-btn-link-disabled[disabled]:focus{outline:0;background:0 0}"]
                }] }
    ];
    PlayerActionsComponent.propDecorators = {
        actionButtons: [{ type: Input }],
        buttonClick: [{ type: Output }]
    };
    return PlayerActionsComponent;
}());
export { PlayerActionsComponent };
if (false) {
    /** @type {?} */
    PlayerActionsComponent.prototype.actionButtons;
    /** @type {?} */
    PlayerActionsComponent.prototype.buttonClick;
    /** @type {?} */
    PlayerActionsComponent.prototype.showFullscreen;
    /** @type {?} */
    PlayerActionsComponent.prototype.fullscreenButton;
}
