// @project-sunbird/common-consumption-v8/esm5/lib/layout/library-cards-hlist/library-cards-hlist.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, TemplateRef } from '@angular/core';
import { defaultLibraryCardsGrid } from '../library-cards.data';
import { LibraryCardTypes, ButtonPosition } from '../../card/models';
import { staticContent } from './library-cards-hlist.data';
var LibraryCardsHlistComponent = /** @class */ (function () {
    function LibraryCardsHlistComponent() {
        this.contentList = staticContent;
        this.viewMoreButtonText = defaultLibraryCardsGrid.viewMoreButtonText;
        this.maxCardCount = 10;
        this.isLoading = true;
        this.isMenu = false;
        this.isMobile = false;
        this.isOffline = false;
        this.type = LibraryCardTypes.DESKTOP_TEXTBOOK;
        this.moreInfoLabel = 'Medium';
        this.section = '';
        this.btnlabel = '';
        this.btnicon = '';
        this.isSelected = false;
        this.viewMoreButtonPosition = ButtonPosition.TOP;
        this.viewMoreClick = new EventEmitter();
        this.cardClick = new EventEmitter();
        this.menuClick = new EventEmitter();
    }
    Object.defineProperty(LibraryCardsHlistComponent.prototype, "LibraryCardTypes", {
        get: /**
         * @return {?}
         */
        function () { return LibraryCardTypes; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LibraryCardsHlistComponent.prototype, "ButtonPosition", {
        get: /**
         * @return {?}
         */
        function () { return ButtonPosition; },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    LibraryCardsHlistComponent.prototype.onViewMoreClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.viewMoreClick.emit(event);
    };
    /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    LibraryCardsHlistComponent.prototype.onCardClick = /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    function (event, index) {
        this.cardClick.emit({ 'data': event.data, 'index': index });
    };
    /**
     * @param {?} event
     * @return {?}
     */
    LibraryCardsHlistComponent.prototype.onMenuClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.menuClick.emit(event);
    };
    /**
     * @param {?} maxCardCounter
     * @return {?}
     */
    LibraryCardsHlistComponent.prototype.range = /**
     * @param {?} maxCardCounter
     * @return {?}
     */
    function (maxCardCounter) {
        return new Array(maxCardCounter);
    };
    LibraryCardsHlistComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-library-cards-hlist',
                    template: "<div class=\"layout-course\" *ngIf=\"!isLoading\">\r\n    <div class=\"header\">\r\n        <div class=\"title\"> {{title}} <hr> </div>\r\n        <button class=\"sb-btn sb-btn-normal sb-btn-outline-secondary sb-btn-white\" *ngIf=\"contentList?.length > maxCardCount && viewMoreButtonPosition === ButtonPosition.TOP\" (click)=\"onViewMoreClick($event)\">\r\n            {{viewMoreButtonText}}\r\n        </button>\r\n    </div>\r\n\r\n    <div class=\"sb--scrolling-wrapper-flexbox\" *ngIf=\"!multiRow\">\r\n        <div *ngFor=\"let content of contentList | slice:0:(maxCardCount ? maxCardCount : contentList?.length); let index = index\" class=\"sb--card--container\">\r\n            <sb-library-card [indexToDisplay]=\"index\" [layoutConfig]=\"layoutConfig\" [content]=\"content\" [type]=\"type\" [isMobile]=\"isMobile\" [isOffline]=\"(isOffline && !content.isAvailableLocally)\" [moreInfoLabel]=\"moreInfoLabel\" [section]=\"section\" [btnlabel]=\"btnlabel\" [btnicon]=\"btnicon\" [isMenu]=\"isMenu\" [isSelected]=\"isSelected\" (cardClick)=\"onCardClick($event, index)\" (menuClick)=\"onMenuClick($event)\" [cardImg]=\"content?.cardImg || content?.contentData?.cardImg\"></sb-library-card>\r\n        </div>\r\n        <div class=\"view-all\">\r\n            <button class=\"sb-btn sb-btn-normal sb-btn-outline-secondary sb-btn-white sb-btn-oval\" *ngIf=\"contentList?.length > maxCardCount && viewMoreButtonPosition === ButtonPosition.RIGHT\" (click)=\"onViewMoreClick($event)\">\r\n                {{viewMoreButtonText}}\r\n            </button>\r\n        </div>\r\n    </div>\r\n    \r\n    <div class=\"sb--scrolling-wrapper-flexbox\" *ngIf=\"multiRow\">\r\n        <span *ngFor=\"let content of contentList | slice:0:(maxCardCount ? maxCardCount : contentList?.length); let index = index\">\r\n            <div *ngIf=\"index%2 == 0\" class=\"sb--card--container\">\r\n                <sb-library-card class=\"sb--card--seperator\" *ngIf=\"contentList[index]!=null\" [indexToDisplay]=\"index\" [layoutConfig]=\"layoutConfig\" [content]=\"contentList[index]\" [type]=\"type\" [isMobile]=\"isMobile\" [isOffline]=\"isOffline && !content.isAvailableLocally\" [moreInfoLabel]=\"moreInfoLabel\" [section]=\"section\" [btnlabel]=\"btnlabel\" [btnicon]=\"btnicon\" [isMenu]=\"isMenu\" [isSelected]=\"isSelected\" (cardClick)=\"onCardClick($event, index)\" (menuClick)=\"onMenuClick($event)\" [cardImg]=\"contentList[index]?.cardImg || contentList[index]?.contentData?.cardImg\"></sb-library-card>\r\n                <sb-library-card class=\"sb--card--seperator\" *ngIf=\"index+1 < contentList.length\" [indexToDisplay]=\"index+1\" [layoutConfig]=\"layoutConfig\" [content]=\"contentList[index+1]\" [type]=\"type\" [isMobile]=\"isMobile\" [isOffline]=\"isOffline && !content.isAvailableLocally\" [moreInfoLabel]=\"moreInfoLabel\" [section]=\"section\" [btnlabel]=\"btnlabel\" [btnicon]=\"btnicon\" [isMenu]=\"isMenu\" [isSelected]=\"isSelected\" (cardClick)=\"onCardClick($event, index+1)\" (menuClick)=\"onMenuClick($event)\" [cardImg]=\"contentList[index+1]?.cardImg || contentList[index+1]?.contentData?.cardImg\"></sb-library-card>\r\n            </div>\r\n        </span>\r\n        <div class=\"view-all\">\r\n            <button class=\"sb-btn sb-btn-normal sb-btn-outline-secondary sb-btn-white sb-btn-oval\" *ngIf=\"contentList?.length > maxCardCount && viewMoreButtonPosition === ButtonPosition.RIGHT\" (click)=\"onViewMoreClick($event)\">\r\n                {{viewMoreButtonText}}\r\n            </button>\r\n        </div>\r\n    </div>\r\n</div>\r\n\r\n<div class=\"layout-course\" *ngIf=\"isLoading\">\r\n    <div class=\"header\" style=\"width:120px;\">\r\n        <div class=\"title loading-animation sb--card--highlighter\"></div>\r\n    </div>\r\n\r\n    <div class=\"sb--scrolling-wrapper-flexbox\" *ngIf=\"maxCardCount > 0 && !multiRow\">\r\n        <div *ngFor=\"let content of range(maxCardCount); let index=index\" class=\"sb--card--container\">\r\n            <sb-library-card [indexToDisplay]=\"index\" [isLoading]=\"isLoading\" [type]=\"type\"></sb-library-card>\r\n        </div>\r\n    </div>\r\n\r\n    <div class=\"sb--scrolling-wrapper-flexbox\" *ngIf=\"maxCardCount > 0 && multiRow\">\r\n        <div *ngFor=\"let content of range(maxCardCount); let index=index\" class=\"sb--card--container\">\r\n            <sb-library-card [indexToDisplay]=\"index\" [isLoading]=\"isLoading\" [type]=\"type\"></sb-library-card>\r\n            <sb-library-card [indexToDisplay]=\"index\" [isLoading]=\"isLoading\" [type]=\"type\"></sb-library-card>\r\n        </div>\r\n    </div>\r\n\r\n</div>\r\n",
                    styles: [".loading-animation{-webkit-animation-duration:1s;animation-duration:1s;-webkit-animation-fill-mode:forwards;animation-fill-mode:forwards;-webkit-animation-iteration-count:infinite;animation-iteration-count:infinite;-webkit-animation-name:placeload;animation-name:placeload;-webkit-animation-timing-function:linear;animation-timing-function:linear;background:var(--lazyload-bg);background:var(--grey-nuance-lighter);background:linear-gradient(to right,var(--grey-nuance-lighter) 8%,var(--grey-nuance-darker) 18%,var(--grey-nuance-lighter) 33%);background-size:75rem 6.5rem;position:relative}@-webkit-keyframes placeload{0%{background-position:-29.25rem 0}100%{background-position:29.25rem 0}}@keyframes placeload{0%{background-position:-29.25rem 0}100%{background-position:29.25rem 0}}.sb--card--highlighter{height:.75rem;width:100%;margin-top:.3125rem}:host .layout-library{display:flex;flex-direction:column}:host .sb--scrolling-wrapper-flexbox{display:flex;flex-wrap:nowrap;overflow-x:auto;-webkit-overflow-scrolling:touch;padding-bottom:.5em}:host .sb--card--container{flex:0 0 auto;width:17.5rem;margin:0 .5rem 0 0}html[dir=rtl] :host .sb--card--container{margin-right:inherit;margin-left:1rem}:host .header{display:flex;flex-direction:row;padding:0;align-items:center}[dir=ltr] :host .header,[dir=rtl] :host .header{margin:.7rem 0 .5rem}:host .header .title{font-family:inherit;text-overflow:ellipsis;overflow:hidden;flex:1;height:1.5rem;color:var(--black);font-size:1.125rem;font-weight:700;text-transform:capitalize;line-height:1.625rem;display:flex;align-items:center}:host .header .title hr{flex:1;margin-left:1rem;display:none}:host .sb-btn-oval{border-radius:50%;padding:0;height:80%;top:10%}"]
                }] }
    ];
    LibraryCardsHlistComponent.propDecorators = {
        contentList: [{ type: Input }],
        title: [{ type: Input }],
        viewMoreButtonText: [{ type: Input }],
        maxCardCount: [{ type: Input }],
        isLoading: [{ type: Input }],
        isMenu: [{ type: Input }],
        layoutConfig: [{ type: Input }],
        multiRow: [{ type: Input }],
        isMobile: [{ type: Input }],
        isOffline: [{ type: Input }],
        type: [{ type: Input }],
        moreInfoLabel: [{ type: Input }],
        section: [{ type: Input }],
        gridTemplate: [{ type: Input, args: ['hover-template',] }],
        btnlabel: [{ type: Input }],
        btnicon: [{ type: Input }],
        indexToDisplay: [{ type: Input }],
        isSelected: [{ type: Input }],
        viewMoreButtonPosition: [{ type: Input }],
        viewMoreClick: [{ type: Output }],
        cardClick: [{ type: Output }],
        menuClick: [{ type: Output }]
    };
    return LibraryCardsHlistComponent;
}());
export { LibraryCardsHlistComponent };
if (false) {
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.contentList;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.title;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.viewMoreButtonText;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.maxCardCount;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.isLoading;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.isMenu;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.layoutConfig;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.multiRow;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.isMobile;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.isOffline;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.type;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.moreInfoLabel;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.section;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.gridTemplate;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.btnlabel;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.btnicon;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.indexToDisplay;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.isSelected;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.viewMoreButtonPosition;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.viewMoreClick;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.cardClick;
    /** @type {?} */
    LibraryCardsHlistComponent.prototype.menuClick;
}
