// @project-sunbird/common-consumption-v8/esm5/lib/accordion/accordion-item/accordion-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, Host, Output, EventEmitter } from '@angular/core';
import { AccordionComponent } from '../accordion.component';
var AccordionItemComponent = /** @class */ (function () {
    function AccordionItemComponent(accordion) {
        this.accordion = accordion;
        this.collapsedValue = true;
        this.collapsedChange = new EventEmitter();
    }
    Object.defineProperty(AccordionItemComponent.prototype, "collapsed", {
        /**
         * Item is collapse (`true` by default)
         * type {boolean}
         */
        get: /**
         * Item is collapse (`true` by default)
         * type {boolean}
         * @return {?}
         */
        function () {
            return this.collapsedValue;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            /** @type {?} */
            var newVal = convertToBoolProperty(val);
            if (this.collapsedValue === newVal) {
                return;
            }
            this.collapsedValue = newVal;
            this.collapsedChange.emit(this.collapsedValue);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AccordionItemComponent.prototype, "expanded", {
        /**
         * Item is expanded (`false` by default)
         * type {boolean}
         */
        get: /**
         * Item is expanded (`false` by default)
         * type {boolean}
         * @return {?}
         */
        function () {
            return !this.collapsed;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this.collapsedValue = !convertToBoolProperty(val);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    AccordionItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.accordion.openCloseItems
            .subscribe((/**
         * @param {?} collapsed
         * @return {?}
         */
        function (collapsed) {
            _this.collapsed = collapsed;
        }));
    };
    /**
     * Open/close the item
     */
    /**
     * Open/close the item
     * @return {?}
     */
    AccordionItemComponent.prototype.toggle = /**
     * Open/close the item
     * @return {?}
     */
    function () {
        // we need this temporary variable as `openCloseItems.next` will change current value we need to save
        /** @type {?} */
        var willSet = !this.collapsed;
        if (!this.accordion.multi) {
            this.accordion.openCloseItems.next(true);
        }
        this.collapsed = willSet;
    };
    /**
     * @return {?}
     */
    AccordionItemComponent.prototype.open = /**
     * @return {?}
     */
    function () {
        this.collapsed = false;
    };
    /**
     * @return {?}
     */
    AccordionItemComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        this.collapsed = true;
    };
    AccordionItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-accordion-item',
                    template: "<div class=\"sbaccordion sbaccordion--toc\">\r\n  <div class=\"sbaccordion__panel\">\r\n    <ng-content select=\"sb-accordion-header\"></ng-content>\r\n    <ng-content select=\"sb-accordion-body\"></ng-content>\r\n  </div>\r\n</div>",
                    styles: [".sbaccordion{width:100%;box-shadow:0 .1875rem .25rem 0 rgba(var(--rc-rgba-primary),.1)!important}.sbaccordion__panel{margin-bottom:.5rem}.sbaccordion__panel-header{background:var(--white);padding:.5rem 1rem;display:flex;align-items:center;justify-content:space-between;cursor:pointer;height:3.5rem;border-radius:.25rem;border:1px solid var(--primary-300)}.accordion-item-header-collapsed .sbaccordion__panel-header__title{font-weight:400}"]
                }] }
    ];
    /** @nocollapse */
    AccordionItemComponent.ctorParameters = function () { return [
        { type: AccordionComponent, decorators: [{ type: Host }] }
    ]; };
    AccordionItemComponent.propDecorators = {
        collapsed: [{ type: Input, args: ['collapsed',] }, { type: HostBinding, args: ['class.collapsed',] }],
        expanded: [{ type: Input, args: ['expanded',] }, { type: HostBinding, args: ['class.expanded',] }],
        collapsedChange: [{ type: Output }]
    };
    return AccordionItemComponent;
}());
export { AccordionItemComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AccordionItemComponent.prototype.collapsedValue;
    /** @type {?} */
    AccordionItemComponent.prototype.collapsedChange;
    /**
     * @type {?}
     * @private
     */
    AccordionItemComponent.prototype.accordion;
}
/**
 * @param {?} val
 * @return {?}
 */
export function convertToBoolProperty(val) {
    if (typeof val === 'string') {
        val = val.toLowerCase().trim();
        return (val === 'true' || val === '');
    }
    return !!val;
}
