// @project-sunbird/common-consumption-v8/esm5/lib/card/toc-card/toc-card.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { MimeTypeMasterData } from './../../pipes-module/mime-type';
import { Component, Input, EventEmitter, Output } from '@angular/core';
import { COMMON_CONSUMPTION_CONSTANTS } from '../../common-consumption.constants';
import { PlatformType, TocCardType } from '../models';
var TocCardComponent = /** @class */ (function () {
    function TocCardComponent() {
        this.type = TocCardType.TEXTBOOK;
        this.contentStatus = [];
        this.trackableDefaultImage = '';
        this.platform = PlatformType.MOBILE;
        this.tocCardClick = new EventEmitter();
        this.playButtonClick = new EventEmitter();
        this.scoreLabel = '';
        this.disabled = '';
        this.fallbackImg = COMMON_CONSUMPTION_CONSTANTS.TOC_CARD_FALLBACK_IMG;
        this.isContentStarted = false;
        this.isCourseCompleted = false;
        this.isLastAttempt = false;
        this.isDisabled = false;
    }
    Object.defineProperty(TocCardComponent.prototype, "TocCardType", {
        get: /**
         * @return {?}
         */
        function () { return TocCardType; },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TocCardComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.type === TocCardType.COURSE) {
            this.createIconMap();
        }
        this.getBestScore();
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.activeContent && _this.activeContent.sbUniqueIdentifier === _this.content.sbUniqueIdentifier) {
                _this.tocCardClick.emit({ event: {}, data: tslib_1.__assign({}, _this.content) });
            }
        }));
        if (!this.playBtnConfig) {
            this.playBtnConfig = {
                label: 'Play',
                show: false
            };
        }
        if (!this.platform) {
            this.platform = PlatformType.MOBILE;
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    TocCardComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes && changes.contentStatus) {
            if (this.type === TocCardType.COURSE) {
                this.createIconMap();
            }
            this.getBestScore();
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TocCardComponent.prototype.onTocCardClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                if (this.isLastAttempt) {
                    event.isLastAttempt = true;
                    this.tocCardClick.emit({ event: event, data: tslib_1.__assign({}, this.content) });
                }
                else if (this.isDisabled) {
                    event.isDisabled = true;
                    this.tocCardClick.emit({ event: event, data: tslib_1.__assign({}, this.content) });
                }
                else if (this.activeContent && this.activeContent.sbUniqueIdentifier !== this.content.sbUniqueIdentifier) {
                    this.tocCardClick.emit({ event: event, data: tslib_1.__assign({}, this.content) });
                }
                else if (this.type = TocCardType.COURSE) {
                    this.tocCardClick.emit({ event: event, data: tslib_1.__assign({}, this.content) });
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * @return {?}
     */
    TocCardComponent.prototype.createIconMap = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var mimeTypesData = MimeTypeMasterData;
        this.isCourseCompleted = this.contentStatus.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (item.contentId === _this.content.identifier && item.status === 2) {
                return true;
            }
        }));
        this.isContentStarted = !!(this.contentStatus.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (item.contentId === _this.content.identifier && item.status === 1) {
                return true;
            }
        })));
        if (this.isCourseCompleted) {
            this.content.appIcon = 'assets/common-consumption/images/sprite.svg#circle-with-check-symbol';
        }
        else {
            this.iconPathMap = [
                [mimeTypesData.COLLECTION, 'assets/common-consumption/images/sprite.svg#doc'],
                [mimeTypesData.VIDEO, 'assets/common-consumption/images/sprite.svg#play'],
                [mimeTypesData.AUDIO, 'assets/common-consumption/images/sprite.svg#play'],
                [mimeTypesData.INTERACTIVE, 'assets/common-consumption/images/sprite.svg#touch'],
                [mimeTypesData.DOCS, 'assets/common-consumption/images/sprite.svg#doc'],
            ];
            this.content.appIcon = this.getIconPath(this.content.mimeType);
        }
    };
    /**
     * @return {?}
     */
    TocCardComponent.prototype.getBestScore = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.contentStatus.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (item.contentId === _this.content.identifier && item.bestScore && _this.scoreLabel && _this.maxAttempts) {
                _this.bestScoreLabel = _this.scoreLabel + ' ' + (Math.round(item.bestScore.totalScore * 100) / 100).toString() + '/' + item.bestScore.totalMaxScore.toString();
                if (_this.maxAttempts - item.score.length === 1) {
                    _this.isLastAttempt = true;
                    _this.isDisabled = false;
                }
                if (item.score.length >= _this.maxAttempts) {
                    _this.isDisabled = true;
                    _this.isLastAttempt = false;
                }
            }
        }));
    };
    /**
     * @param {?} contentMimeType
     * @return {?}
     */
    TocCardComponent.prototype.getIconPath = /**
     * @param {?} contentMimeType
     * @return {?}
     */
    function (contentMimeType) {
        var e_1, _a;
        try {
            for (var _b = tslib_1.__values(this.iconPathMap), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = tslib_1.__read(_c.value, 2), key = _d[0], value = _d[1];
                if (key.indexOf(contentMimeType) >= 0) {
                    return value;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return '';
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TocCardComponent.prototype.onPlayButtonClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.playButtonClick.emit({ event: event, data: tslib_1.__assign({}, this.content) });
    };
    /**
     * @param {?} content
     * @return {?}
     */
    TocCardComponent.prototype.swapCard = /**
     * @param {?} content
     * @return {?}
     */
    function (content) {
        if (this.platform === PlatformType.MOBILE) {
            return false;
        }
        if (this.isTrackable(content)) {
            return false;
        }
        return true;
    };
    /**
     * @param {?} content
     * @return {?}
     */
    TocCardComponent.prototype.isTrackable = /**
     * @param {?} content
     * @return {?}
     */
    function (content) {
        if ((content.contentData && content.contentData.trackable && content.contentData.trackable.enabled === 'Yes') ||
            (content.trackable && content.trackable.enabled === 'Yes')) {
            return true;
        }
        return false;
    };
    /**
     * @param {?} content
     * @return {?}
     */
    TocCardComponent.prototype.getContentImage = /**
     * @param {?} content
     * @return {?}
     */
    function (content) {
        if (this.isTrackable(content) && this.trackableDefaultImage) {
            return this.trackableDefaultImage;
        }
        return content.appIcon || (content.contentData && content.contentData.appIcon);
    };
    TocCardComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-toc-card',
                    template: "<!-- For card type textbook -->\r\n<div *ngIf=\"type === TocCardType.TEXTBOOK\">\r\n  <div class=\"sbchapter__item\" (click)=\"onTocCardClick($event)\"\r\n    [ngClass]=\"{'sbchapter__item--active': activeContent?.sbUniqueIdentifier === content?.sbUniqueIdentifier}\">\r\n    <h4 class=\"sbchapter__item__title sbchapter__item__title text-left sb__ellipsis sb__ellipsis--two mr-8\" title=\"{{content?.name}}\">\r\n      {{content?.name}}\r\n    </h4>\r\n    <div class=\"sbchapter__img\"><img [src]=\"content?.appIcon || fallbackImg\" (error)=\"img.src = fallbackImg\" alt=\"\"\r\n        #img />\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<!-- For card type TRACKABLE -->\r\n<div *ngIf=\"type === TocCardType.TRACKABLE\">\r\n  <div class=\"sbchapter__item\" [ngClass]=\"{'sb-card-swap': swapCard(content)}\" (click)=\"onTocCardClick($event)\"\r\n    [ngClass]=\"{'sbchapter__item--active': activeContent?.sbUniqueIdentifier === content?.sbUniqueIdentifier}\">\r\n    <div class=\"relative\">\r\n      <img src=\"assets/common-consumption/images/completed.svg\" *ngIf=\"content?.isAvailableLocally\" class=\"sb_success_check\" alt=\"\">\r\n      <div class=\"sbchapter__img\">\r\n        <img [src]=\"getContentImage(content) || fallbackImg\" alt=\"\" #img />\r\n      </div>\r\n    </div>\r\n    <div class=\"arrange-card-items certificate-course-contentt pl-8\">\r\n      <h4 class=\"sbchapter__item__title text-left sb__ellipsis sb__ellipsis--two mr-8 ml-8\" title=\"{{content?.name}}\">\r\n        {{content?.name}}\r\n      </h4>\r\n      <h4 *ngIf=\"bestScoreLabel\" class=\"sbchapter__item__score pr-8 ml-auto\" title=\"{{bestScoreLabel}}\">\r\n        {{bestScoreLabel}}\r\n      </h4>\r\n    </div>\r\n    <button class=\"sbchapter__play__btn\" *ngIf=\"!isTrackable(content) && playBtnConfig.show\"\r\n      (click)=\"$event.stopImmediatePropagation(); $event.preventDefault(); onPlayButtonClick($event)\">\r\n      <svg class=\"icon icon-svg--primary\">\r\n        <use xlink:href=\"assets/common-consumption/images/sprite.svg#play\"></use>\r\n      </svg>\r\n      {{playBtnConfig.label}}\r\n    </button>\r\n  </div>\r\n</div>\r\n\r\n<!-- For card type course -->\r\n<div *ngIf=\"type === TocCardType.COURSE\">\r\n  <div class=\"sbchapter__item\"  (click)=\"onTocCardClick($event)\"\r\n  [ngClass]=\"[(activeContent?.sbUniqueIdentifier === content?.sbUniqueIdentifier) ? 'sbchapter__item--active': '', isDisabled ? disabled: '']\">\r\n  <div class=\"sbchapter__img background-transparent\" [ngClass]=\"{'sbchapter__img--inprogress':isContentStarted}\">\r\n    {{createIconMap()}}\r\n    <i class=\"icon-svg icon-svg--sm\" *ngIf=\"content?.appIcon\">\r\n      <svg class=\"icon icon-svg--primary\">\r\n        <use [attr.xlink:href]=\"content?.appIcon\"></use>\r\n      </svg>\r\n    </i>\r\n    <img [src]=\"fallbackImg\" *ngIf=\"!content?.appIcon\" alt=\"\" #img />\r\n  </div>\r\n  <div class=\"certificate-course-content pl-8\">\r\n    <div class=\"sbchapter__item__title text-left sb__ellipsis sb__ellipsis--two mr-8 pt-4\" title=\"{{content?.name}}\">\r\n      {{content?.name}}\r\n    </div>\r\n      <div *ngIf=\"bestScoreLabel\" class=\"sbchapter__item__score pr-8 ml-auto\" title=\"{{bestScoreLabel}}\">\r\n        {{bestScoreLabel}}\r\n      </div>\r\n  </div>\r\n  </div>\r\n</div>\r\n\r\n<!-- For card type NON TRACKABLE -->\r\n<div *ngIf=\"type === TocCardType.NON_TRACKABLE\">\r\n  <div class=\"sbchapter__item\"  (click)=\"onTocCardClick($event)\"\r\n    [ngClass]=\"[(activeContent?.sbUniqueIdentifier === content?.sbUniqueIdentifier) ? 'sbchapter__item--active': '', isDisabled ? disabled: '']\">\r\n    <div class=\"sbchapter__img background-transparent\" [ngClass]=\"{'sbchapter__img--inprogress':isContentStarted}\">\r\n      {{createIconMap()}}\r\n      <i class=\"icon-svg icon-svg--sm\" *ngIf=\"content?.appIcon\">\r\n        <svg class=\"icon icon-svg--primary\">\r\n          <use [attr.xlink:href]=\"content?.appIcon\"></use>\r\n        </svg>\r\n      </i>\r\n      <img [src]=\"fallbackImg\" *ngIf=\"!content?.appIcon\" alt=\"\" #img />\r\n    </div>\r\n    <div class=\"certificate-course-content pl-8\">\r\n      <div class=\"sbchapter__item__title text-left sb__ellipsis sb__ellipsis--two mr-8 pt-4\" title=\"{{content?.name}}\">\r\n        {{content?.name}}\r\n      </div>\r\n      <div *ngIf=\"bestScoreLabel\" class=\"sbchapter__item__score pr-8 ml-auto\" title=\"{{bestScoreLabel}}\">\r\n        {{bestScoreLabel}}\r\n      </div> \r\n    </div>\r\n    <button class=\"sbchapter__play__btn\" *ngIf=\"!(content?.contentData?.trackable?.enabled === 'Yes') && playBtnConfig.show\" \r\n      (click)=\"$event.stopImmediatePropagation(); $event.preventDefault(); onPlayButtonClick($event)\">\r\n      <svg class=\"icon icon-svg--primary\">\r\n        <use xlink:href=\"assets/common-consumption/images/sprite.svg#play\"></use>\r\n      </svg>\r\n      {{playBtnConfig.label}}\r\n    </button>\r\n  </div>\r\n</div>",
                    styles: [".sbchapter__item{background:var(--white);border:.0625rem solid var(--primary-300);box-shadow:0 .1875rem .25rem 0 rgba(var(--rc-rgba-primary),.1);display:flex;align-items:center;margin-bottom:.5rem;padding:.5rem;cursor:pointer;min-height:3.5rem;transition:.2s ease-in}.sbchapter__item .sbchapter__img{width:1.5rem;height:1.5rem;flex:0 0 1.5rem;display:flex;align-items:center;background:var(--rc-dddddd)}.sbchapter__item .sbchapter__img img{width:100%;max-height:100%;-o-object-fit:cover;object-fit:cover}.sbchapter__item .sbchapter__img .icon-svg svg.icon.icon-svg--primary{fill:#9cb9ff}.sbchapter__item .sbchapter__img--inprogress .icon-svg svg.icon.icon-svg--primary{fill:var(--primary-color)}.sbchapter__item--active,.sbchapter__item--active:hover,.sbchapter__item:hover{background:var(--primary-100)}.sbchapter__item--active .sbchapter__item__title{font-weight:700}.sbchapter__item__title{font-size:.75rem;line-height:1.25rem;color:var(--primary-400);margin:0 2rem 0 0;padding:0;font-weight:400}.sbchapter__item__score{text-transform:uppercase;color:#3f19c3;font-size:.6875rem;text-align:right;white-space:nowrap}.certificate-course-content{display:flex;align-items:center;width:100%}@media (max-width:767px){.sbchapter__item__title{margin-right:0;margin-bottom:.5rem}.sbchapter__item__score{text-align:left}.certificate-course-content{display:block}}.pl-8{padding-left:.5rem}.background-transparent{background:0 0!important}.sbchapter__play__btn{border:.0625rem solid var(--primary-color);border-radius:1rem;color:var(--primary-color);background-color:var(--white);padding:.25rem}.sbchapter__play__btn svg{height:1rem;width:1rem;vertical-align:middle}.sb-card-swap{flex-direction:row-reverse}.sb-card-swap .arrange-card-items{display:contents}.sb_success_check{position:absolute;top:-.3rem;right:-.3rem}.pr-8{padding-right:.5rem}.ml-auto{margin-left:auto}"]
                }] }
    ];
    TocCardComponent.propDecorators = {
        content: [{ type: Input }],
        activeContent: [{ type: Input }],
        type: [{ type: Input }],
        contentStatus: [{ type: Input }],
        refresh: [{ type: Input }],
        playBtnConfig: [{ type: Input }],
        trackableDefaultImage: [{ type: Input }],
        platform: [{ type: Input }],
        tocCardClick: [{ type: Output }],
        playButtonClick: [{ type: Output }],
        scoreLabel: [{ type: Input }],
        maxAttempts: [{ type: Input }],
        disabled: [{ type: Input }]
    };
    return TocCardComponent;
}());
export { TocCardComponent };
if (false) {
    /** @type {?} */
    TocCardComponent.prototype.content;
    /** @type {?} */
    TocCardComponent.prototype.activeContent;
    /** @type {?} */
    TocCardComponent.prototype.type;
    /** @type {?} */
    TocCardComponent.prototype.contentStatus;
    /** @type {?} */
    TocCardComponent.prototype.refresh;
    /** @type {?} */
    TocCardComponent.prototype.playBtnConfig;
    /** @type {?} */
    TocCardComponent.prototype.trackableDefaultImage;
    /** @type {?} */
    TocCardComponent.prototype.platform;
    /** @type {?} */
    TocCardComponent.prototype.tocCardClick;
    /** @type {?} */
    TocCardComponent.prototype.playButtonClick;
    /** @type {?} */
    TocCardComponent.prototype.scoreLabel;
    /** @type {?} */
    TocCardComponent.prototype.maxAttempts;
    /** @type {?} */
    TocCardComponent.prototype.disabled;
    /** @type {?} */
    TocCardComponent.prototype.fallbackImg;
    /** @type {?} */
    TocCardComponent.prototype.iconPathMap;
    /** @type {?} */
    TocCardComponent.prototype.isContentStarted;
    /** @type {?} */
    TocCardComponent.prototype.isCourseCompleted;
    /** @type {?} */
    TocCardComponent.prototype.bestScoreLabel;
    /** @type {?} */
    TocCardComponent.prototype.isLastAttempt;
    /** @type {?} */
    TocCardComponent.prototype.isDisabled;
}
