// common-form-elements/esm5/lib/dynamic-textarea/dynamic-textarea.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic-textarea/dynamic-textarea.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { FormControl } from '@angular/forms';
import { FieldConfigValidationType } from '../common-form-config';
import { map, startWith } from 'rxjs/operators';
var DynamicTextareaComponent = /** @class */ (function () {
    function DynamicTextareaComponent() {
    }
    /**
     * @return {?}
     */
    DynamicTextareaComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var maxLengthValidation = this.field.validations &&
            this.field.validations.find((/**
             * @param {?} validation
             * @return {?}
             */
            function (validation) { return validation.type === FieldConfigValidationType.MAXLENGTH; }));
        if (maxLengthValidation) {
            this.remainderValidLength$ = this.formControlRef.valueChanges.pipe(startWith(''), map((/**
             * @param {?} v
             * @return {?}
             */
            function (v) { return ((typeof maxLengthValidation.value === 'number') ? maxLengthValidation.value : 0) - (v || '').length; })));
        }
    };
    DynamicTextareaComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-dynamic-textarea',
                    template: "<div class=\"sb-textarea-container\">\n  <label>{{label}}</label>\n  <textarea class=\"sb-textarea\" [formControl]=\"formControlRef\" rows=\"5\" name=\"sb-textarea\" placeholder={{placeholder}} [attr.disabled]=\"disabled ? true : null\">\n  </textarea>\n  <div class=\"remaining-length\" *ngIf=\"remainderValidLength$\">\n     {{remainderValidLength$ | async}} Characters left\n  </div>\n  <ng-container *ngFor=\"let validation of validations\">\n    <div class=\"cf-error\"\n      *ngIf=\"(validation.type && (validation.type).toLowerCase() && validation.message && formControlRef.errors && formControlRef.errors[(validation.type).toLowerCase()] && (formControlRef.dirty || formControlRef.touched))\">\n      {{ validation.message }}\n    </div>\n  </ng-container>\n</div>\n",
                    styles: ["label{display:block;padding:8px 0 0;font-size:14px}.sb-textarea{width:100%;padding:8px;box-sizing:border-box;border:1px solid #ccc;border-radius:.25rem;resize:vertical}.sb-textarea-container{margin-bottom:10px;position:relative}.remaining-length{position:absolute;bottom:.5rem;right:1rem;font-size:10px;color:#999}::-webkit-input-placeholder{padding:.25rem;color:#999;font-family:\"Noto Sans\";font-size:14px;letter-spacing:0;line-height:20px}::-moz-placeholder{padding:.25rem;color:#999;font-family:\"Noto Sans\";font-size:14px;letter-spacing:0;line-height:20px}:-ms-input-placeholder{padding:.25rem;color:#999;font-family:\"Noto Sans\";font-size:14px;letter-spacing:0;line-height:20px}::-ms-input-placeholder{padding:.25rem;color:#999;font-family:\"Noto Sans\";font-size:14px;letter-spacing:0;line-height:20px}::placeholder{padding:.25rem;color:#999;font-family:\"Noto Sans\";font-size:14px;letter-spacing:0;line-height:20px}.cf-error{color:red;font-family:\"Noto Sans\";font-size:12px}"]
                }] }
    ];
    /** @nocollapse */
    DynamicTextareaComponent.ctorParameters = function () { return []; };
    DynamicTextareaComponent.propDecorators = {
        label: [{ type: Input }],
        placeholder: [{ type: Input }],
        formControlRef: [{ type: Input }],
        field: [{ type: Input }],
        validations: [{ type: Input }],
        default: [{ type: Input }],
        disabled: [{ type: Input }]
    };
    return DynamicTextareaComponent;
}());
export { DynamicTextareaComponent };
if (false) {
    /** @type {?} */
    DynamicTextareaComponent.prototype.label;
    /** @type {?} */
    DynamicTextareaComponent.prototype.placeholder;
    /** @type {?} */
    DynamicTextareaComponent.prototype.formControlRef;
    /** @type {?} */
    DynamicTextareaComponent.prototype.field;
    /** @type {?} */
    DynamicTextareaComponent.prototype.validations;
    /** @type {?} */
    DynamicTextareaComponent.prototype.default;
    /** @type {?} */
    DynamicTextareaComponent.prototype.disabled;
    /** @type {?} */
    DynamicTextareaComponent.prototype.remainderValidLength$;
}
