// common-form-elements/esm5/lib/dynamic-checkbox/dynamic-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic-checkbox/dynamic-checkbox.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';
import { Subject } from 'rxjs';
import * as _ from 'lodash-es';
import { ValueComparator } from '../utilities/value-comparator';
var DynamicCheckboxComponent = /** @class */ (function () {
    function DynamicCheckboxComponent() {
        this.ValueComparator = ValueComparator;
        this.onChangeFilter = new EventEmitter();
        this.dependencyTerms = [];
        this._ = _;
    }
    /**
     * @return {?}
     */
    DynamicCheckboxComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var desiredValue = (this.field.dataType === 'text') ? (this.field.default === 'Yes' ? true : false) : !!this.field.default;
        this.checked = desiredValue;
    };
    /**
     * @param {?} checked
     * @return {?}
     */
    DynamicCheckboxComponent.prototype.onChecklistChange = /**
     * @param {?} checked
     * @return {?}
     */
    function (checked) {
        this.checked = checked;
        /** @type {?} */
        var desiredValue = (this.field.dataType === 'text') ? (checked === true ? 'Yes' : 'No') : checked;
        this.formControlRef.setValue(desiredValue);
    };
    DynamicCheckboxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-dynamic-checkbox',
                    template: "<div class=\"sb-checkbox sb-checkbox-primary\">\n  <input [attr.disabled]=\"disabled ? true : null\" type=\"checkbox\" [formControl]=\"formControlRef\" id=\"{{_.kebabCase(field.name) || field.code}}\"\n  name=\"{{_.kebabCase(field.name) || field.code}}\" [checked]=\"checked\" (change)=\"onChecklistChange($event.target.checked)\">\n  <label *ngIf=\"label\" for=\"{{_.kebabCase(field.name) || field.code}}\">{{label}}</label>\n</div>\n",
                    styles: ["label{font-size:16px}.sb-checkbox{padding:0 0 0 8px}.container{position:relative;cursor:pointer;font-size:22px;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.container input{position:absolute;opacity:0;cursor:pointer;height:0;width:0}.container .checkmark{position:absolute;top:0;left:0;height:22px;width:22px;border:2px solid #000;border-radius:4px}.container input:checked~.checkmark{background-color:#2196f3}.checkmark:after{content:\"\";position:absolute;display:none}.container input:checked~.checkmark:after{display:block}.container .checkmark:after{left:7px;top:4px;width:5px;height:9px;border:solid #fff;border-width:0 3px 3px 0;-webkit-transform:rotate(45deg);transform:rotate(45deg)}.checkbox-label{padding:2px 0 0 32px}.checkbox-row{display:-ms-grid;display:grid;-ms-grid-columns:10% 90%;grid-template-columns:10% 90%;grid-gap:0}.label-html{font-size:14px}"]
                }] }
    ];
    /** @nocollapse */
    DynamicCheckboxComponent.ctorParameters = function () { return []; };
    DynamicCheckboxComponent.propDecorators = {
        field: [{ type: Input }],
        disabled: [{ type: Input }],
        options: [{ type: Input }],
        label: [{ type: Input }],
        placeholder: [{ type: Input }],
        isMultiple: [{ type: Input }],
        context: [{ type: Input }],
        contextTerms: [{ type: Input }],
        formControlRef: [{ type: Input }],
        formGroup: [{ type: Input }],
        default: [{ type: Input }],
        contextData: [{ type: Input }],
        dataLoadStatusDelegate: [{ type: Input }],
        type: [{ type: Input }],
        styleClass: [{ type: Input }],
        onChangeFilter: [{ type: Output }],
        validations: [{ type: Input }],
        depends: [{ type: Input }],
        dependencyTerms: [{ type: Input }]
    };
    return DynamicCheckboxComponent;
}());
export { DynamicCheckboxComponent };
if (false) {
    /** @type {?} */
    DynamicCheckboxComponent.prototype.ValueComparator;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.field;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.disabled;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.options;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.label;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.placeholder;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.isMultiple;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.context;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.contextTerms;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.formControlRef;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.formGroup;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.default;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.contextData;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.dataLoadStatusDelegate;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.type;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.styleClass;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.onChangeFilter;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.validations;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.depends;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.dependencyTerms;
    /** @type {?} */
    DynamicCheckboxComponent.prototype._;
    /** @type {?} */
    DynamicCheckboxComponent.prototype.checked;
}
