// @project-sunbird/common-consumption-v8/esm2015/lib/accordion/accordion.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { Subject } from 'rxjs';
export class AccordionComponent {
    constructor() {
        this.openCloseItems = new Subject();
        this.isHiidden = true;
        this.cardClick = new EventEmitter();
        this.multiValue = false;
    }
    /**
     * @return {?}
     */
    get multi() {
        return this.multiValue;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set multi(val) {
        this.multiValue = convertToBoolProperty(val);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * Opens all enabled accordion items.
     * @return {?}
     */
    openAll() {
        if (this.multi) {
            this.openCloseItems.next(false);
        }
    }
    /**
     * Closes all enabled accordion items.
     * @return {?}
     */
    closeAll() {
        this.openCloseItems.next(true);
    }
}
AccordionComponent.decorators = [
    { type: Component, args: [{
                selector: 'sb-accordion',
                template: "<div class=\"sb-accordion-item\">\r\n<ng-content select=\"sb-accordion-item\"></ng-content>\r\n</div>",
                styles: ["::ng-deep .expanded .sbaccordion{margin-bottom:1.5rem}"]
            }] }
];
/** @nocollapse */
AccordionComponent.ctorParameters = () => [];
AccordionComponent.propDecorators = {
    multi: [{ type: Input, args: ['multi',] }],
    isHiidden: [{ type: Input }],
    cardClick: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    AccordionComponent.prototype.openCloseItems;
    /** @type {?} */
    AccordionComponent.prototype.isHiidden;
    /** @type {?} */
    AccordionComponent.prototype.cardClick;
    /**
     * @type {?}
     * @private
     */
    AccordionComponent.prototype.multiValue;
}
/**
 * @param {?} val
 * @return {?}
 */
export function convertToBoolProperty(val) {
    if (typeof val === 'string') {
        val = val.toLowerCase().trim();
        return (val === 'true' || val === '');
    }
    return !!val;
}
