// common-form-elements/esm5/lib/keywords/keywords.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/keywords/keywords.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { FormControl } from '@angular/forms';
import * as _ from 'lodash-es';
var KeywordsComponent = /** @class */ (function () {
    function KeywordsComponent() {
        this.inputText = '';
    }
    /**
     * @return {?}
     */
    KeywordsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!_.isEmpty(this.default)) {
            this.items = this.default;
        }
    };
    KeywordsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-keywords',
                    template: "<div class=\"sb-keywords\">\n    <label *ngIf=\"label\">{{label}}</label>\n    <tag-input class=\"sb-keywordbox\" [formControl]=\"formControlRef\" [modelAsStrings]=\"true\" [(ngModel)]='items' [placeholder]=\"field.placeholder || 'Enter Keyword'\"\n    [disable]=\"disabled\"></tag-input>\n    <ng-container *ngFor=\"let validation of validations\">\n        <div class=\"cf-error\"\n          *ngIf=\"(validation.type && (validation.type).toLowerCase() && validation.message && formControlRef.errors && formControlRef.errors[(validation.type).toLowerCase()] && (formControlRef.dirty || formControlRef.touched))\">\n          {{ validation.message }}\n        </div>\n      </ng-container>\n</div>\n",
                    styles: [".sb-keywords{margin:1rem 0}label{display:block;font-size:1rem;margin:0}.sb-keywordbox{width:100%;padding:8px 16px;border-bottom:.5px solid #333;box-sizing:border-box}.cf-error{color:red;font-family:\"Noto Sans\";font-size:12px}"]
                }] }
    ];
    /** @nocollapse */
    KeywordsComponent.ctorParameters = function () { return []; };
    KeywordsComponent.propDecorators = {
        label: [{ type: Input }],
        placeholder: [{ type: Input }],
        formControlRef: [{ type: Input }],
        field: [{ type: Input }],
        validations: [{ type: Input }],
        disabled: [{ type: Input }],
        default: [{ type: Input }]
    };
    return KeywordsComponent;
}());
export { KeywordsComponent };
if (false) {
    /** @type {?} */
    KeywordsComponent.prototype.label;
    /** @type {?} */
    KeywordsComponent.prototype.placeholder;
    /** @type {?} */
    KeywordsComponent.prototype.formControlRef;
    /** @type {?} */
    KeywordsComponent.prototype.field;
    /** @type {?} */
    KeywordsComponent.prototype.validations;
    /** @type {?} */
    KeywordsComponent.prototype.disabled;
    /** @type {?} */
    KeywordsComponent.prototype.default;
    /** @type {?} */
    KeywordsComponent.prototype.items;
    /** @type {?} */
    KeywordsComponent.prototype.inputText;
}
