// @project-sunbird/common-consumption-v8/esm5/lib/card/models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function ICardClick() { }
if (false) {
    /** @type {?} */
    ICardClick.prototype.event;
    /** @type {?} */
    ICardClick.prototype.data;
}
/**
 * @record
 */
export function IClick() { }
if (false) {
    /** @type {?} */
    IClick.prototype.event;
    /** @type {?} */
    IClick.prototype.data;
}
/** @enum {string} */
var LibraryCardTypes = {
    RECENTLY_VIEWED: 'recently_viewed',
    MOBILE_TEXTBOOK: 'mobile_textbook',
    DESKTOP_TEXTBOOK: 'desktop_textbook',
    QRCODE_RESULT: 'qrcode_result',
    PORTAL_QRCODE_FLATRESULT: 'portal_qrcode_flatresult',
};
export { LibraryCardTypes };
/** @enum {string} */
var CourseCardGridTypes = {
    MY_COURSE_CARD_GRID: "my_course_card_grid",
    COURSE_CARD_GRID: "course_card_grid",
    MY_COURSE_RECENTLY_VIEWED_CARD_GRID: "my_course_recently_viewed_card_grid",
    CURICULUM_COURSE_CARD_GRID: "curiculum_course_card_grid",
};
export { CourseCardGridTypes };
/** @enum {string} */
var CardGridTypes = {
    INFINITE_CARD_GRID: "infinite_grid",
};
export { CardGridTypes };
/** @enum {string} */
var CourseCardTypes = {
    RECENTLY_VIEWED: 'recently_viewed',
    MY_COURSE_CARD: 'my_course_card',
};
export { CourseCardTypes };
/** @enum {string} */
var LibraryCardGridTypes = {
    INFINITE_CARD_GRID: 'infinite_card_grid',
    INFINITE_CARD_GRID_WITH_HOVER: 'infinite_card_grid_with_hover',
};
export { LibraryCardGridTypes };
/** @enum {string} */
var LibraryCardStackTypes = {
    INFINITE_CARD_STACK: 'infinite_card_stack',
};
export { LibraryCardStackTypes };
/** @enum {string} */
var TocCardType = {
    TEXTBOOK: 'textbook',
    COURSE: 'course',
    TRACKABLE: 'trackable',
    NON_TRACKABLE: 'non_trackable',
};
export { TocCardType };
/**
 * @record
 */
export function IBadge() { }
if (false) {
    /** @type {?} */
    IBadge.prototype.issuerId;
    /** @type {?} */
    IBadge.prototype.assertionId;
    /** @type {?} */
    IBadge.prototype.badgeClassImage;
    /** @type {?} */
    IBadge.prototype.badgeId;
    /** @type {?} */
    IBadge.prototype.badgeClassName;
    /** @type {?} */
    IBadge.prototype.createdTS;
    /** @type {?} */
    IBadge.prototype.status;
}
/**
 * @record
 */
export function IContent() { }
if (false) {
    /** @type {?} */
    IContent.prototype.appIcon;
    /** @type {?} */
    IContent.prototype.name;
    /** @type {?} */
    IContent.prototype.subject;
    /** @type {?} */
    IContent.prototype.gradeLevel;
    /** @type {?} */
    IContent.prototype.badgeAssertions;
    /** @type {?} */
    IContent.prototype.resourceType;
    /** @type {?|undefined} */
    IContent.prototype.medium;
    /** @type {?|undefined} */
    IContent.prototype.creator;
    /** @type {?|undefined} */
    IContent.prototype.organisation;
    /** @type {?|undefined} */
    IContent.prototype.body;
    /** @type {?|undefined} */
    IContent.prototype.mimeType;
    /** @type {?|undefined} */
    IContent.prototype.contentType;
    /** @type {?|undefined} */
    IContent.prototype.identifier;
    /** @type {?|undefined} */
    IContent.prototype.versionKey;
    /** @type {?|undefined} */
    IContent.prototype.me_averageRating;
    /** @type {?|undefined} */
    IContent.prototype.description;
    /** @type {?|undefined} */
    IContent.prototype.artifactUrl;
    /** @type {?|undefined} */
    IContent.prototype.audience;
    /** @type {?|undefined} */
    IContent.prototype.code;
    /** @type {?|undefined} */
    IContent.prototype.collections;
    /** @type {?|undefined} */
    IContent.prototype.concepts;
    /** @type {?|undefined} */
    IContent.prototype.createdBy;
    /** @type {?|undefined} */
    IContent.prototype.createdOn;
    /** @type {?|undefined} */
    IContent.prototype.contributors;
    /** @type {?|undefined} */
    IContent.prototype.attributions;
    /** @type {?|undefined} */
    IContent.prototype.creators;
    /** @type {?|undefined} */
    IContent.prototype.owner;
    /** @type {?|undefined} */
    IContent.prototype.framework;
    /** @type {?|undefined} */
    IContent.prototype.language;
    /** @type {?|undefined} */
    IContent.prototype.languageCode;
    /** @type {?|undefined} */
    IContent.prototype.lastUpdatedOn;
    /** @type {?|undefined} */
    IContent.prototype.mediaType;
    /** @type {?|undefined} */
    IContent.prototype.organisationIds;
    /** @type {?|undefined} */
    IContent.prototype.osId;
    /** @type {?|undefined} */
    IContent.prototype.publisher;
    /** @type {?|undefined} */
    IContent.prototype.status;
    /** @type {?|undefined} */
    IContent.prototype.usedByContent;
    /** @type {?|undefined} */
    IContent.prototype.userId;
    /** @type {?|undefined} */
    IContent.prototype.userName;
    /** @type {?|undefined} */
    IContent.prototype.pkgVersion;
    /** @type {?|undefined} */
    IContent.prototype.visibility;
    /** @type {?|undefined} */
    IContent.prototype.hoverData;
}
/**
 * @record
 */
export function ICourse() { }
if (false) {
    /** @type {?} */
    ICourse.prototype.ownershipType;
    /** @type {?} */
    ICourse.prototype.copyright;
    /** @type {?|undefined} */
    ICourse.prototype.board;
    /** @type {?} */
    ICourse.prototype.channel;
    /** @type {?} */
    ICourse.prototype.downloadUrl;
    /** @type {?} */
    ICourse.prototype.organisation;
    /** @type {?} */
    ICourse.prototype.language;
    /** @type {?} */
    ICourse.prototype.mimeType;
    /** @type {?} */
    ICourse.prototype.variants;
    /** @type {?} */
    ICourse.prototype.leafNodes;
    /** @type {?} */
    ICourse.prototype.objectType;
    /** @type {?} */
    ICourse.prototype.appIcon;
    /** @type {?} */
    ICourse.prototype.children;
    /** @type {?} */
    ICourse.prototype.appId;
    /** @type {?} */
    ICourse.prototype.contentEncoding;
    /** @type {?} */
    ICourse.prototype.lockKey;
    /** @type {?} */
    ICourse.prototype.totalCompressedSize;
    /** @type {?} */
    ICourse.prototype.mimeTypesCount;
    /** @type {?} */
    ICourse.prototype.contentType;
    /** @type {?} */
    ICourse.prototype.contentCredits;
    /** @type {?} */
    ICourse.prototype.identifier;
    /** @type {?} */
    ICourse.prototype.lastUpdatedBy;
    /** @type {?} */
    ICourse.prototype.audience;
    /** @type {?} */
    ICourse.prototype.visibility;
    /** @type {?} */
    ICourse.prototype.toc_url;
    /** @type {?} */
    ICourse.prototype.contentTypesCount;
    /** @type {?} */
    ICourse.prototype.consumerId;
    /** @type {?} */
    ICourse.prototype.childNodes;
    /** @type {?} */
    ICourse.prototype.mediaType;
    /** @type {?} */
    ICourse.prototype.osId;
    /** @type {?} */
    ICourse.prototype.graph_id;
    /** @type {?} */
    ICourse.prototype.nodeType;
    /** @type {?} */
    ICourse.prototype.lastPublishedBy;
    /** @type {?} */
    ICourse.prototype.version;
    /** @type {?} */
    ICourse.prototype.license;
    /** @type {?} */
    ICourse.prototype.prevState;
    /** @type {?} */
    ICourse.prototype.lastPublishedOn;
    /** @type {?} */
    ICourse.prototype.size;
    /** @type {?} */
    ICourse.prototype.IL_FUNC_OBJECT_TYPE;
    /** @type {?} */
    ICourse.prototype.name;
    /** @type {?} */
    ICourse.prototype.topic;
    /** @type {?} */
    ICourse.prototype.c_diksha_load_open_batch_count;
    /** @type {?} */
    ICourse.prototype.status;
    /** @type {?} */
    ICourse.prototype.code;
    /** @type {?} */
    ICourse.prototype.purpose;
    /** @type {?} */
    ICourse.prototype.prevStatus;
    /** @type {?} */
    ICourse.prototype.description;
    /** @type {?} */
    ICourse.prototype.medium;
    /** @type {?} */
    ICourse.prototype.idealScreenSize;
    /** @type {?} */
    ICourse.prototype.createdOn;
    /** @type {?} */
    ICourse.prototype.reservedDialcodes;
    /** @type {?} */
    ICourse.prototype.copyrightYear;
    /** @type {?} */
    ICourse.prototype.contentDisposition;
    /** @type {?} */
    ICourse.prototype.lastUpdatedOn;
    /** @type {?} */
    ICourse.prototype.SYS_INTERNAL_LAST_UPDATED_ON;
    /** @type {?} */
    ICourse.prototype.dialcodeRequired;
    /** @type {?} */
    ICourse.prototype.creator;
    /** @type {?} */
    ICourse.prototype.createdFor;
    /** @type {?} */
    ICourse.prototype.lastStatusChangedOn;
    /** @type {?} */
    ICourse.prototype.IL_SYS_NODE_TYPE;
    /** @type {?} */
    ICourse.prototype.os;
    /** @type {?} */
    ICourse.prototype.pkgVersion;
    /** @type {?} */
    ICourse.prototype.versionKey;
    /** @type {?} */
    ICourse.prototype.idealScreenDensity;
    /** @type {?} */
    ICourse.prototype.framework;
    /** @type {?} */
    ICourse.prototype.depth;
    /** @type {?} */
    ICourse.prototype.s3Key;
    /** @type {?} */
    ICourse.prototype.dialcodes;
    /** @type {?} */
    ICourse.prototype.lastSubmittedOn;
    /** @type {?} */
    ICourse.prototype.createdBy;
    /** @type {?} */
    ICourse.prototype.compatibilityLevel;
    /** @type {?} */
    ICourse.prototype.leafNodesCount;
    /** @type {?} */
    ICourse.prototype.IL_UNIQUE_ID;
    /** @type {?} */
    ICourse.prototype.resourceType;
    /** @type {?} */
    ICourse.prototype.node_id;
}
/**
 * @record
 */
export function IHoverData() { }
if (false) {
    /** @type {?|undefined} */
    IHoverData.prototype.note;
    /** @type {?} */
    IHoverData.prototype.actions;
    /** @type {?} */
    IHoverData.prototype.hoverData;
}
/** @enum {string} */
var FlattenedType = {
    EXPAND: 'expand',
    COLLAPSE: 'collapse',
};
export { FlattenedType };
/**
 * @record
 */
export function IButtonConfig() { }
if (false) {
    /** @type {?|undefined} */
    IButtonConfig.prototype.label;
    /** @type {?|undefined} */
    IButtonConfig.prototype.show;
}
/** @enum {string} */
var PlatformType = {
    PORTAL: 'portal',
    MOBILE: 'mobile',
};
export { PlatformType };
/** @enum {string} */
var NotificationStatus = {
    READ: 'read',
    UNREAD: 'unread',
};
export { NotificationStatus };
/**
 * @record
 */
export function NotificationViewConfig() { }
if (false) {
    /** @type {?} */
    NotificationViewConfig.prototype.title;
    /** @type {?} */
    NotificationViewConfig.prototype.subTitle;
    /** @type {?} */
    NotificationViewConfig.prototype.clearText;
    /** @type {?} */
    NotificationViewConfig.prototype.moreText;
    /** @type {?} */
    NotificationViewConfig.prototype.lessText;
    /** @type {?} */
    NotificationViewConfig.prototype.minNotificationViewCount;
}
/** @enum {string} */
var ExpandMode = {
    SINGLE: 'single',
    MULTIPLE: 'multiple',
};
export { ExpandMode };
/** @enum {string} */
var ExpandBehavior = {
    NONE: 'none',
    EXPAND_FIRST: 'expand-first',
};
export { ExpandBehavior };
/**
 * @record
 */
export function IAccordianConfig() { }
if (false) {
    /** @type {?} */
    IAccordianConfig.prototype.expandMode;
    /** @type {?} */
    IAccordianConfig.prototype.expandBehavior;
}
/** @enum {string} */
var ButtonPosition = {
    TOP: "top",
    RIGHT: "right",
    BOTTOM: "bottom",
};
export { ButtonPosition };
