// common-form-elements/esm2015/lib/utilities/value-comparator.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utilities/value-comparator.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class ValueComparator {
    /**
     * @param {?} v1
     * @param {?} v2
     * @return {?}
     */
    static valueComparator(v1, v2) {
        if (typeof v1 === 'object' && typeof v2 === 'object') {
            return ObjectUtil.equals(v1, v2);
        }
        else if (v1 === v2) {
            return true;
        }
        else if (!v1 && !v2) {
            return true;
        }
        return false;
    }
}
class ObjectUtil {
    /**
     * @param {?} a
     * @param {?} b
     * @return {?}
     */
    static equals(a, b) {
        /** @type {?} */
        const countProps = (/**
         * @param {?} obj
         * @return {?}
         */
        (obj) => {
            /** @type {?} */
            let count = 0;
            for (const k in obj) {
                if (obj.hasOwnProperty(k)) {
                    count++;
                }
            }
            return count;
        });
        /** @type {?} */
        const objectEquals = (/**
         * @param {?} v1
         * @param {?} v2
         * @return {?}
         */
        (v1, v2) => {
            if (typeof (v1) !== typeof (v2)) {
                return false;
            }
            if (typeof (v1) === 'function') {
                return v1.toString() === v2.toString();
            }
            if (v1 instanceof Object && v2 instanceof Object) {
                if (countProps(v1) !== countProps(v2)) {
                    return false;
                }
                /** @type {?} */
                let r = true;
                for (const k in v1) {
                    r = objectEquals(v1[k], v2[k]);
                    if (!r) {
                        return false;
                    }
                }
                return true;
            }
            else {
                return v1 === v2;
            }
        });
        return objectEquals(a, b);
    }
    /**
     * @param {?} newObj
     * @param {?} oldObj
     * @return {?}
     */
    static getPropDiff(newObj, oldObj) {
        return Object.keys(newObj).reduce((/**
         * @param {?} acc
         * @param {?} key
         * @return {?}
         */
        (acc, key) => {
            if (ObjectUtil.equals(newObj[key], oldObj[key])) {
                return acc;
            }
            acc.push(key);
            return acc;
        }), []);
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    static getTruthyProps(obj) {
        return Object.keys(obj).filter((/**
         * @param {?} key
         * @return {?}
         */
        (key) => !!obj[key]));
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    static toOrderedString(obj) {
        return JSON.stringify(Object.keys(obj).sort().reduce((/**
         * @param {?} acc
         * @param {?} k
         * @return {?}
         */
        (acc, k) => {
            acc[k] = obj[k];
            return acc;
        }), {}));
    }
}
