// @project-sunbird/common-consumption-v8/esm5/lib/certificate/certificate-actions/certificate-actions/certificate-actions.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
/**
 * @record
 */
export function IPreviewActionButton() { }
if (false) {
    /** @type {?} */
    IPreviewActionButton.prototype.label;
    /** @type {?} */
    IPreviewActionButton.prototype.name;
    /** @type {?} */
    IPreviewActionButton.prototype.show;
}
/**
 * @record
 */
export function IPreviewActionButtonClick() { }
if (false) {
    /** @type {?} */
    IPreviewActionButtonClick.prototype.event;
    /** @type {?} */
    IPreviewActionButtonClick.prototype.name;
}
var CertificateActionsComponent = /** @class */ (function () {
    function CertificateActionsComponent(domSanitizer) {
        this.domSanitizer = domSanitizer;
        this.select = { show: false, label: 'Select', name: 'select' };
        this.preview = { show: false, label: 'Preview', name: 'preview' };
        this.remove = { show: false, label: 'Remove', name: 'remove' };
        this.isSelectedTemplate = false;
        this.actionButtons = [];
        this.buttonClick = new EventEmitter();
        this.showOverlay = true;
        this.templateUrl = '';
    }
    /**
     * @param {?} event
     * @param {?} name
     * @return {?}
     */
    CertificateActionsComponent.prototype.onButtonClick = /**
     * @param {?} event
     * @param {?} name
     * @return {?}
     */
    function (event, name) {
        if (name === 'select') {
            this.isSelectedTemplate = true;
            this.showOverlay = false;
        }
        this.buttonClick.emit({ event: event, name: name });
    };
    /**
     * @return {?}
     */
    CertificateActionsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.template) {
            this.templateUrl = this.domSanitizer.bypassSecurityTrustResourceUrl(this.template);
        }
    };
    CertificateActionsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-certificate-actions',
                    template: "<div class=\"select-template-container d-flex flex-ai-center flex-jc-center flex-dc mb-16\">\r\n  <div class=\"select-template-container__content\">\r\n    <div class=\"template-checkbox\">\r\n      <img *ngIf=\"isSelectedTemplate\" src=\"assets/common-consumption/images/checked-icon.svg\" width=\"18px\" height=\"18px\" alt=\"\">\r\n    </div>\r\n    <div class=\"template-image\">\r\n      <embed [src]=\"templateUrl\" class=\"template-img\" type=\"image/svg+xml\" /></div>\r\n  </div>\r\n  <div [ngClass]=\"{'overlay-container': showOverlay}\">\r\n    <div class=\"d-flex flex-ai-center flex-jc-center flex-dc overlay-buttons\">\r\n      <button type=\"button\" *ngIf=\"select?.show\" (click)=\"onButtonClick($event, select?.name)\"\r\n        class=\"d-flex sb-btn sb-btn-normal sb-btn-no-border mb-16\">\r\n        <img src=\"assets/common-consumption/images/select-icon.svg\" width=\"16px\" height=\"16px\" class=\"mr-8\" alt=\"\">\r\n        <span class=\"sb-color-primary font-weight-bold\">{{select?.label}}</span>\r\n      </button>\r\n\r\n      <button type=\"button\" *ngIf=\"remove?.show\" (click)=\"onButtonClick($event, remove?.name)\"\r\n        class=\"d-flex sb-btn sb-btn-normal sb-btn-no-border mb-16\">\r\n        <img src=\"assets/common-consumption/images/bin.svg\" width=\"16px\" height=\"16px\" class=\"mr-8\" alt=\"\">\r\n        <span class=\"sb-color-error font-weight-bold\">{{remove?.label}}</span>\r\n      </button>\r\n\r\n      <button type=\"button\" *ngIf=\"preview?.show\" (click)=\"onButtonClick($event, preview?.name)\"\r\n        class=\"d-flex sb-btn sb-btn-normal sb-btn-no-border\">\r\n        <img src=\"assets/common-consumption/images/preview-icon.svg\" width=\"14px\" height=\"14px\" class=\"mr-8\" alt=\"\">\r\n        <span class=\"sb-color-primary font-weight-bold\">{{preview?.label}}</span></button>\r\n    </div>\r\n  </div>\r\n</div>\r\n",
                    styles: [".select-template-container{position:relative}.select-template-container__content{display:block;width:100%;height:auto}.select-template-container__content .template-checkbox{position:absolute;top:.25rem;right:.25rem}.select-template-container__content .template-image{border-radius:.25rem;width:100%}.select-template-container__content .template-image .template-img{max-width:100%;border-radius:.25rem;width:100%;background:var(--white)}.overlay-container{position:absolute;top:0;bottom:0;left:0;right:0;height:100%;width:100%;opacity:0;transition:.5s;background-color:rgba(0,0,0,.2)}.select-template-container:hover .overlay-container{opacity:1}.overlay-buttons{position:absolute;width:100%;height:100%}.overlay-buttons .sb-btn-no-border{box-shadow:0 1.25rem 2.5rem -.625rem rgba(2,79,157,.3);background:var(--white);border:0}.sb-btn-select{padding:.5rem 1.3rem}.sb-btn-content{max-width:200px;white-space:normal;word-break:break-word;overflow:hidden;text-overflow:ellipsis;-webkit-line-clamp:2;display:-webkit-box;-webkit-box-orient:vertical}"]
                }] }
    ];
    /** @nocollapse */
    CertificateActionsComponent.ctorParameters = function () { return [
        { type: DomSanitizer }
    ]; };
    CertificateActionsComponent.propDecorators = {
        select: [{ type: Input }],
        preview: [{ type: Input }],
        remove: [{ type: Input }],
        template: [{ type: Input }],
        isSelectedTemplate: [{ type: Input }],
        actionButtons: [{ type: Input }],
        buttonClick: [{ type: Output }]
    };
    return CertificateActionsComponent;
}());
export { CertificateActionsComponent };
if (false) {
    /** @type {?} */
    CertificateActionsComponent.prototype.select;
    /** @type {?} */
    CertificateActionsComponent.prototype.preview;
    /** @type {?} */
    CertificateActionsComponent.prototype.remove;
    /** @type {?} */
    CertificateActionsComponent.prototype.template;
    /** @type {?} */
    CertificateActionsComponent.prototype.isSelectedTemplate;
    /** @type {?} */
    CertificateActionsComponent.prototype.actionButtons;
    /** @type {?} */
    CertificateActionsComponent.prototype.buttonClick;
    /** @type {?} */
    CertificateActionsComponent.prototype.showOverlay;
    /** @type {?} */
    CertificateActionsComponent.prototype.templateUrl;
    /** @type {?} */
    CertificateActionsComponent.prototype.domSanitizer;
}
