// @project-sunbird/common-consumption-v8/esm2015/lib/faq/faq/faq.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output, TemplateRef } from '@angular/core';
import { staticData } from './faq.component.data';
export class FaqComponent {
    constructor() {
        this.data = staticData;
        this.toggleGroupEvent = new EventEmitter();
        this.yesClickedEvent = new EventEmitter();
        this.noClickedEvent = new EventEmitter();
        this.submitClickedEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.constants = this.data.constants;
        this.faqs = this.data.faqs;
        // tslint:disable-next-line:prefer-for-of
        for (let i = 0; i < this.data.faqs.length; i++) {
            if (this.data.faqs[i].topic.includes('{{APP_NAME}}')) {
                this.data.faqs[i].topic = this.data.faqs[i].topic.replace('{{APP_NAME}}', this.appName);
            }
            else {
                this.data.faqs[i].topic = this.data.faqs[i].topic;
            }
            if (this.data.faqs[i].description.includes('{{APP_NAME}}')) {
                this.data.faqs[i].description = this.data.faqs[i].description.replace('{{APP_NAME}}', this.appName);
            }
            else {
                this.data.faqs[i].description = this.data.faqs[i].description;
            }
        }
    }
    /**
     * @param {?} i
     * @param {?} event
     * @return {?}
     */
    toggleGroup(i, event) {
        this.isNoClicked = false;
        this.isYesClicked = false;
        this.isSubmitted = false;
        /** @type {?} */
        let isCollapsed = true;
        this.value = {};
        this.value.action = 'toggle-clicked';
        this.value.position = i;
        this.value.value = {};
        this.value.value.topic = this.data.faqs[i].topic;
        this.value.value.description = this.data.faqs[i].description;
        if (this.isGroupShown(i)) {
            isCollapsed = false;
            this.shownGroup = null;
            this.value.isOpened = false;
        }
        else {
            isCollapsed = false;
            this.shownGroup = i;
            this.value.isOpened = true;
        }
        this.toggleGroupEvent.emit({ event: event, data: this.value });
    }
    // to check whether the card is toggled or not
    /**
     * @param {?} group
     * @return {?}
     */
    isGroupShown(group) {
        return this.shownGroup === group;
    }
    /**
     * @param {?} i
     * @param {?} event
     * @return {?}
     */
    noClicked(i, event) {
        this.value = {};
        if (!this.isNoClicked) {
            this.isNoClicked = true;
        }
        this.value.action = 'no-clicked';
        this.value.position = i;
        this.value.value = {};
        this.value.value.topic = this.data.faqs[i].topic;
        this.value.value.description = this.data.faqs[i].description;
        window.parent.postMessage(this.value, '*');
        this.noClickedEvent.emit({ event: event, data: this.value });
    }
    /**
     * @param {?} i
     * @param {?} event
     * @return {?}
     */
    yesClicked(i, event) {
        this.value = {};
        if (!this.isYesClicked) {
            this.isYesClicked = true;
        }
        this.value.action = 'yes-clicked';
        this.value.position = i;
        this.value.value = {};
        this.value.value.topic = this.data.faqs[i].topic;
        this.value.value.description = this.data.faqs[i].description;
        window.parent.postMessage(this.value, '*');
        this.yesClickedEvent.emit({ event: event, data: this.value });
    }
    /**
     * @param {?} textValue
     * @param {?} i
     * @param {?} event
     * @return {?}
     */
    submitClicked(textValue, i, event) {
        this.isSubmitted = true;
        this.value = {};
        this.value.action = 'submit-clicked';
        this.value.position = i;
        this.value.value = {};
        this.value.value.topic = this.data.faqs[i].topic;
        this.value.value.description = this.data.faqs[i].description;
        this.value.value.knowMoreText = textValue;
        window.parent.postMessage(this.value, '*');
        this.submitClickedEvent.emit({ event: event, data: this.value });
        this.textValue = '';
    }
}
FaqComponent.decorators = [
    { type: Component, args: [{
                selector: 'sb-faq',
                template: "<div class=\"panel panel-default\" *ngFor=\"let faq of faqs; let i = index\">\r\n  <div data-toggle=\"collapse\" (click)=\"toggleGroup(i,$event)\">\r\n    <div class=\"panel-heading\">\r\n      <div class=\"text-left\" [ngClass]=\"{'panel-title': !isGroupShown(i), 'panel-title-blue': isGroupShown(i)}\">\r\n        {{ faq?.topic}}\r\n        <span>\r\n          <i class=\"icon-svg icon-svg--xs ml-4\" *ngIf=\"!isGroupShown(i)\">\r\n            <svg class=\"icon icon-svg--primary\">\r\n              <use xlink:href=\"assets/common-consumption/images/sprite.svg#arrow-down\"></use>\r\n            </svg>\r\n          </i>\r\n          <i class=\"icon-svg icon-svg--xs ml-4\" *ngIf=\"isGroupShown(i)\">\r\n            <svg class=\"icon icon-svg--primary\">\r\n              <use xlink:href=\"assets/common-consumption/images/sprite.svg#arrow-up\"></use>\r\n            </svg>\r\n          </i>\r\n        </span>\r\n        <!--<img class=\"btn-arrow\" src=\"../../assets/faq/resources/images/Arrow.png\" *ngIf=\"!isGroupShown(i)\">\r\n        <img class=\"btn-arrow\" src=\"../../assets/faq/resources/images/UpArrow.png\" *ngIf=\"isGroupShown(i)\">-->\r\n      </div>\r\n    </div>\r\n  </div>\r\n  <div id=\"collapse' + key + '\" class=\"panel-collapse collapse\" *ngIf=\"isGroupShown(i)\">\r\n    <div class=\"panel-body\">\r\n      <p class=\"panelText text-left\" [innerHtml]=\"faq.description\"></p>\r\n    </div>\r\n    <div class=\"panel-interact\" *ngIf=\"!isNoClicked && !isYesClicked\">\r\n      <p class=\"panelText\">{{data?.constants?.helpMsg}}&lrm;</p>\r\n      <div class=\"d-flex flex-jc-center\">\r\n        <button type=\"button\" class=\"sb-btn sb-btn-normal sb-btn-success mr-8\" id=\"btn-yes\"\r\n          (click)=\"yesClicked(i,$event)\">{{data?.constants?.yesMsg}}</button>\r\n      <button type=\"button\" class=\"sb-btn sb-btn-normal sb-btn-outline-primary\" id=\"btn-no\"\r\n        (click)=\"noClicked(i,$event)\">{{data?.constants?.noMsg}}</button>\r\n        </div>\r\n    </div>\r\n    <div class=\"panel-interact\" *ngIf=\"isYesClicked || isSubmitted\">\r\n      <p class=\"yes-clicked\"> {{data?.constants?.thanksMsg}}</p>\r\n      <div *ngIf=\"isNoClicked && isSubmitted && extraTemplate\" class=\"extraTemplate\">\r\n        <ng-container *ngTemplateOutlet=\"extraTemplate\">\r\n        </ng-container>\r\n      </div>\r\n    </div>\r\n    <div class=\"panel-info\" *ngIf=\"!isSubmitted && isNoClicked\">\r\n      <h6>{{data?.constants?.sorryMsg}}</h6>\r\n      <p>{{data?.constants?.knowMoreMsg}}</p>\r\n      <form action=\"#\" id=\"know-more-form\" class=\"know-more-form\" *ngIf=\"!isSubmitted\">\r\n        <textarea type=\"text\" name=\"moreInfo\" placeholder=\"{{constants?.typeHere}}\" class=\"input-text\" maxlength=\"1000\"\r\n          [(ngModel)]=\"textValue\"></textarea>\r\n        <input type=\"submit\" value=\"{{constants?.submitButton}}\" class=\"submit-button\"\r\n          (click)=\"submitClicked(textValue,i,$event)\">\r\n      </form>\r\n      <p class=\"no-clicked\" hidden=\"true\" *ngIf=\"isSubmitted\"> {{data?.constants?.thanksMsg}} </p>\r\n    </div>\r\n  </div>\r\n</div>",
                styles: [".contentPadding{padding-bottom:0}.scroll-content{padding:0!important}iframe{transform:scale(1,1);height:100%;width:100%}body{background-color:var(--primary-100);font-family:var(--font-stack-en)}.panelText{font-size:.875rem;padding:1rem}.rotate{transform:rotate(180deg)}.btn-arrow{padding:.25rem;margin-top:.25rem;margin-left:.3125rem}.yesBtn{color:var(--secondary-color)}.btn{cursor:pointer}p{font-size:.875rem}.help-header,.help-header-send-email{background:var(--primary-color);color:var(--white);padding:1rem}.help-header-send-email>h4,.help-header>h4{padding:0;font-weight:700;max-width:64rem;margin:0 auto}.help-header-send-email>p,.help-header>p{font-weight:100;max-width:64rem;margin:0 auto}.panel-title{display:flex;align-items:flex-start;justify-content:space-between}.panel-title-blue{display:flex;align-items:flex-start;justify-content:space-between;color:var(--primary-color)}.info-msg{background-color:var(--primary-0)}.report-text{padding-left:.375rem}.icon-svg{position:relative;top:.0625rem;display:inline-block;font-style:normal;font-weight:400;line-height:1;-webkit-font-smoothing:antialiased;opacity:1}.icon-svg svg.icon{width:100%;height:inherit}.icon-svg--xs{width:var(--icon-svg-xs);height:var(--icon-svg-xs)}.icon-svg--sm{width:var(--icon-svg-sm);height:var(--icon-svg-sm)}.icon-svg--md{width:var(--icon-svg-md);height:var(--icon-svg-md)}.icon-svg--lg{width:var(--icon-svg-lg);height:var(--icon-svg-lg)}.icon-svg--xl{width:var(--icon-svg-xl);height:var(--icon-svg-xl)}.icon-svg--white{fill:var(--white)}.icon-svg--primary{fill:var(--primary-color)}.icon-svg--secondary{fill:var(--secondary-color)}.icon-svg--tertiary{fill:var(--tertiary-color)}.icon-svg--red{fill:var(--red-400)}.icon-svg--gray{fill:var(--gray-800)}@media only screen and (max-width:768px){.info-msg-send-email>p,.info-msg>p{padding:.5rem 1rem;max-width:64rem;margin:0 auto}}@media only screen and (min-width:768px){.info-msg-send-email>p,.info-msg>p{padding:1rem 0;max-width:64rem;margin:0 auto}}.icon{color:var(--primary-color);float:right}.panel{border-radius:0!important;border-top:0;border-left:0;border-right:0;box-shadow:none!important;border-bottom:.0625rem solid #d3d3d3;font-size:1rem}.panel-heading{padding:1rem;cursor:pointer}.panel-group{margin:0 auto;max-width:64rem}.panel-group .panel+.panel{border-radius:0;margin-top:0}.panel-default>.panel-heading{background-color:var(--white);border-radius:0!important}.panel-body{background-color:transparent}.panel-interact{text-align:center;padding:1.25rem}.panel-interact>button{margin:.625rem .3125rem 0}.panel-interact>p{margin:0}.no-clicked,.yes-clicked{padding:.625rem!important;margin:.625rem auto auto!important;background:var(--primary-0);width:12.5rem;border-radius:.25rem;color:var(--secondary-color);text-align:center}.panel-info{text-align:center;max-width:64rem;margin:0 auto}.panel-info>h6{font-weight:700;margin-top:1.25rem;margin-bottom:0}.input-text{width:100%;height:4.375rem;border-radius:.3125rem;box-shadow:none;border-width:.0625rem;border-style:solid;border-color:var(--rc-c3c3c3);padding:.5rem 0 0 .5rem;resize:none}.submit-button{width:100%;color:var(--white);background:var(--primary-color);height:2.5rem;border-radius:.25rem;border-style:solid;border-color:var(--primary-color);margin:.625rem 0 1rem;outline:0}@media only screen and (max-width:768px){.report-button{min-width:90vw;color:var(--white);background:var(--primary-color);height:2.5rem;border-radius:.25rem;border-style:solid;border-color:var(--primary-color);outline:0}p{margin:0;font-weight:400}.panel-body{padding:.5rem 1rem}#textareainfo{opacity:.5;position:absolute;bottom:-.8125rem;right:0;font-weight:400;font-size:.75rem}.initiate-email-info-button{left:0;right:0;position:fixed;top:80vh;padding:0 1rem}}@media only screen and (min-width:768px){.report-button{min-width:64rem;color:var(--white);background:var(--primary-color);height:2.5rem;border-radius:.25rem;border-style:solid;border-color:var(--primary-color);outline:0}}img>.report-button{padding-bottom:.25rem;padding-right:.0625rem}.send-email{text-align:center;padding:.9375rem;background:#f0f8ff;--padding-bottom:0px}.send-email-form{text-align:center;margin:0 1rem}#send-email-form{height:40vh;position:relative}.send-email-info{padding:.625rem 0;font-weight:400;font-size:.75rem;text-align:justify}#know-more-form{margin:0 1rem}@media only screen and (min-width:768px){.send-email-form{max-width:64.0625rem;margin:auto}.send-email-info{max-width:35.625rem;padding:.625rem 0!important;margin:left;float:left}.submit-button{width:9.375rem;float:right;color:var(--white);background:var(--primary-color);border-radius:0;border-style:solid;border-color:var(--primary-color);margin:.625rem 0 1rem;height:2rem;outline:0}.panel-info{overflow:hidden;text-align:left}.panel-info>h6,.panel-info>p{margin:0 0 0 .9375rem}#initiate-email{margin-top:.625rem}}.input-text-form{width:100%;height:8rem;border-radius:.3125rem;box-shadow:none;border-width:.0625rem;border-style:solid;border-color:var(--rc-c3c3c3);padding:.9375rem;resize:none}::-webkit-input-placeholder{color:var(--gray-200);font-size:.875rem;font-weight:400}::-moz-placeholder{color:var(--gray-200);font-size:.875rem;font-weight:400}:-ms-input-placeholder{color:var(--gray-200);font-size:.875rem;font-weight:400}::-ms-input-placeholder{color:var(--gray-200);font-size:.875rem;font-weight:400}::placeholder{color:var(--gray-200);font-size:.875rem;font-weight:400}a{color:var(--gray)}a:active,a:hover{text-decoration:none;color:none}.extraTemplate{display:flex;align-items:center;justify-content:center;margin-top:18px}"]
            }] }
];
/** @nocollapse */
FaqComponent.ctorParameters = () => [];
FaqComponent.propDecorators = {
    data: [{ type: Input }],
    appName: [{ type: Input }],
    extraTemplate: [{ type: Input }],
    toggleGroupEvent: [{ type: Output }],
    yesClickedEvent: [{ type: Output }],
    noClickedEvent: [{ type: Output }],
    submitClickedEvent: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FaqComponent.prototype.data;
    /** @type {?} */
    FaqComponent.prototype.appName;
    /** @type {?} */
    FaqComponent.prototype.extraTemplate;
    /** @type {?} */
    FaqComponent.prototype.toggleGroupEvent;
    /** @type {?} */
    FaqComponent.prototype.yesClickedEvent;
    /** @type {?} */
    FaqComponent.prototype.noClickedEvent;
    /** @type {?} */
    FaqComponent.prototype.submitClickedEvent;
    /** @type {?} */
    FaqComponent.prototype.constants;
    /** @type {?} */
    FaqComponent.prototype.faqs;
    /** @type {?} */
    FaqComponent.prototype.isNoClicked;
    /** @type {?} */
    FaqComponent.prototype.isYesClicked;
    /** @type {?} */
    FaqComponent.prototype.isSubmitted;
    /** @type {?} */
    FaqComponent.prototype.shownGroup;
    /** @type {?} */
    FaqComponent.prototype.value;
    /** @type {?} */
    FaqComponent.prototype.textValue;
}
