// common-form-elements/esm5/lib/option-group/option-group.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/option-group/option-group.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component } from '@angular/core';
var OptionGroupComponent = /** @class */ (function () {
    function OptionGroupComponent() {
    }
    /**
     * @return {?}
     */
    OptionGroupComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    OptionGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-option-group',
                    template: "<p>\n  option-group works!\n</p>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    OptionGroupComponent.ctorParameters = function () { return []; };
    return OptionGroupComponent;
}());
export { OptionGroupComponent };
