// @project-sunbird/common-consumption-v8/esm2015/lib/pipes-module/mime-type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
export class MimeTypeMasterData {
}
MimeTypeMasterData.COLLECTION = 'application/vnd.ekstep.content-collection';
MimeTypeMasterData.VIDEO = ['video/avi', 'video/mpeg', 'video/quicktime', 'video/3gpp', 'video/mpeg', 'video/mp4',
    'video/ogg', 'video/webm'];
MimeTypeMasterData.AUDIO = ['audio/mp3', 'audio/mp4', 'audio/mpeg', 'audio/ogg', 'audio/webm', 'audio/x-wav', 'audio/wav'];
MimeTypeMasterData.INTERACTIVE = ['application/vnd.ekstep.ecml-archive', 'application/vnd.ekstep.html-archive',
    'application/vnd.android.package-archive', 'application/vnd.ekstep.content-archive',
    'application/vnd.ekstep.plugin-archive', 'application/vnd.ekstep.h5p-archive'];
MimeTypeMasterData.DOCS = ['application/pdf', 'application/epub', 'application/msword'];
MimeTypeMasterData.ALL = ['video/mp4', 'video/x-youtube', 'video/webm', 'application/pdf', 'application/epub',
    'application/pdf', 'application/epub', 'application/vnd.ekstep.ecml-archive', 'application/vnd.ekstep.h5p-archive',
    'application/vnd.ekstep.html-archive'
];
if (false) {
    /** @type {?} */
    MimeTypeMasterData.COLLECTION;
    /** @type {?} */
    MimeTypeMasterData.VIDEO;
    /** @type {?} */
    MimeTypeMasterData.AUDIO;
    /** @type {?} */
    MimeTypeMasterData.INTERACTIVE;
    /** @type {?} */
    MimeTypeMasterData.DOCS;
    /** @type {?} */
    MimeTypeMasterData.ALL;
}
/*
  Contents are filtered based on given mimetype
*/
export class MimeTypePipe {
    /**
     * @param {?=} item
     * @param {?=} mimeTypes
     * @param {?=} isTextbookTocPage
     * @return {?}
     */
    transform(item, mimeTypes = ['all'], isTextbookTocPage = false) {
        if (mimeTypes.indexOf('all') > -1 && item != null) {
            if (item.mimeType !== MimeTypeMasterData.COLLECTION && !item.children) {
                return true;
            }
            else {
                if (item.mimeType === MimeTypeMasterData.COLLECTION && item.children && !item.children.length) {
                    return false;
                }
                if (item.children && item.children.length) {
                    return this.getFilteredItems(item.children, MimeTypeMasterData.ALL);
                }
                else {
                    return true;
                }
            }
        }
        if (item != null && item.mimeType !== MimeTypeMasterData.COLLECTION && (!item.children || !item.children.length)) {
            return this.getFilteredItems([item], mimeTypes);
        }
        if (item != null) {
            return this.getFilteredItems(item.children, mimeTypes);
        }
        else {
            return null;
        }
    }
    /**
     * @param {?} contents
     * @param {?} mimeTypes
     * @return {?}
     */
    getFilteredItems(contents, mimeTypes) {
        if (contents) {
            return this.flattenDeep(contents)
                .some((/**
             * @param {?} c
             * @return {?}
             */
            (c) => !!mimeTypes.find((/**
             * @param {?} m
             * @return {?}
             */
            m => m === c.mimeType))));
        }
    }
    /**
     * @private
     * @param {?} contents
     * @return {?}
     */
    flattenDeep(contents) {
        if (contents) {
            return contents.reduce((/**
             * @param {?} acc
             * @param {?} val
             * @return {?}
             */
            (acc, val) => {
                if (val.children) {
                    acc.push(val);
                    return acc.concat(this.flattenDeep(val.children));
                }
                else {
                    return acc.concat(val);
                }
            }), []);
        }
    }
}
MimeTypePipe.decorators = [
    { type: Pipe, args: [{
                name: 'hasMimeType',
            },] }
];
