// common-form-elements/esm5/lib/common-form-config.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/common-form-config.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var FieldConfigInputType = {
    INPUT: "input",
    CHECKBOX: "checkbox",
    SELECT: "select",
    LABEL: "label",
    TEXTAREA: "textarea",
    NESTED_SELECT: "nested_select",
    NESTED_GROUP: "nested_group",
    MULTIPLE_SELECT: "multiple_select",
};
export { FieldConfigInputType };
/** @enum {string} */
var FieldConfigValidationType = {
    REQUIRED: "required",
    MAXLENGTH: "maxLength",
    REQUIRED_TRUE: "required_true",
    MINLENGTH: "minLength",
    PATTERN: "pattern",
};
export { FieldConfigValidationType };
/**
 * @record
 * @template T
 */
export function FieldConfigOption() { }
if (false) {
    /** @type {?} */
    FieldConfigOption.prototype.label;
    /** @type {?} */
    FieldConfigOption.prototype.value;
    /** @type {?|undefined} */
    FieldConfigOption.prototype.extras;
}
/**
 * @record
 * @template T
 */
export function FieldConfigOptionAssociations() { }
/**
 * @record
 */
export function FieldConfigAsyncValidation() { }
if (false) {
    /** @type {?} */
    FieldConfigAsyncValidation.prototype.marker;
    /** @type {?|undefined} */
    FieldConfigAsyncValidation.prototype.message;
    /** @type {?|undefined} */
    FieldConfigAsyncValidation.prototype.trigger;
    /** @type {?|undefined} */
    FieldConfigAsyncValidation.prototype.asyncValidatorFactory;
}
/**
 * @record
 * @template T
 */
export function FieldConfig() { }
if (false) {
    /** @type {?} */
    FieldConfig.prototype.code;
    /** @type {?|undefined} */
    FieldConfig.prototype.type;
    /** @type {?|undefined} */
    FieldConfig.prototype.fieldName;
    /** @type {?|undefined} */
    FieldConfig.prototype.default;
    /** @type {?|undefined} */
    FieldConfig.prototype.context;
    /** @type {?|undefined} */
    FieldConfig.prototype.children;
    /** @type {?|undefined} */
    FieldConfig.prototype.templateOptions;
    /** @type {?|undefined} */
    FieldConfig.prototype.validations;
    /** @type {?|undefined} */
    FieldConfig.prototype.asyncValidation;
    /** @type {?|undefined} */
    FieldConfig.prototype.visible;
    /** @type {?|undefined} */
    FieldConfig.prototype.editable;
    /** @type {?|undefined} */
    FieldConfig.prototype.dataType;
    /** @type {?|undefined} */
    FieldConfig.prototype.renderingHints;
    /** @type {?|undefined} */
    FieldConfig.prototype.name;
    /** @type {?|undefined} */
    FieldConfig.prototype.description;
    /** @type {?|undefined} */
    FieldConfig.prototype.inputType;
    /** @type {?|undefined} */
    FieldConfig.prototype.index;
    /** @type {?|undefined} */
    FieldConfig.prototype.placeholder;
    /** @type {?|undefined} */
    FieldConfig.prototype.required;
    /** @type {?|undefined} */
    FieldConfig.prototype.label;
    /** @type {?|undefined} */
    FieldConfig.prototype.options;
    /** @type {?|undefined} */
    FieldConfig.prototype.collections;
    /** @type {?|undefined} */
    FieldConfig.prototype.value;
    /** @type {?|undefined} */
    FieldConfig.prototype.terms;
    /** @type {?|undefined} */
    FieldConfig.prototype.range;
    /** @type {?|undefined} */
    FieldConfig.prototype.depends;
    /** @type {?|undefined} */
    FieldConfig.prototype.dependencyTerms;
    /** @type {?|undefined} */
    FieldConfig.prototype.output;
    /** @type {?|undefined} */
    FieldConfig.prototype.sourceCategory;
    /** @type {?|undefined} */
    FieldConfig.prototype.association;
}
/** @enum {string} */
var FilterType = {
    FACET: "facet",
};
export { FilterType };
/**
 * @record
 */
export function Validator() { }
if (false) {
    /** @type {?|undefined} */
    Validator.prototype.name;
    /** @type {?|undefined} */
    Validator.prototype.validator;
    /** @type {?|undefined} */
    Validator.prototype.message;
    /** @type {?|undefined} */
    Validator.prototype.type;
    /** @type {?|undefined} */
    Validator.prototype.value;
}
/**
 * @record
 * @template T
 */
export function SectionConfig() { }
if (false) {
    /** @type {?} */
    SectionConfig.prototype.name;
    /** @type {?} */
    SectionConfig.prototype.fields;
}
