// common-form-elements/esm2015/lib/filters/filters.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/filters/filters.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { FilterType } from '../common-form-config';
export class FiltersComponent {
    constructor() {
        this.facets = [];
        this.selectedFilter = new EventEmitter();
        this.FilterType = FilterType;
        this.formattedFacets = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.formatFacets();
    }
    /**
     * @return {?}
     */
    formatFacets() {
        this.facets.forEach((/**
         * @param {?} facet
         * @param {?} index
         * @return {?}
         */
        (facet, index) => {
            /** @type {?} */
            let facetObj = {};
            facetObj['name'] = facet['name'];
            facetObj['type'] = this.config[facet['name']] && this.config[facet['name']]['type'] ? this.config[facet['name']]['type'] : 'dropdown';
            facetObj['index'] = this.config[facet['name']] && this.config[facet['name']]['index'] ? this.config[facet['name']]['index'] : index;
            facetObj['label'] = this.config[facet['name']] && this.config[facet['name']]['label'] ? this.config[facet['name']]['label'] : facet['name'];
            facetObj['placeHolder'] = this.config[facet['name']] && this.config[facet['name']]['placeHolder'] ? this.config[facet['name']]['placeHolder'] : 'Select ' + facet['name'];
            facetObj['default'] = this.config[facet['name']] && this.config[facet['name']]['default'] ? this.config[facet['name']]['default'] : '';
            facetObj['data'] = [];
            facet['values'].forEach((/**
             * @param {?} facetValue
             * @return {?}
             */
            facetValue => {
                facetObj['data'].push({
                    label: facetValue['name'],
                    value: facetValue['name']
                });
            }));
            this.formattedFacets.push(facetObj);
            this.formattedFacets.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            (a, b) => a.index - b.index));
        }));
    }
}
FiltersComponent.decorators = [
    { type: Component, args: [{
                selector: 'sb-filters',
                template: "<div *ngIf=\"type === FilterType.FACET\">\n  <div *ngFor=\"let facet of formattedFacets\">\n\n    <!-- Start of Dropdown -->\n    <sb-dropdown *ngIf=\"facet.type === 'dropdown'\" [type]=\"type\" [label]=\"facet.label\" [options]=\"facet\"\n      [placeHolder]=\"facet.placeHolder\" [default]=\"facet.default\" [styleClass]=\"styleClass\"\n      (onChangeFilter)=\"selectedFilter.emit($event)\">\n    </sb-dropdown>\n    <!-- End of Dropdown -->\n\n    <!-- Start of pills -->\n    <div *ngIf=\"facet.type === 'pills'\">\n      <sb-pills [label]=\"facet.label\" [options]=\"facet\" [styleClass]=\"styleClass\"\n        (onChangeFilter)=\"selectedFilter.emit($event)\"></sb-pills>\n    </div>\n    <!-- End of pills -->\n\n  </div>\n</div>",
                styles: [""]
            }] }
];
/** @nocollapse */
FiltersComponent.ctorParameters = () => [];
FiltersComponent.propDecorators = {
    type: [{ type: Input }],
    facets: [{ type: Input }],
    config: [{ type: Input }],
    styleClass: [{ type: Input }],
    selectedFilter: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FiltersComponent.prototype.type;
    /** @type {?} */
    FiltersComponent.prototype.facets;
    /** @type {?} */
    FiltersComponent.prototype.config;
    /** @type {?} */
    FiltersComponent.prototype.styleClass;
    /** @type {?} */
    FiltersComponent.prototype.selectedFilter;
    /** @type {?} */
    FiltersComponent.prototype.FilterType;
    /** @type {?} */
    FiltersComponent.prototype.formGroup;
    /** @type {?} */
    FiltersComponent.prototype.formattedFacets;
}
