// @project-sunbird/common-consumption-v8/esm5/lib/pipes-module/mime-type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
var MimeTypeMasterData = /** @class */ (function () {
    function MimeTypeMasterData() {
    }
    MimeTypeMasterData.COLLECTION = 'application/vnd.ekstep.content-collection';
    MimeTypeMasterData.VIDEO = ['video/avi', 'video/mpeg', 'video/quicktime', 'video/3gpp', 'video/mpeg', 'video/mp4',
        'video/ogg', 'video/webm'];
    MimeTypeMasterData.AUDIO = ['audio/mp3', 'audio/mp4', 'audio/mpeg', 'audio/ogg', 'audio/webm', 'audio/x-wav', 'audio/wav'];
    MimeTypeMasterData.INTERACTIVE = ['application/vnd.ekstep.ecml-archive', 'application/vnd.ekstep.html-archive',
        'application/vnd.android.package-archive', 'application/vnd.ekstep.content-archive',
        'application/vnd.ekstep.plugin-archive', 'application/vnd.ekstep.h5p-archive'];
    MimeTypeMasterData.DOCS = ['application/pdf', 'application/epub', 'application/msword'];
    MimeTypeMasterData.ALL = ['video/mp4', 'video/x-youtube', 'video/webm', 'application/pdf', 'application/epub',
        'application/pdf', 'application/epub', 'application/vnd.ekstep.ecml-archive', 'application/vnd.ekstep.h5p-archive',
        'application/vnd.ekstep.html-archive'
    ];
    return MimeTypeMasterData;
}());
export { MimeTypeMasterData };
if (false) {
    /** @type {?} */
    MimeTypeMasterData.COLLECTION;
    /** @type {?} */
    MimeTypeMasterData.VIDEO;
    /** @type {?} */
    MimeTypeMasterData.AUDIO;
    /** @type {?} */
    MimeTypeMasterData.INTERACTIVE;
    /** @type {?} */
    MimeTypeMasterData.DOCS;
    /** @type {?} */
    MimeTypeMasterData.ALL;
}
/*
  Contents are filtered based on given mimetype
*/
var MimeTypePipe = /** @class */ (function () {
    function MimeTypePipe() {
    }
    /**
     * @param {?=} item
     * @param {?=} mimeTypes
     * @param {?=} isTextbookTocPage
     * @return {?}
     */
    MimeTypePipe.prototype.transform = /**
     * @param {?=} item
     * @param {?=} mimeTypes
     * @param {?=} isTextbookTocPage
     * @return {?}
     */
    function (item, mimeTypes, isTextbookTocPage) {
        if (mimeTypes === void 0) { mimeTypes = ['all']; }
        if (isTextbookTocPage === void 0) { isTextbookTocPage = false; }
        if (mimeTypes.indexOf('all') > -1 && item != null) {
            if (item.mimeType !== MimeTypeMasterData.COLLECTION && !item.children) {
                return true;
            }
            else {
                if (item.mimeType === MimeTypeMasterData.COLLECTION && item.children && !item.children.length) {
                    return false;
                }
                if (item.children && item.children.length) {
                    return this.getFilteredItems(item.children, MimeTypeMasterData.ALL);
                }
                else {
                    return true;
                }
            }
        }
        if (item != null && item.mimeType !== MimeTypeMasterData.COLLECTION && (!item.children || !item.children.length)) {
            return this.getFilteredItems([item], mimeTypes);
        }
        if (item != null) {
            return this.getFilteredItems(item.children, mimeTypes);
        }
        else {
            return null;
        }
    };
    /**
     * @param {?} contents
     * @param {?} mimeTypes
     * @return {?}
     */
    MimeTypePipe.prototype.getFilteredItems = /**
     * @param {?} contents
     * @param {?} mimeTypes
     * @return {?}
     */
    function (contents, mimeTypes) {
        if (contents) {
            return this.flattenDeep(contents)
                .some((/**
             * @param {?} c
             * @return {?}
             */
            function (c) { return !!mimeTypes.find((/**
             * @param {?} m
             * @return {?}
             */
            function (m) { return m === c.mimeType; })); }));
        }
    };
    /**
     * @private
     * @param {?} contents
     * @return {?}
     */
    MimeTypePipe.prototype.flattenDeep = /**
     * @private
     * @param {?} contents
     * @return {?}
     */
    function (contents) {
        var _this = this;
        if (contents) {
            return contents.reduce((/**
             * @param {?} acc
             * @param {?} val
             * @return {?}
             */
            function (acc, val) {
                if (val.children) {
                    acc.push(val);
                    return acc.concat(_this.flattenDeep(val.children));
                }
                else {
                    return acc.concat(val);
                }
            }), []);
        }
    };
    MimeTypePipe.decorators = [
        { type: Pipe, args: [{
                    name: 'hasMimeType',
                },] }
    ];
    return MimeTypePipe;
}());
export { MimeTypePipe };
