// @project-sunbird/common-consumption-v8/esm5/lib/accordion/accordion.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { Subject } from 'rxjs';
var AccordionComponent = /** @class */ (function () {
    function AccordionComponent() {
        this.openCloseItems = new Subject();
        this.isHiidden = true;
        this.cardClick = new EventEmitter();
        this.multiValue = false;
    }
    Object.defineProperty(AccordionComponent.prototype, "multi", {
        get: /**
         * @return {?}
         */
        function () {
            return this.multiValue;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this.multiValue = convertToBoolProperty(val);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    AccordionComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * Opens all enabled accordion items.
     */
    /**
     * Opens all enabled accordion items.
     * @return {?}
     */
    AccordionComponent.prototype.openAll = /**
     * Opens all enabled accordion items.
     * @return {?}
     */
    function () {
        if (this.multi) {
            this.openCloseItems.next(false);
        }
    };
    /**
     * Closes all enabled accordion items.
     */
    /**
     * Closes all enabled accordion items.
     * @return {?}
     */
    AccordionComponent.prototype.closeAll = /**
     * Closes all enabled accordion items.
     * @return {?}
     */
    function () {
        this.openCloseItems.next(true);
    };
    AccordionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-accordion',
                    template: "<div class=\"sb-accordion-item\">\r\n<ng-content select=\"sb-accordion-item\"></ng-content>\r\n</div>",
                    styles: ["::ng-deep .expanded .sbaccordion{margin-bottom:1.5rem}"]
                }] }
    ];
    /** @nocollapse */
    AccordionComponent.ctorParameters = function () { return []; };
    AccordionComponent.propDecorators = {
        multi: [{ type: Input, args: ['multi',] }],
        isHiidden: [{ type: Input }],
        cardClick: [{ type: Output }]
    };
    return AccordionComponent;
}());
export { AccordionComponent };
if (false) {
    /** @type {?} */
    AccordionComponent.prototype.openCloseItems;
    /** @type {?} */
    AccordionComponent.prototype.isHiidden;
    /** @type {?} */
    AccordionComponent.prototype.cardClick;
    /**
     * @type {?}
     * @private
     */
    AccordionComponent.prototype.multiValue;
}
/**
 * @param {?} val
 * @return {?}
 */
export function convertToBoolProperty(val) {
    if (typeof val === 'string') {
        val = val.toLowerCase().trim();
        return (val === 'true' || val === '');
    }
    return !!val;
}
