// @project-sunbird/common-consumption-v8/esm2015/lib/pills-grid/models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const PillShape = {
    CIRCLE: 'circle',
    BOX: 'box',
    IMAGE_OVERLAP: 'image_overlap',
    DEFAULT: 'default',
};
export { PillShape };
/** @enum {string} */
const PillsViewType = {
    SCROLL: 'scroll',
    NONE: 'none',
};
export { PillsViewType };
/** @enum {string} */
const PillBorder = {
    ROUNDED: 'rounded',
    NONE: 'none',
};
export { PillBorder };
/** @enum {string} */
const SelectMode = {
    SINGLE: 'single',
    MULTIPLE: 'multiple',
};
export { SelectMode };
/** @enum {string} */
const PillsMultiRow = {
    DEFAULT: 'default',
    SINGLE: 'single',
    DOUBLE_VIEW_ROW: 'double_view_row',
    DOUBLE_VIEW_COLUMN: 'double_view_column',
};
export { PillsMultiRow };
/** @enum {string} */
const ShowMoreViewType = {
    EXPAND: 'expand',
    NEW_SCREEN: 'new_screen',
};
export { ShowMoreViewType };
/**
 * @record
 */
export function PillsColorTheme() { }
if (false) {
    /** @type {?|undefined} */
    PillsColorTheme.prototype.iconBgColor;
    /** @type {?|undefined} */
    PillsColorTheme.prototype.pillBgColor;
}
