// @project-sunbird/common-consumption-v8/esm5/lib/layout/course-cards-hlist/course-cards-hlist.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { courseSection } from './course-section-data';
import { myCourses } from './my-courses-data';
import { CourseCardGridTypes, CourseCardTypes, ButtonPosition } from '../../card/models';
var CourseCardsHlistComponent = /** @class */ (function () {
    function CourseCardsHlistComponent() {
        /* Title for the grid */
        this.title = courseSection.name;
        /* Max card count to be shown */
        this.maxCardCount = 10;
        this.viewMoreButtonText = 'View More';
        /* Show Menu on each card */
        this.isMenu = false;
        this.viewMoreButtonPosition = ButtonPosition.TOP;
        this.viewMoreClick = new EventEmitter();
        this.cardClick = new EventEmitter();
        this.menuClick = new EventEmitter();
    }
    Object.defineProperty(CourseCardsHlistComponent.prototype, "ButtonPosition", {
        get: /**
         * @return {?}
         */
        function () { return ButtonPosition; },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    CourseCardsHlistComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.type != null) {
            if (this.type == CourseCardGridTypes.MY_COURSE_CARD_GRID ||
                this.type == CourseCardGridTypes.MY_COURSE_RECENTLY_VIEWED_CARD_GRID) {
                if (!(this.contentList != null)) {
                    this.contentList = myCourses.courses;
                }
            }
            else {
                if (!(this.contentList != null)) {
                    this.contentList = courseSection.contents;
                }
            }
        }
        else {
            this.type = CourseCardGridTypes.COURSE_CARD_GRID;
            if (!(this.contentList != null)) {
                this.contentList = courseSection.contents;
            }
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    CourseCardsHlistComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
    };
    Object.defineProperty(CourseCardsHlistComponent.prototype, "CourseCardGridTypes", {
        get: /**
         * @return {?}
         */
        function () { return CourseCardGridTypes; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CourseCardsHlistComponent.prototype, "CourseCardTypes", {
        get: /**
         * @return {?}
         */
        function () { return CourseCardTypes; },
        enumerable: true,
        configurable: true
    });
    /**
     * Triggers event on `View More` Click
     * @param event HTML click event
     */
    /**
     * Triggers event on `View More` Click
     * @param {?} event HTML click event
     * @return {?}
     */
    CourseCardsHlistComponent.prototype.onViewMoreClick = /**
     * Triggers event on `View More` Click
     * @param {?} event HTML click event
     * @return {?}
     */
    function (event) {
        this.viewMoreClick.emit({ event: event, data: this.contentList });
    };
    /**
     * Triggers event on card click
     * @param event HTML Click event
     * @param data Content data for selected card
     */
    /**
     * Triggers event on card click
     * @param {?} event HTML Click event
     * @param {?} data Content data for selected card
     * @return {?}
     */
    CourseCardsHlistComponent.prototype.onCardClick = /**
     * Triggers event on card click
     * @param {?} event HTML Click event
     * @param {?} data Content data for selected card
     * @return {?}
     */
    function (event, data) {
        this.cardClick.emit({ event: event, data: data });
    };
    /**
     * @param {?} maxCardCounter
     * @return {?}
     */
    CourseCardsHlistComponent.prototype.range = /**
     * @param {?} maxCardCounter
     * @return {?}
     */
    function (maxCardCounter) {
        return new Array(maxCardCounter);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CourseCardsHlistComponent.prototype.onCardMenuClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.menuClick.emit(event);
    };
    CourseCardsHlistComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-course-cards-hlist',
                    template: "<div class=\"layout-course\" *ngIf=\"!isLoading\">\r\n    <div class=\"header\">\r\n        <div class=\"title\">\r\n            {{title}}\r\n            <hr>\r\n        </div>\r\n        <button class=\"sb-btn sb-btn-normal sb-btn-outline-secondary sb-btn-white\"\r\n            *ngIf=\"contentList?.length > maxCardCount && viewMoreButtonPosition === ButtonPosition.TOP\"\r\n            (click)=\"onViewMoreClick($event)\">\r\n            {{viewMoreButtonText}}\r\n        </button>\r\n    </div>\r\n    <div class=\"sb--scrolling-wrapper-flexbox\" *ngIf=\"!multiRow\">\r\n      <div  *ngFor=\"let content of contentList | slice:0:(maxCardCount ? maxCardCount : contentList?.length); let index = index\" class=\"sb--card--container\">\r\n        <sb-course-card [course]=\"content\" (click)=\"onCardClick($event, content)\"  [cardImg]=\"content?.cardImg\" [section]=\"content?.info_to_display\" *ngIf=\"type === CourseCardGridTypes.COURSE_CARD_GRID\" [isMenu]=\"isMenu\"  (menuClick)=\"onCardMenuClick($event)\"></sb-course-card>\r\n        <sb-my-course-card [hideProgress]=\"hideProgress\" [course]=\"content\" (click)=\"onCardClick($event, content)\" [cardImg]=\"content?.cardImg\" [section]=\"content?.info_to_display\" *ngIf=\"type === CourseCardGridTypes.MY_COURSE_CARD_GRID\"></sb-my-course-card>\r\n        <sb-my-course-card [hideProgress]=\"hideProgress\" [course]=\"content\" (click)=\"onCardClick($event, content)\" [cardImg]=\"content?.cardImg\" [section]=\"content?.info_to_display\" [type]=\"CourseCardTypes.RECENTLY_VIEWED\" *ngIf=\"type === CourseCardGridTypes.MY_COURSE_RECENTLY_VIEWED_CARD_GRID\"></sb-my-course-card>\r\n        <sb-course-curiculum-card [cardData]=\"content\" [index]=\"index\" (click)=\"onCardClick($event, content)\" *ngIf=\"type === CourseCardGridTypes.CURICULUM_COURSE_CARD_GRID\"></sb-course-curiculum-card>\r\n      </div>\r\n      <div class=\"view-all\">\r\n        <button class=\"sb-btn sb-btn-normal sb-btn-outline-secondary sb-btn-white sb-btn-oval\" *ngIf=\"contentList?.length > maxCardCount && viewMoreButtonPosition === ButtonPosition.RIGHT\" (click)=\"onViewMoreClick($event)\">\r\n          {{viewMoreButtonText}}\r\n        </button>\r\n      </div>\r\n    </div>\r\n    <div class=\"sb--scrolling-wrapper-flexbox\" *ngIf=\"multiRow\">\r\n      <span *ngFor=\"let content of contentList | slice:0:(maxCardCount ? maxCardCount : contentList?.length); let index = index\">\r\n        <div *ngIf=\"index%2 == 0\"  class=\"sb--card--container\">\r\n            <sb-course-card class=\"sb--card--seperator\" *ngIf=\"contentList[index]!=null && type === CourseCardGridTypes.COURSE_CARD_GRID\" [course]=\"contentList[index]\" (click)=\"onCardClick($event, contentList[index])\"  [cardImg]=\"contentList[index]?.cardImg\" [section]=\"contentList[index]?.info_to_display\" [isMenu]=\"isMenu\"  (menuClick)=\"onCardMenuClick($event)\"></sb-course-card>\r\n            <sb-course-card class=\"sb--card--seperator\" *ngIf=\"index+1 < contentList.length && type === CourseCardGridTypes.COURSE_CARD_GRID\" [course]=\"contentList[index+1]\" (click)=\"onCardClick($event, contentList[index+1])\"  [cardImg]=\"contentList[index+1]?.cardImg\" [section]=\"contentList[index+1]?.info_to_display\"></sb-course-card>\r\n\r\n            <sb-my-course-card [hideProgress]=\"hideProgress\" class=\"sb--card--seperator\" *ngIf=\"contentList[index]!=null && type === CourseCardGridTypes.MY_COURSE_CARD_GRID\" [course]=\"contentList[index]\" (click)=\"onCardClick($event, contentList[index])\" [cardImg]=\"contentList[index]?.cardImg\" [section]=\"contentList[index]?.info_to_display\"></sb-my-course-card>\r\n            <sb-my-course-card [hideProgress]=\"hideProgress\" class=\"sb--card--seperator\" *ngIf=\"index+1 < contentList.length && type === CourseCardGridTypes.MY_COURSE_CARD_GRID\" [course]=\"contentList[index+1]\" (click)=\"onCardClick($event, contentList[index+1])\" [cardImg]=\"contentList[index+1]?.cardImg\" [section]=\"contentList[index+1]?.info_to_display\"></sb-my-course-card>\r\n            \r\n            <sb-my-course-card [hideProgress]=\"hideProgress\" class=\"sb--card--seperator\" *ngIf=\"contentList[index]!=null && type === CourseCardGridTypes.MY_COURSE_RECENTLY_VIEWED_CARD_GRID\" [course]=\"contentList[index]\" (click)=\"onCardClick($event, contentList[index])\" [cardImg]=\"contentList[index]?.cardImg\" [section]=\"contentList[index]?.info_to_display\" [type]=\"CourseCardTypes.RECENTLY_VIEWED\"></sb-my-course-card>\r\n            <sb-my-course-card [hideProgress]=\"hideProgress\" class=\"sb--card--seperator\" *ngIf=\"index+1 < contentList.length && type === CourseCardGridTypes.MY_COURSE_RECENTLY_VIEWED_CARD_GRID\" [course]=\"contentList[index+1]\" (click)=\"onCardClick($event, contentList[index+1])\" [cardImg]=\"contentList[index+1]?.cardImg\" [section]=\"contentList[index+1]?.info_to_display\" [type]=\"CourseCardTypes.RECENTLY_VIEWED\"></sb-my-course-card>\r\n        \r\n            <sb-course-curiculum-card class=\"sb--card--seperator\" [index]=\"index\" [cardData]=\"contentList[index]\" (click)=\"onCardClick($event, contentList[index])\" *ngIf=\"contentList[index]!=null && type === CourseCardGridTypes.CURICULUM_COURSE_CARD_GRID\"></sb-course-curiculum-card>\r\n            <sb-course-curiculum-card class=\"sb--card--seperator\" [index]=\"index+1\" [cardData]=\"contentList[index+1]\" (click)=\"onCardClick($event, contentList[index+1])\" *ngIf=\"index+1 < contentList.length && contentList[index+1]!=null && type === CourseCardGridTypes.CURICULUM_COURSE_CARD_GRID\"></sb-course-curiculum-card>\r\n       </div>\r\n      </span>\r\n      <div class=\"view-all\">\r\n        <button class=\"sb-btn sb-btn-normal sb-btn-outline-secondary sb-btn-white sb-btn-oval\" *ngIf=\"contentList?.length > maxCardCount && viewMoreButtonPosition === ButtonPosition.RIGHT\" (click)=\"onViewMoreClick($event)\">\r\n          {{viewMoreButtonText}}\r\n        </button>\r\n      </div>\r\n    </div>\r\n</div>\r\n<div class=\"layout-course\" *ngIf=\"isLoading\">\r\n    <div class=\"header\" style=\"width:120px;\">\r\n        <div class=\"title loading-animation sb--card--highlighter\">\r\n            \r\n        </div>\r\n    </div>\r\n    <div class=\"sb--scrolling-wrapper-flexbox\" *ngIf=\"maxCardCount > 0 && !multiRow\">\r\n        <div *ngFor=\"let content of range(maxCardCount); let i=index\" class=\"sb--card--container\">\r\n          <sb-course-card [isLoading]=\"isLoading\" *ngIf=\"type === CourseCardGridTypes.COURSE_CARD_GRID\"></sb-course-card>\r\n          <sb-my-course-card [hideProgress]=\"hideProgress\" [isLoading]=\"isLoading\" *ngIf=\"type === CourseCardGridTypes.MY_COURSE_CARD_GRID\"></sb-my-course-card>\r\n          <sb-my-course-card [hideProgress]=\"hideProgress\" [isLoading]=\"isLoading\" [type]=\"CourseCardTypes.RECENTLY_VIEWED\" *ngIf=\"type === CourseCardGridTypes.MY_COURSE_RECENTLY_VIEWED_CARD_GRID\"></sb-my-course-card>\r\n          <sb-course-curiculum-card [index]=\"i\" [isLoading]=\"isLoading\" *ngIf=\"type === CourseCardGridTypes.CURICULUM_COURSE_CARD_GRID\"></sb-course-curiculum-card>\r\n        </div>\r\n      </div>\r\n      <div class=\"sb--scrolling-wrapper-flexbox\" *ngIf=\"maxCardCount > 0 && multiRow\">\r\n          <div *ngFor=\"let content of range(maxCardCount); let i=index\" class=\"sb--card--container\">\r\n              <sb-course-card [isLoading]=\"isLoading\" *ngIf=\"type === CourseCardGridTypes.COURSE_CARD_GRID\"></sb-course-card>\r\n              <sb-course-card [isLoading]=\"isLoading\" *ngIf=\"type === CourseCardGridTypes.COURSE_CARD_GRID\"></sb-course-card>\r\n              <sb-my-course-card [hideProgress]=\"hideProgress\" [isLoading]=\"isLoading\" *ngIf=\"type === CourseCardGridTypes.MY_COURSE_CARD_GRID\"></sb-my-course-card>\r\n              <sb-my-course-card [hideProgress]=\"hideProgress\" [isLoading]=\"isLoading\" *ngIf=\"type === CourseCardGridTypes.MY_COURSE_CARD_GRID\"></sb-my-course-card>\r\n              <sb-my-course-card [hideProgress]=\"hideProgress\" [isLoading]=\"isLoading\" [type]=\"CourseCardTypes.RECENTLY_VIEWED\" *ngIf=\"type === CourseCardGridTypes.MY_COURSE_RECENTLY_VIEWED_CARD_GRID\"></sb-my-course-card>\r\n              <sb-my-course-card [hideProgress]=\"hideProgress\" [isLoading]=\"isLoading\" [type]=\"CourseCardTypes.RECENTLY_VIEWED\" *ngIf=\"type === CourseCardGridTypes.MY_COURSE_RECENTLY_VIEWED_CARD_GRID\"></sb-my-course-card>\r\n              <sb-course-curiculum-card [index]=\"i\" [isLoading]=\"isLoading\" *ngIf=\"type === CourseCardGridTypes.CURICULUM_COURSE_CARD_GRID\"></sb-course-curiculum-card>\r\n              <sb-course-curiculum-card [index]=\"i+1\" [isLoading]=\"isLoading\" *ngIf=\"type === CourseCardGridTypes.CURICULUM_COURSE_CARD_GRID\"></sb-course-curiculum-card>\r\n            </div>\r\n      </div>\r\n\r\n</div>",
                    styles: [".loading-animation{-webkit-animation-duration:1s;animation-duration:1s;-webkit-animation-fill-mode:forwards;animation-fill-mode:forwards;-webkit-animation-iteration-count:infinite;animation-iteration-count:infinite;-webkit-animation-name:placeload;animation-name:placeload;-webkit-animation-timing-function:linear;animation-timing-function:linear;background:var(--lazyload-bg);background:var(--grey-nuance-lighter);background:linear-gradient(to right,var(--grey-nuance-lighter) 8%,var(--grey-nuance-darker) 18%,var(--grey-nuance-lighter) 33%);background-size:75rem 6.5rem;position:relative}@-webkit-keyframes placeload{0%{background-position:-29.25rem 0}100%{background-position:29.25rem 0}}@keyframes placeload{0%{background-position:-29.25rem 0}100%{background-position:29.25rem 0}}.sb--card--highlighter{height:.75rem;width:100%;margin-top:.3125rem}:host .layout-library{display:flex;flex-direction:column}:host .sb--scrolling-wrapper-flexbox{display:flex;flex-wrap:nowrap;overflow-x:auto;-webkit-overflow-scrolling:touch;padding-bottom:.5em}:host .sb--card--container{flex:0 0 auto;width:17.5rem;margin:0 .5rem 0 0}html[dir=rtl] :host .sb--card--container{margin-right:inherit;margin-left:1rem}:host .header{display:flex;flex-direction:row;padding:0;align-items:center}[dir=ltr] :host .header,[dir=rtl] :host .header{margin:.7rem 0 .5rem}:host .header .title{font-family:inherit;text-overflow:ellipsis;overflow:hidden;flex:1;height:1.5rem;color:var(--secondary-color);font-size:1.125rem;font-weight:700;text-transform:capitalize;line-height:1.625rem;display:flex;align-items:center}:host .header .title hr{flex:1;margin-left:1rem;display:none}:host .sb-btn-oval{border-radius:50%;padding:0;height:80%;top:10%}"]
                }] }
    ];
    /** @nocollapse */
    CourseCardsHlistComponent.ctorParameters = function () { return []; };
    CourseCardsHlistComponent.propDecorators = {
        title: [{ type: Input }],
        contentList: [{ type: Input }],
        maxCardCount: [{ type: Input }],
        viewMoreButtonText: [{ type: Input }],
        isLoading: [{ type: Input }],
        type: [{ type: Input }],
        multiRow: [{ type: Input }],
        hideProgress: [{ type: Input }],
        isMenu: [{ type: Input }],
        viewMoreButtonPosition: [{ type: Input }],
        viewMoreClick: [{ type: Output }],
        cardClick: [{ type: Output }],
        menuClick: [{ type: Output }]
    };
    return CourseCardsHlistComponent;
}());
export { CourseCardsHlistComponent };
if (false) {
    /** @type {?} */
    CourseCardsHlistComponent.prototype.title;
    /** @type {?} */
    CourseCardsHlistComponent.prototype.contentList;
    /** @type {?} */
    CourseCardsHlistComponent.prototype.maxCardCount;
    /** @type {?} */
    CourseCardsHlistComponent.prototype.viewMoreButtonText;
    /** @type {?} */
    CourseCardsHlistComponent.prototype.isLoading;
    /** @type {?} */
    CourseCardsHlistComponent.prototype.type;
    /** @type {?} */
    CourseCardsHlistComponent.prototype.multiRow;
    /** @type {?} */
    CourseCardsHlistComponent.prototype.hideProgress;
    /** @type {?} */
    CourseCardsHlistComponent.prototype.isMenu;
    /** @type {?} */
    CourseCardsHlistComponent.prototype.viewMoreButtonPosition;
    /** @type {?} */
    CourseCardsHlistComponent.prototype.viewMoreClick;
    /** @type {?} */
    CourseCardsHlistComponent.prototype.cardClick;
    /** @type {?} */
    CourseCardsHlistComponent.prototype.menuClick;
}
