// @project-sunbird/common-consumption-v8/esm2015/lib/layout/library-cards-stack/library-cards-stack.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { defaultLibraryCardsGrid } from '../library-cards.data';
import { LibraryCardTypes } from '../../card/models';
export class LibraryCardsStackComponent {
    constructor() {
        this.contentList = defaultLibraryCardsGrid.contentList;
        this.isOffline = false;
        this.viewMoreButtonText = defaultLibraryCardsGrid.viewMoreButtonText;
        this.maxCardCount = defaultLibraryCardsGrid.maxCardCount;
        this.cardType = 'textbook';
        this.viewMoreClick = new EventEmitter();
        this.cardClick = new EventEmitter();
        this.menuClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get LibraryCardTypes() { return LibraryCardTypes; }
    /**
     * @param {?} event
     * @return {?}
     */
    showAllList(event) {
        this.viewMoreClick.emit(event);
    }
    /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    emitCardClick(event, index) {
        this.cardClick.emit({ 'data': event.data, 'index': index });
    }
    /**
     * @param {?} event
     * @return {?}
     */
    emitMenuClick(event) {
        this.menuClick.emit(event);
    }
}
LibraryCardsStackComponent.decorators = [
    { type: Component, args: [{
                selector: 'sb-library-cards-stack',
                template: "<div class=\"card-list-container\">\r\n\r\n    <div class=\"header\" *ngIf=\"!isLoading && title\">\r\n        <div class=\"title\">\r\n            <i class=\"icon-svg icon-svg--sm\">\r\n                <svg class=\"icon\">\r\n                    <use xlink:href=\"assets/common-consumption/images/sprite.svg#book\"></use>\r\n                </svg>\r\n            </i>\r\n            {{title}}\r\n        </div>\r\n    </div>\r\n\r\n    <!--Loading Indicator for header-->\r\n    <div class=\"header\" *ngIf=\"isLoading\">\r\n            <div class=\"loading-animation sb--card--highlighter\" style=\"width:100px;height:20px;\">\r\n                    <i class=\"icon-svg icon-svg--sm\"></i>\r\n            </div>\r\n    </div>\r\n\r\n    <div class=\"body\">\r\n        <ng-container *ngIf=\"cardType==='textbook' || layoutConfig\">\r\n            <sb-library-card\r\n                *ngFor=\"let content of contentList | slice:0:(maxCardCount ? maxCardCount : contentList?.length); let i= index\" [isOffline]=\"!content?.isAvailableLocally && isOffline\"\r\n                [content]=\"content\" [type]=\"LibraryCardTypes.MOBILE_TEXTBOOK\" [isLoading]=\"isLoading\" (cardClick)=\"emitCardClick($event, i)\" [cardImg]=\"content?.cardImg\"\r\n                (menuClick)=\"emitMenuClick($event)\" [isMenu]=\"isMenu\" [layoutConfig]=\"layoutConfig\">\r\n            </sb-library-card>\r\n            <button *ngIf=\"maxCardCount && contentList?.length >maxCardCount\" class=\"sbbtn flat stack-btn full-width\"\r\n                (click)=\"showAllList($event)\">{{viewMoreButtonText}}</button>\r\n        </ng-container>\r\n        <ng-container *ngIf=\"cardType==='course' && !layoutConfig\">\r\n            <sb-course-card *ngFor=\"let content of contentList | slice:0:(maxCardCount ? maxCardCount : contentList?.length); let i= index\"\r\n                [section]=\"null\" [course]=\"content\" [cardImg]=\"content?.cardImg\" (cardClick)=\"emitCardClick($event, i)\" [isMenu]=\"isMenu\"\r\n                (menuClick)=\"emitMenuClick($event)\">\r\n            </sb-course-card>\r\n            <button *ngIf=\"maxCardCount && contentList?.length >maxCardCount\" class=\"sbbtn flat stack-btn full-width\"\r\n                (click)=\"showAllList($event)\">{{viewMoreButtonText}}</button>\r\n        </ng-container>\r\n    </div>\r\n\r\n</div>",
                styles: [".loading-animation{-webkit-animation-duration:1s;animation-duration:1s;-webkit-animation-fill-mode:forwards;animation-fill-mode:forwards;-webkit-animation-iteration-count:infinite;animation-iteration-count:infinite;-webkit-animation-name:placeload;animation-name:placeload;-webkit-animation-timing-function:linear;animation-timing-function:linear;background:var(--lazyload-bg);background:var(--grey-nuance-lighter);background:linear-gradient(to right,var(--grey-nuance-lighter) 8%,var(--grey-nuance-darker) 18%,var(--grey-nuance-lighter) 33%);background-size:75rem 6.5rem;position:relative}@-webkit-keyframes placeload{0%{background-position:-29.25rem 0}100%{background-position:29.25rem 0}}@keyframes placeload{0%{background-position:-29.25rem 0}100%{background-position:29.25rem 0}}.sb--card--highlighter{height:.75rem;width:100%;margin-top:.3125rem}:host .header{display:flex;flex-direction:row;padding:.5rem 0;margin-bottom:.5rem}:host .header .title{font-size:1.125rem;line-height:1.625rem;font-weight:700;color:var(--primary-color);text-overflow:ellipsis;text-transform:capitalize;overflow:hidden;flex:1;display:flex;align-items:center}:host .header .title i{vertical-align:middle;height:1.0625rem;width:.875rem;top:0}[dir=ltr] :host .header .title i{margin-right:.5rem}[dir=rtl] :host .header .title i{margin-left:.5rem}:host .body{box-shadow:0 .125rem .4375rem 0 rgba(var(--rc-rgba-black),.16)}:host .stack-btn{background-color:var(--primary-100);padding:1rem;color:var(--primary-color);font-weight:700;border-style:none;text-transform:uppercase;vertical-align:middle;line-height:1;width:100%}@media (max-width:768px){:host .sb--card{margin-top:.0625rem}:host .sb--card:last-child{box-shadow:none}}"]
            }] }
];
LibraryCardsStackComponent.propDecorators = {
    contentList: [{ type: Input }],
    title: [{ type: Input }],
    isOffline: [{ type: Input }],
    viewMoreButtonText: [{ type: Input }],
    maxCardCount: [{ type: Input }],
    isLoading: [{ type: Input }],
    isMenu: [{ type: Input }],
    layoutConfig: [{ type: Input }],
    cardType: [{ type: Input }],
    viewMoreClick: [{ type: Output }],
    cardClick: [{ type: Output }],
    menuClick: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    LibraryCardsStackComponent.prototype.contentList;
    /** @type {?} */
    LibraryCardsStackComponent.prototype.title;
    /** @type {?} */
    LibraryCardsStackComponent.prototype.isOffline;
    /** @type {?} */
    LibraryCardsStackComponent.prototype.viewMoreButtonText;
    /** @type {?} */
    LibraryCardsStackComponent.prototype.maxCardCount;
    /** @type {?} */
    LibraryCardsStackComponent.prototype.isLoading;
    /** @type {?} */
    LibraryCardsStackComponent.prototype.isMenu;
    /** @type {?} */
    LibraryCardsStackComponent.prototype.layoutConfig;
    /** @type {?} */
    LibraryCardsStackComponent.prototype.cardType;
    /** @type {?} */
    LibraryCardsStackComponent.prototype.viewMoreClick;
    /** @type {?} */
    LibraryCardsStackComponent.prototype.cardClick;
    /** @type {?} */
    LibraryCardsStackComponent.prototype.menuClick;
}
