// @project-sunbird/common-consumption-v8/esm5/lib/card/notification-card/notification-card.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { notificationData } from './notification-card-data';
import { NotificationStatus } from '../models';
var NotificationCardComponent = /** @class */ (function () {
    function NotificationCardComponent() {
        this.notification = notificationData;
        this.notificationClick = new EventEmitter();
        this.deleteNotificationClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    NotificationCardComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.notification['isRead'] = (this.notification.status === NotificationStatus.READ);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NotificationCardComponent.prototype.notificationClickHandler = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var eventData = {
            event: event,
            data: this.notification
        };
        this.notificationClick.emit(eventData);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NotificationCardComponent.prototype.notificationDeleteHandler = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var eventData = {
            event: event,
            data: this.notification
        };
        this.deleteNotificationClick.emit(eventData);
    };
    NotificationCardComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sb-notification-card',
                    template: "<div class=\"sb-notification-list p-8\" (click)=\"notificationClickHandler($event)\">\r\n  <div class=\"sb-notification-list-header d-flex flex-ai-center flex-jc-space-between\">\r\n    <div class=\"notification-info d-flex flex-ai-center mb-8\">\r\n      <div class=\"status\" *ngIf=\"!notification?.isRead\"></div>\r\n      <div class=\"date ml-8\">{{notification?.createdOn | date:'E, d MMMM h:mm'}}</div>\r\n    </div>\r\n    <div class=\"delete-icon\" (click)=\"notificationDeleteHandler($event);$event.stopPropagation()\">\r\n      <img class=\"Delete-gray\" src=\"assets/common-consumption/images/Delete-gray.svg\" alt=\"\" alt=\"delete\" width=\"14px\">\r\n      <img class=\"Delete-red\" src=\"assets/common-consumption/images/Delete-red.svg\" alt=\"\" alt=\"delete\" width=\"14px\">\r\n    </div>\r\n  </div>\r\n  <div class=\"sb-notification-data d-flex flex-ai-center\">\r\n    <div class=\"logo\" *ngIf=\"notification?.data?.actionData?.appIcon\">\r\n      <img [src]=\"notification?.data?.actionData?.appIcon\" width=\"20px\">\r\n    </div>\r\n    <div class=\"ml-12 notification-content\">\r\n      <div class=\"title\" *ngIf=\"notification?.data?.actionData?.title\" [ngClass]=\"{'active': !notification?.isRead}\">{{notification?.data?.actionData?.title}}</div>\r\n      <div class=\"sub-title py-8\" *ngIf=\"notification?.data?.actionData?.description\">{{notification?.data?.actionData?.description}}</div>\r\n    </div>\r\n  </div>\r\n</div>\r\n",
                    styles: [".sb-notification-list{border-radius:.5rem;background-color:var(--white);box-shadow:.3125rem .3125rem .125rem 0 rgba(var(--rc-rgba-black),.1);margin-bottom:.5rem;border:.03125rem solid #ddd}.sb-notification-list:hover{cursor:pointer}.sb-notification-list:hover .sb-notification-list-header .notification-info .status{background-color:var(--orange)}.sb-notification-list:hover .sb-notification-list-header .delete-icon .Delete-red{display:block}.sb-notification-list:hover .sb-notification-list-header .delete-icon .Delete-gray{display:none}.sb-notification-list .sb-notification-list-header .notification-info .status{height:.5rem;width:.5rem;background-color:var(--gray-200);border-radius:.5rem}.sb-notification-list .sb-notification-list-header .notification-info .date{color:#979797;font-size:.75rem}.sb-notification-list .sb-notification-list-header .delete-icon .Delete-red{display:none}.sb-notification-list .sb-notification-data .logo{height:2rem;width:2rem;border-radius:.5rem;background-color:var(--gray-0);display:inline-flex;align-items:center;justify-content:center}.sb-notification-list .sb-notification-data .logo img{height:1.5rem;max-width:1.25rem}.sb-notification-list .sb-notification-data .notification-content{overflow:hidden}.sb-notification-list .sb-notification-data .notification-content .title{color:var(--gray-400);font-size:.875rem;font-weight:700;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}.sb-notification-list .sb-notification-data .notification-content .title.active{color:var(--primary-color)}.sb-notification-list .sb-notification-data .notification-content .sub-title{color:var(--gray-400);font-size:.75rem;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}"]
                }] }
    ];
    /** @nocollapse */
    NotificationCardComponent.ctorParameters = function () { return []; };
    NotificationCardComponent.propDecorators = {
        notification: [{ type: Input }],
        notificationClick: [{ type: Output }],
        deleteNotificationClick: [{ type: Output }]
    };
    return NotificationCardComponent;
}());
export { NotificationCardComponent };
if (false) {
    /** @type {?} */
    NotificationCardComponent.prototype.notification;
    /** @type {?} */
    NotificationCardComponent.prototype.notificationClick;
    /** @type {?} */
    NotificationCardComponent.prototype.deleteNotificationClick;
}
