// common-form-elements/esm2015/lib/dynamic-form/dynamic-form.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic-form/dynamic-form.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { FieldConfigInputType } from '../common-form-config';
import { FormBuilder, Validators } from '@angular/forms';
import { Subject } from 'rxjs';
import { tap } from 'rxjs/operators';
import * as _ from 'lodash-es';
export class DynamicFormComponent {
    /**
     * @param {?} formBuilder
     */
    constructor(formBuilder) {
        this.formBuilder = formBuilder;
        this.initialize = new EventEmitter();
        this.finalize = new EventEmitter();
        this.dataLoadStatusDelegate = new Subject();
        this.valueChanges = new EventEmitter();
        this.statusChanges = new EventEmitter();
        this._ = _;
        this.FieldConfigInputType = FieldConfigInputType;
        this.isSection = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const formGroupData = {};
        /** @type {?} */
        const dependency = [];
        if (changes['config']) {
            if ((changes['config'].currentValue && changes['config'].firstChange)
                || changes['config'].previousValue !== changes['config'].currentValue) {
                this.initialize.emit(this.formGroup);
            }
        }
        if (this.statusChangesSubscription) {
            this.statusChangesSubscription.unsubscribe();
        }
        if (this.valueChangesSubscription) {
            this.valueChangesSubscription.unsubscribe();
        }
        this.isSection = !_.isEmpty(_.find(this.config, 'fields'));
        if (this.isSection) {
            this.config.forEach((/**
             * @param {?} sections
             * @return {?}
             */
            (sections) => {
                sections.fields.forEach((/**
                 * @param {?} element
                 * @param {?} index
                 * @return {?}
                 */
                (element, index) => {
                    /** @type {?} */
                    const formValueList = this.prepareFormValidationData(element, index);
                    if (!_.isEmpty(element.depends)) {
                        dependency.push({ code: element.code, depends: element.depends });
                    }
                    formGroupData[element.code] = formValueList;
                }));
            }));
        }
        else {
            /** @type {?} */
            let defaultSection = [];
            defaultSection = [
                {
                    'name': '',
                    'fields': _.cloneDeep(this.config)
                }
            ];
            this.config = _.cloneDeep(defaultSection);
            defaultSection.forEach((/**
             * @param {?} sections
             * @return {?}
             */
            (sections) => {
                sections.fields.forEach((/**
                 * @param {?} element
                 * @param {?} index
                 * @return {?}
                 */
                (element, index) => {
                    /** @type {?} */
                    const formValueList = this.prepareFormValidationData(element, index);
                    if (!_.isEmpty(element.depends)) {
                        dependency.push({ code: element.code, depends: element.depends });
                    }
                    formGroupData[element.code] = formValueList;
                }));
            }));
        }
        this.flattenSectionFields = this.getFlattenedSectionFields();
        this.formGroup = this.formBuilder.group(formGroupData);
        this.statusChangesSubscription = this.formGroup.valueChanges.pipe(tap((/**
         * @param {?} v
         * @return {?}
         */
        (v) => {
            this.statusChanges.emit({
                isPristine: this.formGroup.pristine,
                isDirty: this.formGroup.dirty,
                isInvalid: this.formGroup.invalid,
                isValid: this.formGroup.valid,
                controls: this.getFormValidationErrors()
            });
        }))).subscribe();
        this.valueChangesSubscription = this.formGroup.valueChanges.pipe(tap((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            this.valueChanges.emit(data);
        }))).subscribe();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.finalize.emit();
        if (this.statusChangesSubscription) {
            this.statusChangesSubscription.unsubscribe();
        }
        if (this.valueChangesSubscription) {
            this.valueChangesSubscription.unsubscribe();
        }
    }
    /**
     * @return {?}
     */
    getFormValidationErrors() {
        /** @type {?} */
        const errors = [];
        _.keys(this.formGroup.controls).forEach((/**
         * @param {?} key
         * @return {?}
         */
        key => {
            /** @type {?} */
            const controlErrors = this.formGroup.get(key).errors;
            if (controlErrors != null) {
                _.keys(controlErrors).forEach((/**
                 * @param {?} keyError
                 * @return {?}
                 */
                keyError => {
                    errors.push({
                        control_name: key,
                        error_name: keyError,
                        error_value: controlErrors[keyError]
                    });
                }));
            }
        }));
        return errors;
    }
    /**
     * @private
     * @param {?} element
     * @param {?} index
     * @return {?}
     */
    prepareFormValidationData(element, index) {
        /** @type {?} */
        const formValueList = [];
        /** @type {?} */
        const validationList = [];
        /** @type {?} */
        let defaultVal = '';
        switch (element.inputType) {
            case 'text':
                defaultVal = element.default || null;
                break;
            case 'textarea':
                defaultVal = element.default || null;
                break;
            case 'select':
                if (element.default) {
                    if (element.dataType === 'list') {
                        if (_.isArray(element.default)) {
                            defaultVal = element.default;
                        }
                        else {
                            defaultVal = _.toArray(element.default);
                        }
                    }
                    else if (element.dataType === 'text') {
                        if (_.isString(element.default)) {
                            defaultVal = element.default;
                        }
                        else {
                            defaultVal = _.toString(element.default);
                        }
                    }
                }
                else {
                    defaultVal = null;
                }
                break;
            case 'multiselect':
                if (element.default) {
                    if (element.dataType === 'list' && _.isArray(element.default)) {
                        defaultVal = element.default;
                    }
                    else if (element.dataType === 'list' && _.isString(element.default)) {
                        if (_.includes(element.default, ',')) {
                            defaultVal = _.split(element.default, ',');
                        }
                        else {
                            defaultVal = [element.default];
                        }
                    }
                    else if (element.dataType === 'text') {
                        if (_.includes(element.default, ',')) {
                            defaultVal = _.split(element.default, ',');
                        }
                        else {
                            defaultVal = [element.default];
                        }
                    }
                }
                else {
                    defaultVal = [];
                }
                break;
            case 'nestedselect':
                defaultVal = element.dataType === 'list' ?
                    (element.default && Array.isArray(element.default) ? element.default :
                        _.isEmpty(element.default) ? [] : [element.default]) :
                    (element.default || null);
                break;
            case 'checkbox':
                defaultVal = (element.dataType === 'text') ? (element.default === 'Yes' ? 'Yes' : 'No') : !!element.default;
                break;
        }
        formValueList.push(defaultVal);
        if (element.validations && element.validations.length) {
            element.validations.forEach((/**
             * @param {?} data
             * @param {?} i
             * @return {?}
             */
            (data, i) => {
                switch (data.type) {
                    case 'required':
                        if (element.inputType === 'select' || element.inputType === 'multiselect' || element.inputType === 'nestedselect') {
                            validationList.push(Validators.required);
                        }
                        else if (element.type === 'checkbox') {
                            validationList.push(Validators.requiredTrue);
                        }
                        else {
                            validationList.push(Validators.required);
                        }
                        break;
                    case 'pattern':
                        validationList.push(Validators.pattern((/** @type {?} */ (element.validations[i].value))));
                        break;
                    case 'min':
                        validationList.push(Validators.minLength((/** @type {?} */ (element.validations[i].value))));
                        break;
                    case 'max':
                        validationList.push(Validators.maxLength((/** @type {?} */ (element.validations[i].value))));
                        break;
                }
            }));
        }
        formValueList.push(Validators.compose(validationList));
        return formValueList;
    }
    /**
     * @param {?} config
     * @param {?} context
     * @return {?}
     */
    fetchContextTerms(config, context) {
        return _.get(_.find(config, { 'code': context }), 'terms') || null;
    }
    /**
     * @param {?} code
     * @param {?} depends
     * @return {?}
     */
    getAllDependsFormControl(code, depends) {
        /** @type {?} */
        const fieldDepends = {};
        _.forEach(depends, (/**
         * @param {?} depend
         * @return {?}
         */
        depend => {
            if (this.formGroup.get(depend)) {
                fieldDepends[depend] = this.formGroup.get(depend);
            }
        }));
        return fieldDepends || null;
    }
    /**
     * @param {?} code
     * @param {?} depends
     * @return {?}
     */
    fetchDependencyTerms(code, depends) {
        /** @type {?} */
        const dependsTerms = _.map(_.filter(this.flattenSectionFields, (/**
         * @param {?} c
         * @return {?}
         */
        c => {
            return _.includes(depends, c.code);
        })), (/**
         * @param {?} depend
         * @return {?}
         */
        (depend) => {
            return depend.terms || depend.range;
        }));
        return _.flatten(dependsTerms);
    }
    /**
     * @param {?} config
     * @param {?} val
     * @return {?}
     */
    getAppIcon(config, val) {
        if (val) {
            return config.filter((/**
             * @param {?} field
             * @return {?}
             */
            field => {
                return field.code === 'appicon';
            }));
        }
        else {
            return config.filter((/**
             * @param {?} field
             * @return {?}
             */
            field => {
                return field.code !== 'appicon';
            }));
        }
    }
    /**
     * @param {?} config
     * @return {?}
     */
    groupBySection(config) {
        /** @type {?} */
        const fields = this.getAppIcon(config, false);
        return _.groupBy(fields, 'section.index');
    }
    /**
     * @return {?}
     */
    getFlattenedSectionFields() {
        return _.flatten(_.map(this.config, 'fields'));
    }
}
DynamicFormComponent.decorators = [
    { type: Component, args: [{
                selector: 'sb-dynamic-form',
                template: "\n<div class=\"dynamic-form\" [formGroup]=\"formGroup\">\n  <div *ngFor=\"let section of config\" class=\"formSection grid two-column-grid\">\n      <ng-container *ngFor=\"let field of section['fields']\" sbDynamicField\n      [depends]=\"getAllDependsFormControl(field.code, field.depends)\" [dependencyTerms]=\"field.depends ? fetchDependencyTerms(field.code, field.depends) : []\"\n      [context]=\"field.context ? formGroup.get(field.context) : null\" [contextTerms]=\"fetchContextTerms(config, field.context)\"\n      [options]=\"field.terms\" [field]=\"field\" [formGroup]=\"formGroup\" [default]=\"field.default || ''\" [label]=\"field.label\" [formControlRef]=\"formGroup.get(field.code)\"\n      [dataLoadStatusDelegate]=\"dataLoadStatusDelegate\" [placeholder]=\"field.placeholder || ''\" [validations]=\"field.validations || []\"\n      [disabled]=\"field.editable === true ? false : true\">\n    </ng-container>\n  </div>\n</div>\n\n",
                styles: [""]
            }] }
];
/** @nocollapse */
DynamicFormComponent.ctorParameters = () => [
    { type: FormBuilder }
];
DynamicFormComponent.propDecorators = {
    config: [{ type: Input }],
    initialize: [{ type: Output }],
    finalize: [{ type: Output }],
    dataLoadStatusDelegate: [{ type: Input }],
    valueChanges: [{ type: Output }],
    statusChanges: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DynamicFormComponent.prototype.config;
    /** @type {?} */
    DynamicFormComponent.prototype.initialize;
    /** @type {?} */
    DynamicFormComponent.prototype.finalize;
    /** @type {?} */
    DynamicFormComponent.prototype.dataLoadStatusDelegate;
    /** @type {?} */
    DynamicFormComponent.prototype.valueChanges;
    /** @type {?} */
    DynamicFormComponent.prototype.statusChanges;
    /**
     * @type {?}
     * @private
     */
    DynamicFormComponent.prototype.statusChangesSubscription;
    /**
     * @type {?}
     * @private
     */
    DynamicFormComponent.prototype.valueChangesSubscription;
    /** @type {?} */
    DynamicFormComponent.prototype._;
    /** @type {?} */
    DynamicFormComponent.prototype.formGroup;
    /** @type {?} */
    DynamicFormComponent.prototype.FieldConfigInputType;
    /** @type {?} */
    DynamicFormComponent.prototype.fieldDependency;
    /** @type {?} */
    DynamicFormComponent.prototype.isSection;
    /** @type {?} */
    DynamicFormComponent.prototype.flattenSectionFields;
    /**
     * @type {?}
     * @private
     */
    DynamicFormComponent.prototype.formBuilder;
}
